--- a/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
+++ b/net/minecraft/world/entity/ai/behavior/InteractWithDoor.java
@@ -60,8 +60,15 @@
                         })) {
                             DoorBlock doorblock = (DoorBlock) blockstate.getBlock();
 
-                            if (!doorblock.isOpen(blockstate)) {
-                                doorblock.setOpen(livingentity, serverlevel, blockstate, blockpos, true);
+                            if (!blockdoor.isOpen(iblockdata)) {
+                                // CraftBukkit start - entities opening doors
+                                org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(entityliving.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(entityliving.level(), blockposition));
+                                entityliving.level().getCraftServer().getPluginManager().callEvent(event);
+                                if (event.isCancelled()) {
+                                    return false;
+                                }
+                                // CraftBukkit end
+                                blockdoor.setOpen(entityliving, worldserver, iblockdata, blockposition, true);
                             }
 
                             optional = rememberDoorToClose(memoryaccessor1, optional, serverlevel, blockpos);
@@ -75,9 +82,16 @@
                         })) {
                             DoorBlock doorblock1 = (DoorBlock) blockstate1.getBlock();
 
-                            if (!doorblock1.isOpen(blockstate1)) {
-                                doorblock1.setOpen(livingentity, serverlevel, blockstate1, blockpos1, true);
-                                optional = rememberDoorToClose(memoryaccessor1, optional, serverlevel, blockpos1);
+                            if (!blockdoor1.isOpen(iblockdata1)) {
+                                // CraftBukkit start - entities opening doors
+                                org.bukkit.event.entity.EntityInteractEvent event = new org.bukkit.event.entity.EntityInteractEvent(entityliving.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(entityliving.level(), blockposition1));
+                                entityliving.level().getCraftServer().getPluginManager().callEvent(event);
+                                if (event.isCancelled()) {
+                                    return false;
+                                }
+                                // CraftBukkit end
+                                blockdoor1.setOpen(entityliving, worldserver, iblockdata1, blockposition1, true);
+                                optional = rememberDoorToClose(memoryaccessor1, optional, worldserver, blockposition1);
                             }
                         }
 
