--- a/net/minecraft/world/item/enchantment/DamageEnchantment.java
+++ b/net/minecraft/world/item/enchantment/DamageEnchantment.java
@@ -3,13 +3,14 @@
 import net.minecraft.world.effect.MobEffectInstance;
 import net.minecraft.world.effect.MobEffects;
 import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.EnumMonsterType;
 import net.minecraft.world.entity.EquipmentSlot;
 import net.minecraft.world.entity.LivingEntity;
-import net.minecraft.world.entity.MobType;
 import net.minecraft.world.item.AxeItem;
 import net.minecraft.world.item.ItemStack;
 
 public class DamageEnchantment extends Enchantment {
+
     public static final int ALL = 0;
     public static final int UNDEAD = 1;
     public static final int ARTHROPODS = 2;
@@ -26,12 +27,12 @@
 
     @Override
     public int getMinCost(int enchantmentLevel) {
-        return MIN_COST[this.type] + (enchantmentLevel - 1) * LEVEL_COST[this.type];
+        return DamageEnchantment.MIN_COST[this.type] + (enchantmentLevel - 1) * DamageEnchantment.LEVEL_COST[this.type];
     }
 
     @Override
     public int getMaxCost(int enchantmentLevel) {
-        return this.getMinCost(enchantmentLevel) + LEVEL_COST_SPAN[this.type];
+        return this.getMinCost(enchantmentLevel) + DamageEnchantment.LEVEL_COST_SPAN[this.type];
     }
 
     @Override
@@ -40,14 +41,8 @@
     }
 
     @Override
-    public float getDamageBonus(int level, MobType creatureType) {
-        if (this.type == 0) {
-            return 1.0F + (float)Math.max(0, level - 1) * 0.5F;
-        } else if (this.type == 1 && creatureType == MobType.UNDEAD) {
-            return (float)level * 2.5F;
-        } else {
-            return this.type == 2 && creatureType == MobType.ARTHROPOD ? (float)level * 2.5F : 0.0F;
-        }
+    public float getDamageBonus(int level, EnumMonsterType creatureType) {
+        return this.type == 0 ? 1.0F + (float) Math.max(0, level - 1) * 0.5F : (this.type == 1 && creatureType == EnumMonsterType.UNDEAD ? (float) level * 2.5F : (this.type == 2 && creatureType == EnumMonsterType.ARTHROPOD ? (float) level * 2.5F : 0.0F));
     }
 
     @Override
@@ -57,14 +52,20 @@
 
     @Override
     public boolean canEnchant(ItemStack stack) {
-        return stack.getItem() instanceof AxeItem || super.canEnchant(stack);
+        return stack.getItem() instanceof AxeItem ? true : super.canEnchant(stack);
     }
 
     @Override
     public void doPostAttack(LivingEntity user, Entity target, int level) {
-        if (target instanceof LivingEntity livingEntity && this.type == 2 && level > 0 && livingEntity.getMobType() == MobType.ARTHROPOD) {
-            int i = 20 + user.getRandom().nextInt(10 * level);
-            livingEntity.addEffect(new MobEffectInstance(MobEffects.MOVEMENT_SLOWDOWN, i, 3));
+        if (target instanceof LivingEntity) {
+            LivingEntity entityliving1 = (LivingEntity) target;
+
+            if (this.type == 2 && level > 0 && entityliving1.getMobType() == EnumMonsterType.ARTHROPOD) {
+                int j = 20 + user.getRandom().nextInt(10 * level);
+
+                entityliving1.addEffect(new MobEffectInstance(MobEffects.MOVEMENT_SLOWDOWN, j, 3), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
+            }
         }
+
     }
 }
