--- a/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
@@ -23,13 +22,26 @@
         this.states = new BlockState[size];
         int i = 0;
 
-        for (short s : positions) {
-            this.positions[i] = s;
-            this.states[i] = section.getBlockState(SectionPos.sectionRelativeX(s), SectionPos.sectionRelativeY(s), SectionPos.sectionRelativeZ(s));
-            i++;
+        this.positions = new short[i];
+        this.states = new IBlockData[i];
+        int j = 0;
+
+        for (ShortIterator shortiterator = positions.iterator(); shortiterator.hasNext(); ++j) {
+            short short0 = (Short) shortiterator.next();
+
+            this.positions[j] = short0;
+            this.states[j] = (section != null) ? section.getBlockState(SectionPos.sectionRelativeX(short0), SectionPos.sectionRelativeY(short0), SectionPos.sectionRelativeZ(short0)) : net.minecraft.world.level.block.Blocks.AIR.defaultBlockState(); // CraftBukkit - SPIGOT-6076, Mojang bug when empty chunk section notified
         }
     }
 
+    // CraftBukkit start - Add constructor
+    public ClientboundSectionBlocksUpdatePacket(SectionPos sectionposition, ShortSet shortset, IBlockData[] states) {
+        this.sectionPos = sectionposition;
+        this.positions = shortset.toShortArray();
+        this.states = states;
+    }
+    // CraftBukkit end
+
     public ClientboundSectionBlocksUpdatePacket(FriendlyByteBuf buffer) {
         this.sectionPos = SectionPos.of(buffer.readLong());
         int varInt = buffer.readVarInt();
