--- a/net/minecraft/world/level/storage/loot/functions/LootingEnchantFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/LootingEnchantFunction.java
@@ -60,9 +57,14 @@
 
         if (entity instanceof LivingEntity) {
             int i = EnchantmentHelper.getMobLooting((LivingEntity) entity);
+            // CraftBukkit start - use lootingModifier if set by plugin
+            if (context.hasParam(LootContextParams.LOOTING_MOD)) {
+                i = context.getParamOrNull(LootContextParams.LOOTING_MOD);
+            }
+            // CraftBukkit end
 
-            if (i == 0) {
-                return itemstack;
+            if (i <= 0) { // CraftBukkit - account for possible negative looting values from Bukkit
+                return stack;
             }
 
             float f = (float) i * this.value.getFloat(lootcontext);
