--- a/net/minecraft/world/effect/HealOrHarmMobEffect.java
+++ b/net/minecraft/world/effect/HealOrHarmMobEffect.java
@@ -16,7 +17,7 @@
     public void applyEffectTick(LivingEntity livingEntity, int amplifier) {
         super.applyEffectTick(livingEntity, amplifier);
         if (this.isHarm == livingEntity.isInvertedHealAndHarm()) {
-            livingEntity.heal((float)Math.max(4 << amplifier, 0));
+            livingEntity.heal((float) Math.max(4 << amplifier, 0), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
             livingEntity.hurt(livingEntity.damageSources().magic(), (float)(6 << amplifier));
         }
@@ -25,8 +29,8 @@
     @Override
     public void applyInstantenousEffect(@Nullable Entity source, @Nullable Entity indirectSource, LivingEntity livingEntity, int amplifier, double health) {
         if (this.isHarm == livingEntity.isInvertedHealAndHarm()) {
-            int i = (int)(health * (double)(4 << amplifier) + 0.5);
-            livingEntity.heal((float)i);
+            j = (int) (health * (double) (4 << amplifier) + 0.5D);
+            livingEntity.heal((float) j, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
             int i = (int)(health * (double)(6 << amplifier) + 0.5);
             if (source == null) {
