--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -104,10 +111,15 @@
                 return;
             }
 
-            for (Bee bee : entitiesOfClass) {
-                if (bee.getTarget() == null) {
-                    Player player = Util.getRandom(entitiesOfClass1, level.random);
-                    bee.setTarget(player);
+            Iterator iterator = list.iterator();
+
+            while (iterator.hasNext()) {
+                Bee entitybee = (Bee) iterator.next();
+
+                if (entitybee.getTarget() == null) {
+                    Player entityhuman = (Player) Util.getRandom(list1, level.random);
+
+                    entitybee.setTarget(entityhuman, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER, true); // CraftBukkit
                 }
             }
         }
