--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -10,6 +11,9 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
+// CraftBukkit start
+import org.bukkit.event.block.CauldronLevelChangeEvent;
+// CraftBukkit end
 
 public class CauldronBlock extends AbstractCauldronBlock {
     public static final MapCodec<CauldronBlock> CODEC = simpleCodec(CauldronBlock::new);
@@ -57,14 +63,12 @@
     @Override
     protected void receiveStalactiteDrip(BlockState state, Level level, BlockPos pos, Fluid fluid) {
         if (fluid == Fluids.WATER) {
-            BlockState blockState = Blocks.WATER_CAULDRON.defaultBlockState();
-            level.setBlockAndUpdate(pos, blockState);
-            level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(blockState));
+            iblockdata1 = Blocks.WATER_CAULDRON.defaultBlockState();
+            LayeredCauldronBlock.changeLevel(state, level, pos, iblockdata1, null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
             level.levelEvent(1047, pos, 0);
         } else if (fluid == Fluids.LAVA) {
-            BlockState blockState = Blocks.LAVA_CAULDRON.defaultBlockState();
-            level.setBlockAndUpdate(pos, blockState);
-            level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(blockState));
+            iblockdata1 = Blocks.LAVA_CAULDRON.defaultBlockState();
+            LayeredCauldronBlock.changeLevel(state, level, pos, iblockdata1, null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
             level.levelEvent(1046, pos, 0);
         }
     }
