--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -22,6 +22,8 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
+
 public class ChorusFlowerBlock extends Block {
     public static final MapCodec<ChorusFlowerBlock> CODEC = RecordCodecBuilder.mapCodec(
         instance -> instance.group(
@@ -95,9 +103,13 @@
                     flag = true;
                 }
 
-                if (flag && allNeighborsEmpty(level, blockPos, null) && level.isEmptyBlock(pos.above(2))) {
-                    level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
-                    this.placeGrownFlower(level, blockPos, i);
+                if (flag && allNeighborsEmpty(level, blockposition1, (Direction) null) && level.isEmptyBlock(pos.above(2))) {
+                    // CraftBukkit start - add event
+                    if (CraftEventFactory.handleBlockSpreadEvent(level, pos, blockposition1, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(i)), 2)) {
+                        level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
+                        this.placeGrownFlower(level, blockposition1, i);
+                    }
+                    // CraftBukkit end
                 } else if (i < 4) {
                     int i1 = random.nextInt(4);
                     if (flag1) {
@@ -106,24 +118,35 @@
 
                     boolean flag2 = false;
 
-                    for (int i3 = 0; i3 < i1; i3++) {
-                        Direction randomDirection = Direction.Plane.HORIZONTAL.getRandomDirection(random);
-                        BlockPos blockPos1 = pos.relative(randomDirection);
-                        if (level.isEmptyBlock(blockPos1)
-                            && level.isEmptyBlock(blockPos1.below())
-                            && allNeighborsEmpty(level, blockPos1, randomDirection.getOpposite())) {
-                            this.placeGrownFlower(level, blockPos1, i + 1);
-                            flag2 = true;
+                    for (int l = 0; l < j; ++l) {
+                        Direction enumdirection = Direction.Plane.HORIZONTAL.getRandomDirection(random);
+                        BlockPos blockposition2 = pos.relative(enumdirection);
+
+                        if (level.isEmptyBlock(blockposition2) && level.isEmptyBlock(blockposition2.below()) && allNeighborsEmpty(level, blockposition2, enumdirection.getOpposite())) {
+                            // CraftBukkit start - add event
+                            if (CraftEventFactory.handleBlockSpreadEvent(level, pos, blockposition2, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(i + 1)), 2)) {
+                                this.placeGrownFlower(level, blockposition2, i + 1);
+                                flag2 = true;
+                            }
+                            // CraftBukkit end
                         }
                     }
 
                     if (flag2) {
                         level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
                     } else {
-                        this.placeDeadFlower(level, pos);
+                        // CraftBukkit start - add event
+                        if (CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(5)), 2)) {
+                            this.placeDeadFlower(level, pos);
+                        }
+                        // CraftBukkit end
                     }
                 } else {
-                    this.placeDeadFlower(level, pos);
+                    // CraftBukkit start - add event
+                    if (CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(5)), 2)) {
+                        this.placeDeadFlower(level, pos);
+                    }
+                    // CraftBukkit end
                 }
             }
         }
@@ -249,10 +279,16 @@
     }
 
     @Override
-    public void onProjectileHit(Level level, BlockState state, BlockHitResult hit, Projectile projectile) {
-        BlockPos blockPos = hit.getBlockPos();
-        if (!level.isClientSide && projectile.mayInteract(level, blockPos) && projectile.mayBreak(level)) {
-            level.destroyBlock(blockPos, true, projectile);
+    public void onProjectileHit(Level level, IBlockData state, BlockHitResult hit, Projectile projectile) {
+        BlockPos blockposition = hit.getBlockPos();
+
+        if (!level.isClientSide && projectile.mayInteract(level, blockposition) && projectile.mayBreak(level)) {
+            // CraftBukkit
+            if (!CraftEventFactory.callEntityChangeBlockEvent(projectile, blockposition, Blocks.AIR.defaultBlockState())) {
+                return;
+            }
+            // CraftBukkit end
+            level.destroyBlock(blockposition, true, projectile);
         }
     }
 }
