--- a/net/minecraft/world/level/block/DragonEggBlock.java
+++ b/net/minecraft/world/level/block/DragonEggBlock.java
@@ -16,6 +16,7 @@
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockFromToEvent; // CraftBukkit
 
 public class DragonEggBlock extends FallingBlock {
     public static final MapCodec<DragonEggBlock> CODEC = simpleCodec(DragonEggBlock::new);
@@ -49,13 +51,22 @@
     private void teleport(BlockState state, Level level, BlockPos pos) {
         WorldBorder worldBorder = level.getWorldBorder();
 
-        for (int i = 0; i < 1000; i++) {
-            BlockPos blockPos = pos.offset(
-                level.random.nextInt(16) - level.random.nextInt(16),
-                level.random.nextInt(8) - level.random.nextInt(8),
-                level.random.nextInt(16) - level.random.nextInt(16)
-            );
-            if (level.getBlockState(blockPos).isAir() && worldBorder.isWithinBounds(blockPos)) {
+        for (int i = 0; i < 1000; ++i) {
+            BlockPos blockposition1 = pos.offset(level.random.nextInt(16) - level.random.nextInt(16), level.random.nextInt(8) - level.random.nextInt(8), level.random.nextInt(16) - level.random.nextInt(16));
+
+            if (level.getBlockState(blockposition1).isAir() && worldborder.isWithinBounds(blockposition1)) {
+                // CraftBukkit start
+                org.bukkit.block.Block from = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+                org.bukkit.block.Block to = level.getWorld().getBlockAt(blockposition1.getX(), blockposition1.getY(), blockposition1.getZ());
+                BlockFromToEvent event = new BlockFromToEvent(from, to);
+                org.bukkit.Bukkit.getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+
+                blockposition1 = new BlockPos(event.getToBlock().getX(), event.getToBlock().getY(), event.getToBlock().getZ());
+                // CraftBukkit end
                 if (level.isClientSide) {
                     for (int i1 = 0; i1 < 128; i1++) {
                         double randomDouble = level.random.nextDouble();
