--- a/net/minecraft/world/level/block/NetherWartBlock.java
+++ b/net/minecraft/world/level/block/NetherWartBlock.java
@@ -56,8 +53,8 @@
     public void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         int i = state.getValue(AGE);
         if (i < 3 && random.nextInt(10) == 0) {
-            state = state.setValue(AGE, Integer.valueOf(i + 1));
-            level.setBlock(pos, state, 2);
+            state = (IBlockData) state.setValue(NetherWartBlock.AGE, i + 1);
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, state, 2); // CraftBukkit
         }
     }
 
