--- a/net/minecraft/world/level/block/SignBlock.java
+++ b/net/minecraft/world/level/block/SignBlock.java
@@ -103,10 +117,8 @@
                     return InteractionResult.SUCCESS;
                 } else if (flag1) {
                     return InteractionResult.SUCCESS;
-                } else if (!this.otherPlayerIsEditingSign(player, signBlockEntity)
-                    && player.mayBuild()
-                    && this.hasEditableText(player, signBlockEntity, isFacingFrontText)) {
-                    this.openTextEdit(player, signBlockEntity, isFacingFrontText);
+                } else if (!this.otherPlayerIsEditingSign(player, tileentitysign) && player.mayBuild() && this.hasEditableText(player, tileentitysign, flag1)) {
+                    this.openTextEdit(player, tileentitysign, flag1, org.bukkit.event.player.PlayerSignOpenEvent.Cause.INTERACT); // CraftBukkit
                     return this.getInteractionResult(flag);
                 } else {
                     return InteractionResult.PASS;
