--- a/net/minecraft/world/level/block/CauldronBlock.java
+++ b/net/minecraft/world/level/block/CauldronBlock.java
@@ -11,6 +11,9 @@
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.Fluids;
+// CraftBukkit start
+import org.bukkit.event.block.CauldronLevelChangeEvent;
+// CraftBukkit end
 
 public class CauldronBlock extends AbstractCauldronBlock {
 
@@ -65,15 +63,13 @@
         BlockState blockstate1;
 
         if (fluid == Fluids.WATER) {
-            blockstate1 = Blocks.WATER_CAULDRON.defaultBlockState();
-            level.setBlockAndUpdate(blockpos, blockstate1);
-            level.gameEvent(GameEvent.BLOCK_CHANGE, blockpos, GameEvent.Context.of(blockstate1));
-            level.levelEvent(1047, blockpos, 0);
+            iblockdata1 = Blocks.WATER_CAULDRON.defaultBlockState();
+            LayeredCauldronBlock.changeLevel(state, level, pos, iblockdata1, null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
+            level.levelEvent(1047, pos, 0);
         } else if (fluid == Fluids.LAVA) {
-            blockstate1 = Blocks.LAVA_CAULDRON.defaultBlockState();
-            level.setBlockAndUpdate(blockpos, blockstate1);
-            level.gameEvent(GameEvent.BLOCK_CHANGE, blockpos, GameEvent.Context.of(blockstate1));
-            level.levelEvent(1046, blockpos, 0);
+            iblockdata1 = Blocks.LAVA_CAULDRON.defaultBlockState();
+            LayeredCauldronBlock.changeLevel(state, level, pos, iblockdata1, null, CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL); // CraftBukkit
+            level.levelEvent(1046, pos, 0);
         }
 
     }
