--- a/net/minecraft/world/level/block/ChangeOverTimeBlock.java
+++ b/net/minecraft/world/level/block/ChangeOverTimeBlock.java
@@ -19,8 +19,8 @@
         float f = 0.05688889F;
 
         if (randomsource.nextFloat() < 0.05688889F) {
-            this.getNextState(blockstate, serverlevel, blockpos, randomsource).ifPresent((blockstate1) -> {
-                serverlevel.setBlockAndUpdate(blockpos, blockstate1);
+            this.getNextState(iblockdata, worldserver, blockposition, randomsource).ifPresent((iblockdata1) -> {
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(worldserver, blockposition, iblockdata1); // CraftBukkit
             });
         }
 
