--- a/net/minecraft/world/level/block/ComparatorBlock.java
+++ b/net/minecraft/world/level/block/ComparatorBlock.java
@@ -27,6 +27,7 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.ticks.TickPriority;
+import org.bukkit.craftbukkit.event.CraftEventFactory; // CraftBukkit
 
 public class ComparatorBlock extends DiodeBlock implements EntityBlock {
 
@@ -173,9 +167,19 @@
             boolean flag1 = (Boolean) blockstate.getValue(ComparatorBlock.POWERED);
 
             if (flag1 && !flag) {
-                level.setBlock(blockpos, (BlockState) blockstate.setValue(ComparatorBlock.POWERED, false), 2);
+                // CraftBukkit start
+                if (CraftEventFactory.callRedstoneChange(level, pos, 15, 0).getNewCurrent() != 0) {
+                    return;
+                }
+                // CraftBukkit end
+                level.setBlock(pos, (IBlockData) state.setValue(ComparatorBlock.POWERED, false), 2);
             } else if (!flag1 && flag) {
-                level.setBlock(blockpos, (BlockState) blockstate.setValue(ComparatorBlock.POWERED, true), 2);
+                // CraftBukkit start
+                if (CraftEventFactory.callRedstoneChange(level, pos, 0, 15).getNewCurrent() != 15) {
+                    return;
+                }
+                // CraftBukkit end
+                level.setBlock(pos, (IBlockData) state.setValue(ComparatorBlock.POWERED, true), 2);
             }
 
             this.updateNeighborsInFront(level, blockpos, blockstate);
