--- a/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
+++ b/net/minecraft/world/level/block/GrowingPlantHeadBlock.java
@@ -51,8 +47,8 @@
         if ((Integer) blockstate.getValue(GrowingPlantHeadBlock.AGE) < 25 && randomsource.nextDouble() < this.growPerTickProbability) {
             BlockPos blockpos1 = blockpos.relative(this.growthDirection);
 
-            if (this.canGrowInto(serverlevel.getBlockState(blockpos1))) {
-                serverlevel.setBlockAndUpdate(blockpos1, this.getGrowIntoState(blockstate, serverlevel.random));
+            if (this.canGrowInto(level.getBlockState(blockposition1))) {
+                org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, blockposition1, this.getGrowIntoState(state, level.random)); // CraftBukkit
             }
         }
 
