--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -188,14 +174,20 @@
                 if (level.getFluidState(blockpos1).is(FluidTags.WATER)) {
                     Block block = level.getFluidState(blockpos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
 
-                    level.setBlockAndUpdate(blockpos, block.defaultBlockState());
-                    this.fizz(level, blockpos);
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, block.defaultBlockState())) {
+                        this.fizz(level, pos);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
 
-                if (flag && level.getBlockState(blockpos1).is(Blocks.BLUE_ICE)) {
-                    level.setBlockAndUpdate(blockpos, Blocks.BASALT.defaultBlockState());
-                    this.fizz(level, blockpos);
+                if (flag && level.getBlockState(blockposition1).is(Blocks.BLUE_ICE)) {
+                    // CraftBukkit start
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, Blocks.BASALT.defaultBlockState())) {
+                        this.fizz(level, pos);
+                    }
+                    // CraftBukkit end
                     return false;
                 }
             }
