--- a/net/minecraft/world/level/block/MultifaceSpreader.java
+++ b/net/minecraft/world/level/block/MultifaceSpreader.java
@@ -158,7 +156,7 @@
                     levelaccessor.getChunk(multifacespreader_spreadpos.pos()).markPosForPostprocessing(multifacespreader_spreadpos.pos());
                 }
 
-                return levelaccessor.setBlock(multifacespreader_spreadpos.pos(), blockstate1, 2);
+                return org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos.source(), pos.pos(), iblockdata1, 2); // CraftBukkit
             } else {
                 return false;
             }
@@ -175,23 +173,20 @@
 
         SAME_POSITION {
             @Override
-            @Override
-            public MultifaceSpreader.SpreadPos getSpreadPos(BlockPos blockpos, Direction direction, Direction direction1) {
-                return new MultifaceSpreader.SpreadPos(blockpos, direction);
+            public MultifaceSpreader.c getSpreadPos(BlockPos pos, Direction face, Direction spreadDirection) {
+                return new MultifaceSpreader.c(pos, face, pos); // CraftBukkit
             }
         },
         SAME_PLANE {
             @Override
-            @Override
-            public MultifaceSpreader.SpreadPos getSpreadPos(BlockPos blockpos, Direction direction, Direction direction1) {
-                return new MultifaceSpreader.SpreadPos(blockpos.relative(direction), direction1);
+            public MultifaceSpreader.c getSpreadPos(BlockPos pos, Direction face, Direction spreadDirection) {
+                return new MultifaceSpreader.c(pos.relative(face), spreadDirection, pos); // CraftBukkit
             }
         },
         WRAP_AROUND {
             @Override
-            @Override
-            public MultifaceSpreader.SpreadPos getSpreadPos(BlockPos blockpos, Direction direction, Direction direction1) {
-                return new MultifaceSpreader.SpreadPos(blockpos.relative(direction).relative(direction1), direction.getOpposite());
+            public MultifaceSpreader.c getSpreadPos(BlockPos pos, Direction face, Direction spreadDirection) {
+                return new MultifaceSpreader.c(pos.relative(face).relative(spreadDirection), face.getOpposite(), pos); // CraftBukkit
             }
         };
 
@@ -200,7 +195,7 @@
         public abstract MultifaceSpreader.SpreadPos getSpreadPos(BlockPos pos, Direction face, Direction spreadDirection);
     }
 
-    public static record SpreadPos(BlockPos pos, Direction face) {
+    public static record c(BlockPos pos, Direction face, BlockPos source) { // CraftBukkit
 
     }
 }
