--- a/net/minecraft/world/level/block/NetherWartBlock.java
+++ b/net/minecraft/world/level/block/NetherWartBlock.java
@@ -57,9 +52,9 @@
     public void randomTick(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, RandomSource randomsource) {
         int i = (Integer) blockstate.getValue(NetherWartBlock.AGE);
 
-        if (i < 3 && randomsource.nextInt(10) == 0) {
-            blockstate = (BlockState) blockstate.setValue(NetherWartBlock.AGE, i + 1);
-            serverlevel.setBlock(blockpos, blockstate, 2);
+        if (i < 3 && random.nextInt(10) == 0) {
+            state = (IBlockData) state.setValue(NetherWartBlock.AGE, i + 1);
+            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, state, 2); // CraftBukkit
         }
 
     }
