--- a/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -81,8 +77,13 @@
 
         entity.setIsInPowderSnow(true);
         if (!level.isClientSide) {
-            if (entity.isOnFire() && (level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) || entity instanceof Player) && entity.mayInteract(level, blockpos)) {
-                level.destroyBlock(blockpos, false);
+            // CraftBukkit start
+            if (entity.isOnFire() && entity.mayInteract(level, pos)) {
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.AIR.defaultBlockState(), !(level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) || entity instanceof Player))) {
+                    return;
+                }
+                // CraftBukkit end
+                level.destroyBlock(pos, false);
             }
 
             entity.setSharedFlagOnFire(false);
