--- a/net/minecraft/world/level/block/SculkCatalystBlock.java
+++ b/net/minecraft/world/level/block/SculkCatalystBlock.java
@@ -66,9 +66,9 @@
     }
 
     @Override
-    @Override
-    public RenderShape getRenderShape(BlockState blockstate) {
-        return RenderShape.MODEL;
+    public void spawnAfterBreak(IBlockData state, ServerLevel level, BlockPos pos, ItemStack stack, boolean dropExperience) {
+        super.spawnAfterBreak(state, level, pos, stack, dropExperience);
+        // CraftBukkit start - Delegate to getExpDrop
     }
 
     @Override
@@ -79,5 +77,7 @@
             this.tryDropExperience(serverlevel, blockpos, itemstack, this.xpRange);
         }
 
+        return 0;
+        // CraftBukkit end
     }
 }
