--- a/net/minecraft/world/level/block/WitherRoseBlock.java
+++ b/net/minecraft/world/level/block/WitherRoseBlock.java
@@ -69,8 +65,8 @@
             if (entity instanceof LivingEntity) {
                 LivingEntity livingentity = (LivingEntity) entity;
 
-                if (!livingentity.isInvulnerableTo(level.damageSources().wither())) {
-                    livingentity.addEffect(new MobEffectInstance(MobEffects.WITHER, 40));
+                if (!entityliving.isInvulnerableTo(level.damageSources().wither())) {
+                    entityliving.addEffect(new MobEffectInstance(MobEffects.WITHER, 40), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.WITHER_ROSE); // CraftBukkit
                 }
             }
 
