--- a/net/minecraft/world/item/trading/Merchant.java
+++ b/net/minecraft/world/item/trading/Merchant.java
@@ -36,20 +36,22 @@
         return false;
     }
 
-    default void openTradingScreen(Player player, Component component, int i) {
-        OptionalInt optionalint = player.openMenu(new SimpleMenuProvider((j, inventory, player1) -> {
-            return new MerchantMenu(j, inventory, this);
-        }, component));
+    default void openTradingScreen(Player player, Component displayName, int level) {
+        OptionalInt optionalint = player.openMenu(new SimpleMenuProvider((j, playerinventory, entityhuman1) -> {
+            return new MerchantMenu(j, playerinventory, this);
+        }, displayName));
 
         if (optionalint.isPresent()) {
-            MerchantOffers merchantoffers = this.getOffers();
+            MerchantOffers merchantrecipelist = this.getOffers();
 
-            if (!merchantoffers.isEmpty()) {
-                player.sendMerchantOffers(optionalint.getAsInt(), merchantoffers, i, this.getVillagerXp(), this.showProgressBar(), this.canRestock());
+            if (!merchantrecipelist.isEmpty()) {
+                player.sendMerchantOffers(optionalint.getAsInt(), merchantrecipelist, level, this.getVillagerXp(), this.showProgressBar(), this.canRestock());
             }
         }
 
     }
 
     boolean isClientSide();
+
+    org.bukkit.craftbukkit.inventory.CraftMerchant getCraftMerchant(); // CraftBukkit
 }
