--- a/net/minecraft/network/protocol/game/ClientboundSetBorderCenterPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSetBorderCenterPacket.java
@@ -9,9 +9,11 @@
     private final double newCenterX;
     private final double newCenterZ;
 
-    public ClientboundSetBorderCenterPacket(WorldBorder worldborder) {
-        this.newCenterX = worldborder.getCenterX();
-        this.newCenterZ = worldborder.getCenterZ();
+    public ClientboundSetBorderCenterPacket(WorldBorder worldBorder) {
+        // CraftBukkit start - multiply out nether border
+        this.newCenterX = worldBorder.getCenterX() * (worldBorder.world != null ? worldBorder.world.dimensionType().coordinateScale() : 1.0);
+        this.newCenterZ = worldBorder.getCenterZ() * (worldBorder.world != null ? worldBorder.world.dimensionType().coordinateScale() : 1.0);
+        // CraftBukkit end
     }
 
     public ClientboundSetBorderCenterPacket(FriendlyByteBuf friendlybytebuf) {
