--- a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
@@ -26,15 +26,17 @@
         this.warningTime = friendlybytebuf.readVarInt();
     }
 
-    public ClientboundInitializeBorderPacket(WorldBorder worldborder) {
-        this.newCenterX = worldborder.getCenterX();
-        this.newCenterZ = worldborder.getCenterZ();
-        this.oldSize = worldborder.getSize();
-        this.newSize = worldborder.getLerpTarget();
-        this.lerpTime = worldborder.getLerpRemainingTime();
-        this.newAbsoluteMaxSize = worldborder.getAbsoluteMaxSize();
-        this.warningBlocks = worldborder.getWarningBlocks();
-        this.warningTime = worldborder.getWarningTime();
+    public ClientboundInitializeBorderPacket(WorldBorder worldBorder) {
+        // CraftBukkit start - multiply out nether border
+        this.newCenterX = worldBorder.getCenterX() * worldBorder.world.dimensionType().coordinateScale();
+        this.newCenterZ = worldBorder.getCenterZ() * worldBorder.world.dimensionType().coordinateScale();
+        // CraftBukkit end
+        this.oldSize = worldBorder.getSize();
+        this.newSize = worldBorder.getLerpTarget();
+        this.lerpTime = worldBorder.getLerpRemainingTime();
+        this.newAbsoluteMaxSize = worldBorder.getAbsoluteMaxSize();
+        this.warningBlocks = worldBorder.getWarningBlocks();
+        this.warningTime = worldBorder.getWarningTime();
     }
 
     @Override
