--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -13,22 +13,21 @@
 public class SpectralArrow extends AbstractArrow {
 
     private static final ItemStack DEFAULT_ARROW_STACK = new ItemStack(Items.SPECTRAL_ARROW);
-    private int duration = 200;
+    public int duration = 200;
 
-    public SpectralArrow(EntityType<? extends SpectralArrow> entitytype, Level level) {
-        super(entitytype, level, SpectralArrow.DEFAULT_ARROW_STACK);
+    public SpectralArrow(EntityType<? extends SpectralArrow> entityType, Level level) {
+        super(entityType, level, SpectralArrow.DEFAULT_ARROW_STACK);
     }
 
-    public SpectralArrow(Level level, LivingEntity livingentity, ItemStack itemstack) {
-        super(EntityType.SPECTRAL_ARROW, livingentity, level, itemstack);
+    public SpectralArrow(Level world, LivingEntity entityliving, ItemStack itemstack) {
+        super(EntityType.SPECTRAL_ARROW, entityliving, world, itemstack);
     }
 
-    public SpectralArrow(Level level, double d0, double d1, double d2, ItemStack itemstack) {
-        super(EntityType.SPECTRAL_ARROW, d0, d1, d2, level, itemstack);
+    public SpectralArrow(Level world, double d0, double d1, double d2, ItemStack itemstack) {
+        super(EntityType.SPECTRAL_ARROW, d0, d1, d2, world, itemstack);
     }
 
     @Override
-    @Override
     public void tick() {
         super.tick();
         if (this.level().isClientSide && !this.inGround) {
@@ -38,28 +37,25 @@
     }
 
     @Override
-    @Override
-    protected void doPostHurtEffects(LivingEntity livingentity) {
-        super.doPostHurtEffects(livingentity);
-        MobEffectInstance mobeffectinstance = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
+    protected void doPostHurtEffects(LivingEntity living) {
+        super.doPostHurtEffects(living);
+        MobEffectInstance mobeffect = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
 
-        livingentity.addEffect(mobeffectinstance, this.getEffectSource());
+        living.addEffect(mobeffect, this.getEffectSource(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
     }
 
     @Override
-    @Override
-    public void readAdditionalSaveData(CompoundTag compoundtag) {
-        super.readAdditionalSaveData(compoundtag);
-        if (compoundtag.contains("Duration")) {
-            this.duration = compoundtag.getInt("Duration");
+    public void readAdditionalSaveData(CompoundTag compound) {
+        super.readAdditionalSaveData(compound);
+        if (compound.contains("Duration")) {
+            this.duration = compound.getInt("Duration");
         }
 
     }
 
     @Override
-    @Override
-    public void addAdditionalSaveData(CompoundTag compoundtag) {
-        super.addAdditionalSaveData(compoundtag);
-        compoundtag.putInt("Duration", this.duration);
+    public void addAdditionalSaveData(CompoundTag compound) {
+        super.addAdditionalSaveData(compound);
+        compound.putInt("Duration", this.duration);
     }
 }
