--- a/net/minecraft/world/level/BaseCommandBlock.java
+++ b/net/minecraft/world/level/BaseCommandBlock.java
@@ -30,6 +30,10 @@
     private Component lastOutput;
     private String command = "";
     private Component name;
+    // CraftBukkit start
+    @Override
+    public abstract org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper);
+    // CraftBukkit end
 
     public BaseCommandBlock() {
         this.name = BaseCommandBlock.DEFAULT_NAME;
@@ -126,7 +130,7 @@
 
                         });
 
-                        minecraftserver.getCommands().performPrefixedCommand(commandsourcestack, this.command);
+                        minecraftserver.getCommands().dispatchServerCommand(commandlistenerwrapper, this.command); // CraftBukkit
                     } catch (Throwable throwable) {
                         CrashReport crashreport = CrashReport.forThrowable(throwable, "Executing command block");
                         CrashReportCategory crashreportcategory = crashreport.addCategory("Command to be executed");
