--- a/net/minecraft/world/level/LevelAccessor.java
+++ b/net/minecraft/world/level/LevelAccessor.java
@@ -116,4 +114,6 @@
     default void gameEvent(GameEvent gameevent, BlockPos blockpos, GameEvent.Context gameevent_context) {
         this.gameEvent(gameevent, Vec3.atCenterOf(blockpos), gameevent_context);
     }
+
+    net.minecraft.server.level.ServerLevel getMinecraftWorld(); // CraftBukkit
 }
