--- a/net/minecraft/world/level/storage/loot/LootDataManager.java
+++ b/net/minecraft/world/level/storage/loot/LootDataManager.java
@@ -101,6 +101,13 @@
         problemreporter_collector.get().forEach((s, s1) -> {
             LootDataManager.LOGGER.warn("Found loot table element validation problem in {}: {}", s, s1);
         });
+        // CraftBukkit start
+        map1.forEach((key, lootTable) -> {
+            if (object instanceof LootTable table) {
+                table.craftLootTable = new CraftLootTable(CraftNamespacedKey.fromMinecraft(key.location()), table);
+            }
+        });
+        // CraftBukkit end
         this.elements = map1;
         this.typeKeys = com_google_common_collect_immutablemultimap_builder.build();
     }
