--- a/net/minecraft/commands/arguments/selector/EntitySelector.java
+++ b/net/minecraft/commands/arguments/selector/EntitySelector.java
@@ -93,8 +91,8 @@
         return this.usesSelector;
     }
 
-    private void checkPermissions(CommandSourceStack commandsourcestack) throws CommandSyntaxException {
-        if (this.usesSelector && !commandsourcestack.hasPermission(2)) {
+    private void checkPermissions(CommandSourceStack source) throws CommandSyntaxException {
+        if (this.usesSelector && !source.hasPermission(2, "minecraft.command.selector")) { // CraftBukkit
             throw EntityArgument.ERROR_SELECTORS_NOT_ALLOWED.create();
         }
     }
