--- a/net/minecraft/world/item/BowItem.java
+++ b/net/minecraft/world/item/BowItem.java
@@ -65,6 +64,14 @@
                         if (EnchantmentHelper.getItemEnchantmentLevel(Enchantments.FLAMING_ARROWS, itemstack) > 0) {
                             abstractarrow.setSecondsOnFire(100);
                         }
+                        // CraftBukkit start
+                        org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(entityhuman, stack, itemstack1, entityarrow, entityhuman.getUsedItemHand(), f, !flag1);
+                        if (event.isCancelled()) {
+                            event.getProjectile().remove();
+                            return;
+                        }
+                        flag1 = !event.shouldConsumeItem();
+                        // CraftBukkit end
 
                         itemstack.hurtAndBreak(1, player, (player1) -> {
                             player1.broadcastBreakEvent(player.getUsedItemHand());
@@ -73,7 +80,16 @@
                             abstractarrow.pickup = AbstractArrow.Pickup.CREATIVE_ONLY;
                         }
 
-                        level.addFreshEntity(abstractarrow);
+                        // CraftBukkit start
+                        if (event.getProjectile() == entityarrow.getBukkitEntity()) {
+                            if (!level.addFreshEntity(entityarrow)) {
+                                if (entityhuman instanceof net.minecraft.server.level.ServerPlayer) {
+                                    ((net.minecraft.server.level.ServerPlayer) entityhuman).getBukkitEntity().updateInventory();
+                                }
+                                return;
+                            }
+                        }
+                        // CraftBukkit end
                     }
 
                     level.playSound((Player) null, player.getX(), player.getY(), player.getZ(), SoundEvents.ARROW_SHOOT, SoundSource.PLAYERS, 1.0F, 1.0F / (level.getRandom().nextFloat() * 0.4F + 1.2F) + f * 0.5F);
