--- a/net/minecraft/world/item/CrossbowItem.java
+++ b/net/minecraft/world/item/CrossbowItem.java
@@ -240,12 +236,28 @@
 
                 ((Projectile) object).shoot((double) vector3f.x(), (double) vector3f.y(), (double) vector3f.z(), f1, f2);
             }
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(shooter, crossbowStack, ammoStack, (Entity) object, shooter.getUsedItemHand(), soundPitch, true);
+            if (event.isCancelled()) {
+                event.getProjectile().remove();
+                return;
+            }
+            // CraftBukkit end
 
             itemstack.hurtAndBreak(flag1 ? 3 : 1, livingentity, (livingentity1) -> {
                 livingentity1.broadcastBreakEvent(interactionhand);
             });
-            level.addFreshEntity((Entity) object);
-            level.playSound((Player) null, livingentity.getX(), livingentity.getY(), livingentity.getZ(), SoundEvents.CROSSBOW_SHOOT, SoundSource.PLAYERS, 1.0F, f);
+            // CraftBukkit start
+            if (event.getProjectile() == ((Entity) object).getBukkitEntity()) {
+                if (!level.addFreshEntity((Entity) object)) {
+                    if (shooter instanceof ServerPlayer) {
+                        ((ServerPlayer) shooter).getBukkitEntity().updateInventory();
+                    }
+                    return;
+                }
+            }
+            // CraftBukkit end
+            level.playSound((Player) null, shooter.getX(), shooter.getY(), shooter.getZ(), SoundEvents.CROSSBOW_SHOOT, SoundSource.PLAYERS, 1.0F, soundPitch);
         }
     }
 
