--- a/net/minecraft/world/item/DyeItem.java
+++ b/net/minecraft/world/item/DyeItem.java
@@ -12,6 +12,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
+import org.bukkit.event.entity.SheepDyeWoolEvent; // CraftBukkit
 
 public class DyeItem extends Item implements SignApplicator {
 
@@ -33,8 +33,18 @@
             if (sheep.isAlive() && !sheep.isSheared() && sheep.getColor() != this.dyeColor) {
                 sheep.level().playSound(player, (Entity) sheep, SoundEvents.DYE_USE, SoundSource.PLAYERS, 1.0F, 1.0F);
                 if (!player.level().isClientSide) {
-                    sheep.setColor(this.dyeColor);
-                    itemstack.shrink(1);
+                    // CraftBukkit start
+                    byte bColor = (byte) this.dyeColor.getId();
+                    SheepDyeWoolEvent event = new SheepDyeWoolEvent((org.bukkit.entity.Sheep) entitysheep.getBukkitEntity(), org.bukkit.DyeColor.getByWoolData(bColor), (org.bukkit.entity.Player) player.getBukkitEntity());
+                    entitysheep.level().getCraftServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return InteractionResult.PASS;
+                    }
+
+                    entitysheep.setColor(DyeColor.byId((byte) event.getColor().getWoolData()));
+                    // CraftBukkit end
+                    stack.shrink(1);
                 }
 
                 return InteractionResult.sidedSuccess(player.level().isClientSide);
