--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -46,10 +45,15 @@
                     if (level instanceof ServerLevel) {
                         EndCrystal endcrystal = new EndCrystal(level, d0 + 0.5D, d1, d2 + 0.5D);
 
-                        endcrystal.setShowBottom(false);
-                        level.addFreshEntity(endcrystal);
-                        level.gameEvent((Entity) useoncontext.getPlayer(), GameEvent.ENTITY_PLACE, blockpos1);
-                        EndDragonFight enddragonfight = ((ServerLevel) level).getDragonFight();
+                        entityendercrystal.setShowBottom(false);
+                        // CraftBukkit start
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPlaceEvent(context, entityendercrystal).isCancelled()) {
+                            return InteractionResult.FAIL;
+                        }
+                        // CraftBukkit end
+                        world.addFreshEntity(entityendercrystal);
+                        world.gameEvent((Entity) context.getPlayer(), GameEvent.ENTITY_PLACE, blockposition1);
+                        EndDragonFight enderdragonbattle = ((ServerLevel) world).getDragonFight();
 
                         if (enddragonfight != null) {
                             enddragonfight.tryRespawn();
