--- a/net/minecraft/world/item/EnderpearlItem.java
+++ b/net/minecraft/world/item/EnderpearlItem.java
@@ -20,8 +19,7 @@
     public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand interactionhand) {
         ItemStack itemstack = player.getItemInHand(interactionhand);
 
-        level.playSound((Player) null, player.getX(), player.getY(), player.getZ(), SoundEvents.ENDER_PEARL_THROW, SoundSource.NEUTRAL, 0.5F, 0.4F / (level.getRandom().nextFloat() * 0.4F + 0.8F));
-        player.getCooldowns().addCooldown(this, 20);
+        // CraftBukkit start - change order
         if (!level.isClientSide) {
             ThrownEnderpearl thrownenderpearl = new ThrownEnderpearl(level, player);
 
@@ -30,6 +33,10 @@
             level.addFreshEntity(thrownenderpearl);
         }
 
+        level.playSound((Player) null, player.getX(), player.getY(), player.getZ(), SoundEvents.ENDER_PEARL_THROW, SoundSource.NEUTRAL, 0.5F, 0.4F / (level.getRandom().nextFloat() * 0.4F + 0.8F));
+        player.getCooldowns().addCooldown(this, 20);
+        // CraftBukkit end
+
         player.awardStat(Stats.ITEM_USED.get(this));
         if (!player.getAbilities().instabuild) {
             itemstack.shrink(1);
