--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -69,7 +69,7 @@
     public static Integer getMapId(ItemStack itemstack) {
         CompoundTag compoundtag = itemstack.getTag();
 
-        return compoundtag != null && compoundtag.contains("map", 99) ? compoundtag.getInt("map") : null;
+        return nbttagcompound != null && nbttagcompound.contains("map", 99) ? nbttagcompound.getInt("map") : -1; // CraftBukkit - make new maps for no tag
     }
 
     private static int createNewSavedData(Level level, int i, int j, int k, boolean flag, boolean flag1, ResourceKey<Level> resourcekey) {
