--- a/net/minecraft/world/item/MilkBucketItem.java
+++ b/net/minecraft/world/item/MilkBucketItem.java
@@ -32,7 +31,7 @@
         }
 
         if (!level.isClientSide) {
-            livingentity.removeAllEffects();
+            entityLiving.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.MILK); // CraftBukkit
         }
 
         return itemstack.isEmpty() ? new ItemStack(Items.BUCKET) : itemstack;
