--- a/net/minecraft/world/item/RecordItem.java
+++ b/net/minecraft/world/item/RecordItem.java
@@ -46,9 +45,10 @@
         if (blockstate.is(Blocks.JUKEBOX) && !(Boolean) blockstate.getValue(JukeboxBlock.HAS_RECORD)) {
             ItemStack itemstack = useoncontext.getItemInHand();
 
-            if (!level.isClientSide) {
-                Player player = useoncontext.getPlayer();
-                BlockEntity blockentity = level.getBlockEntity(blockpos);
+            if (!world.isClientSide) {
+                if (true) return InteractionResult.SUCCESS; // CraftBukkit - handled in ItemStack
+                Player entityhuman = context.getPlayer();
+                BlockEntity tileentity = world.getBlockEntity(blockposition);
 
                 if (blockentity instanceof JukeboxBlockEntity) {
                     JukeboxBlockEntity jukeboxblockentity = (JukeboxBlockEntity) blockentity;
