--- a/net/minecraft/world/item/SignItem.java
+++ b/net/minecraft/world/item/SignItem.java
@@ -13,8 +13,10 @@
 
 public class SignItem extends StandingAndWallBlockItem {
 
-    public SignItem(Item.Properties item_properties, Block block, Block block1) {
-        super(block, block1, item_properties, Direction.DOWN);
+    public static BlockPos openSign; // CraftBukkit
+
+    public SignItem(Item.Properties properties, Block standingBlock, Block wallBlock) {
+        super(standingBlock, wallBlock, properties, Direction.DOWN);
     }
 
     public SignItem(Item.Properties item_properties, Block block, Block block1, Direction direction) {
@@ -36,7 +37,10 @@
                 if (block instanceof SignBlock) {
                     SignBlock signblock = (SignBlock) block;
 
-                    signblock.openTextEdit(player, signblockentity, true);
+                    // CraftBukkit start - SPIGOT-4678
+                    // blocksign.openTextEdit(entityhuman, tileentitysign, true);
+                    SignItem.openSign = pos;
+                    // CraftBukkit end
                 }
             }
         }
