--- a/net/minecraft/world/item/TridentItem.java
+++ b/net/minecraft/world/item/TridentItem.java
@@ -72,8 +68,10 @@
 
                 if (k <= 0 || player.isInWaterOrRain()) {
                     if (!level.isClientSide) {
-                        itemstack.hurtAndBreak(1, player, (player1) -> {
-                            player1.broadcastBreakEvent(livingentity.getUsedItemHand());
+                        // CraftBukkit - moved down
+                        /*
+                        itemstack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
+                            entityhuman1.broadcastBreakEvent(entityliving.getUsedItemHand());
                         });
                         if (k == 0) {
                             ThrownTrident throwntrident = new ThrownTrident(level, player, itemstack);
@@ -83,18 +82,41 @@
                                 throwntrident.pickup = AbstractArrow.Pickup.CREATIVE_ONLY;
                             }
 
-                            level.addFreshEntity(throwntrident);
-                            level.playSound((Player) null, (Entity) throwntrident, SoundEvents.TRIDENT_THROW, SoundSource.PLAYERS, 1.0F, 1.0F);
-                            if (!player.getAbilities().instabuild) {
-                                player.getInventory().removeItem(itemstack);
+                            // CraftBukkit start
+                            if (!level.addFreshEntity(entitythrowntrident)) {
+                                if (entityhuman instanceof net.minecraft.server.level.ServerPlayer) {
+                                    ((net.minecraft.server.level.ServerPlayer) entityhuman).getBukkitEntity().updateInventory();
+                                }
+                                return;
                             }
+
+                            stack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
+                                entityhuman1.broadcastBreakEvent(entityLiving.getUsedItemHand());
+                            });
+                            entitythrowntrident.pickupItemStack = stack.copy(); // SPIGOT-4511 update since damage call moved
+                            // CraftBukkit end
+
+                            level.playSound((Player) null, (Entity) entitythrowntrident, SoundEvents.TRIDENT_THROW, SoundSource.PLAYERS, 1.0F, 1.0F);
+                            if (!entityhuman.getAbilities().instabuild) {
+                                entityhuman.getInventory().removeItem(stack);
+                            }
+                            // CraftBukkit start - SPIGOT-5458 also need in this branch :(
+                        } else {
+                            stack.hurtAndBreak(1, entityhuman, (entityhuman1) -> {
+                                entityhuman1.broadcastBreakEvent(entityLiving.getUsedItemHand());
+                            });
+                            // CraftBukkkit end
                         }
                     }
 
                     player.awardStat(Stats.ITEM_USED.get(this));
                     if (k > 0) {
-                        float f = player.getYRot();
-                        float f1 = player.getXRot();
+                        // CraftBukkit start
+                        org.bukkit.event.player.PlayerRiptideEvent event = new org.bukkit.event.player.PlayerRiptideEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(stack));
+                        event.getPlayer().getServer().getPluginManager().callEvent(event);
+                        // CraftBukkit end
+                        float f = entityhuman.getYRot();
+                        float f1 = entityhuman.getXRot();
                         float f2 = -Mth.sin(f * 0.017453292F) * Mth.cos(f1 * 0.017453292F);
                         float f3 = -Mth.sin(f1 * 0.017453292F);
                         float f4 = Mth.cos(f * 0.017453292F) * Mth.cos(f1 * 0.017453292F);
