--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -170,7 +165,7 @@
                     MobEffectInstance mobeffectinstance1 = new MobEffectInstance(MobEffects.BAD_OMEN, 120000, i, false, false, true);
 
                     if (!this.level().getGameRules().getBoolean(GameRules.RULE_DISABLE_RAIDS)) {
-                        player.addEffect(mobeffectinstance1);
+                        entityhuman.addEffect(mobeffect1, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.PATROL_CAPTAIN); // CraftBukkit
                     }
                 }
             }
@@ -547,7 +521,7 @@
             while (iterator.hasNext()) {
                 Raider raider = (Raider) iterator.next();
 
-                raider.setTarget(this.mob.getTarget());
+                entityraider.setTarget(this.mob.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
             }
 
         }
@@ -565,8 +538,8 @@
                 while (iterator.hasNext()) {
                     Raider raider = (Raider) iterator.next();
 
-                    raider.setTarget(livingentity);
-                    raider.setAggressive(true);
+                    entityraider.setTarget(this.mob.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER, true); // CraftBukkit
+                    entityraider.setAggressive(true);
                 }
 
                 this.mob.setAggressive(true);
