--- a/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
+++ b/net/minecraft/world/level/levelgen/structure/LegacyStructureDataHandler.java
@@ -236,16 +236,16 @@
         }
     }
 
-    public static LegacyStructureDataHandler getLegacyStructureHandler(ResourceKey<Level> resourcekey, @Nullable DimensionDataStorage dimensiondatastorage) {
-        if (resourcekey == Level.OVERWORLD) {
-            return new LegacyStructureDataHandler(dimensiondatastorage, ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"), ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument"));
+    public static LegacyStructureDataHandler getLegacyStructureHandler(ResourceKey<LevelStem> level, @Nullable DimensionDataStorage storage) { // CraftBukkit
+        if (level == LevelStem.OVERWORLD) { // CraftBukkit
+            return new LegacyStructureDataHandler(storage, ImmutableList.of("Monument", "Stronghold", "Village", "Mineshaft", "Temple", "Mansion"), ImmutableList.of("Village", "Mineshaft", "Mansion", "Igloo", "Desert_Pyramid", "Jungle_Pyramid", "Swamp_Hut", "Stronghold", "Monument"));
         } else {
             ImmutableList immutablelist;
 
-            if (resourcekey == Level.NETHER) {
+            if (level == LevelStem.NETHER) { // CraftBukkit
                 immutablelist = ImmutableList.of("Fortress");
-                return new LegacyStructureDataHandler(dimensiondatastorage, immutablelist, immutablelist);
-            } else if (resourcekey == Level.END) {
+                return new LegacyStructureDataHandler(storage, immutablelist, immutablelist);
+            } else if (level == LevelStem.END) { // CraftBukkit
                 immutablelist = ImmutableList.of("EndCity");
                 return new LegacyStructureDataHandler(dimensiondatastorage, immutablelist, immutablelist);
             } else {
