--- a/net/minecraft/world/level/block/BasePressurePlateBlock.java
+++ b/net/minecraft/world/level/block/BasePressurePlateBlock.java
@@ -9,22 +9,25 @@
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySelector;
+import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.state.BlockBehaviour;
-import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.IBlockData;
 import net.minecraft.world.level.block.state.properties.BlockSetType;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
+import org.bukkit.event.block.BlockRedstoneEvent; // CraftBukkit
 
 public abstract class BasePressurePlateBlock extends Block {
-    protected static final VoxelShape PRESSED_AABB = Block.box(1.0, 0.0, 1.0, 15.0, 0.5, 15.0);
-    protected static final VoxelShape AABB = Block.box(1.0, 0.0, 1.0, 15.0, 1.0, 15.0);
-    protected static final AABB TOUCH_AABB = new AABB(0.0625, 0.0, 0.0625, 0.9375, 0.25, 0.9375);
+
+    protected static final VoxelShape PRESSED_AABB = Block.box(1.0D, 0.0D, 1.0D, 15.0D, 0.5D, 15.0D);
+    protected static final VoxelShape AABB = Block.box(1.0D, 0.0D, 1.0D, 15.0D, 1.0D, 15.0D);
+    protected static final AABB TOUCH_AABB = new AABB(0.0625D, 0.0D, 0.0625D, 0.9375D, 0.25D, 0.9375D);
     protected final BlockSetType type;
 
     protected BasePressurePlateBlock(BlockBehaviour.Properties properties, BlockSetType type) {
@@ -36,8 +39,8 @@
     protected abstract MapCodec<? extends BasePressurePlateBlock> codec();
 
     @Override
-    public VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
-        return this.getSignalForState(state) > 0 ? PRESSED_AABB : AABB;
+    public VoxelShape getShape(IBlockData state, BlockGetter level, BlockPos pos, CollisionContext context) {
+        return this.getSignalForState(state) > 0 ? BasePressurePlateBlock.PRESSED_AABB : BasePressurePlateBlock.AABB;
     }
 
     protected int getPressedTime() {
@@ -45,67 +48,86 @@
     }
 
     @Override
-    public boolean isPossibleToRespawnInThis(BlockState state) {
+    public boolean isPossibleToRespawnInThis(IBlockData state) {
         return true;
     }
 
     @Override
-    public BlockState updateShape(BlockState state, Direction facing, BlockState facingState, LevelAccessor level, BlockPos currentPos, BlockPos facingPos) {
-        return facing == Direction.DOWN && !state.canSurvive(level, currentPos)
-            ? Blocks.AIR.defaultBlockState()
-            : super.updateShape(state, facing, facingState, level, currentPos, facingPos);
+    public IBlockData updateShape(IBlockData state, Direction facing, IBlockData facingState, LevelAccessor level, BlockPos currentPos, BlockPos facingPos) {
+        return facing == Direction.DOWN && !state.canSurvive(level, currentPos) ? Blocks.AIR.defaultBlockState() : super.updateShape(state, facing, facingState, level, currentPos, facingPos);
     }
 
     @Override
-    public boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
-        BlockPos blockPos = pos.below();
-        return canSupportRigidBlock(level, blockPos) || canSupportCenter(level, blockPos, Direction.UP);
+    public boolean canSurvive(IBlockData state, LevelReader level, BlockPos pos) {
+        BlockPos blockposition1 = pos.below();
+
+        return canSupportRigidBlock(level, blockposition1) || canSupportCenter(level, blockposition1, Direction.UP);
     }
 
     @Override
-    public void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        int signalForState = this.getSignalForState(state);
-        if (signalForState > 0) {
-            this.checkPressed(null, level, pos, state, signalForState);
+    public void tick(IBlockData state, ServerLevel level, BlockPos pos, RandomSource random) {
+        int i = this.getSignalForState(state);
+
+        if (i > 0) {
+            this.checkPressed((Entity) null, level, pos, state, i);
         }
+
     }
 
     @Override
-    public void entityInside(BlockState state, Level level, BlockPos pos, Entity entity) {
+    public void entityInside(IBlockData state, Level level, BlockPos pos, Entity entity) {
         if (!level.isClientSide) {
-            int signalForState = this.getSignalForState(state);
-            if (signalForState == 0) {
-                this.checkPressed(entity, level, pos, state, signalForState);
+            int i = this.getSignalForState(state);
+
+            if (i == 0) {
+                this.checkPressed(entity, level, pos, state, i);
             }
+
         }
     }
 
-    private void checkPressed(@Nullable Entity entity, Level level, BlockPos pos, BlockState state, int currentSignal) {
-        int signalStrength = this.getSignalStrength(level, pos);
+    private void checkPressed(@Nullable Entity entity, Level level, BlockPos pos, IBlockData state, int currentSignal) {
+        int j = this.getSignalStrength(level, pos);
         boolean flag = currentSignal > 0;
-        boolean flag1 = signalStrength > 0;
-        if (currentSignal != signalStrength) {
-            BlockState blockState = this.setSignalForState(state, signalStrength);
-            level.setBlock(pos, blockState, 2);
+        boolean flag1 = j > 0;
+
+        // CraftBukkit start - Interact Pressure Plate
+        org.bukkit.World bworld = level.getWorld();
+        org.bukkit.plugin.PluginManager manager = level.getCraftServer().getPluginManager();
+
+        if (flag != flag1) {
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(bworld.getBlockAt(pos.getX(), pos.getY(), pos.getZ()), currentSignal, j);
+            manager.callEvent(eventRedstone);
+
+            flag1 = eventRedstone.getNewCurrent() > 0;
+            j = eventRedstone.getNewCurrent();
+        }
+        // CraftBukkit end
+
+        if (currentSignal != j) {
+            IBlockData iblockdata1 = this.setSignalForState(state, j);
+
+            level.setBlock(pos, iblockdata1, 2);
             this.updateNeighbours(level, pos);
-            level.setBlocksDirty(pos, state, blockState);
+            level.setBlocksDirty(pos, state, iblockdata1);
         }
 
         if (!flag1 && flag) {
-            level.playSound(null, pos, this.type.pressurePlateClickOff(), SoundSource.BLOCKS);
+            level.playSound((Player) null, pos, this.type.pressurePlateClickOff(), SoundSource.BLOCKS);
             level.gameEvent(entity, GameEvent.BLOCK_DEACTIVATE, pos);
         } else if (flag1 && !flag) {
-            level.playSound(null, pos, this.type.pressurePlateClickOn(), SoundSource.BLOCKS);
+            level.playSound((Player) null, pos, this.type.pressurePlateClickOn(), SoundSource.BLOCKS);
             level.gameEvent(entity, GameEvent.BLOCK_ACTIVATE, pos);
         }
 
         if (flag1) {
-            level.scheduleTick(new BlockPos(pos), this, this.getPressedTime());
+            level.scheduleTick(new BlockPos(pos), (Block) this, this.getPressedTime());
         }
+
     }
 
     @Override
-    public void onRemove(BlockState state, Level level, BlockPos pos, BlockState newState, boolean isMoving) {
+    public void onRemove(IBlockData state, Level level, BlockPos pos, IBlockData newState, boolean isMoving) {
         if (!isMoving && !state.is(newState.getBlock())) {
             if (this.getSignalForState(state) > 0) {
                 this.updateNeighbours(level, pos);
@@ -121,27 +143,35 @@
     }
 
     @Override
-    public int getSignal(BlockState blockState, BlockGetter blockAccess, BlockPos pos, Direction side) {
+    public int getSignal(IBlockData blockState, BlockGetter blockAccess, BlockPos pos, Direction side) {
         return this.getSignalForState(blockState);
     }
 
     @Override
-    public int getDirectSignal(BlockState blockState, BlockGetter blockAccess, BlockPos pos, Direction side) {
+    public int getDirectSignal(IBlockData blockState, BlockGetter blockAccess, BlockPos pos, Direction side) {
         return side == Direction.UP ? this.getSignalForState(blockState) : 0;
     }
 
     @Override
-    public boolean isSignalSource(BlockState state) {
+    public boolean isSignalSource(IBlockData state) {
         return true;
     }
 
     protected static int getEntityCount(Level level, AABB box, Class<? extends Entity> entityClass) {
-        return level.getEntitiesOfClass(entityClass, box, EntitySelector.NO_SPECTATORS.and(entity -> !entity.isIgnoringBlockTriggers())).size();
+        // CraftBukkit start
+        return getEntities(level, box, entityClass).size();
     }
 
+    protected static <T extends Entity> java.util.List<T> getEntities(Level world, AABB axisalignedbb, Class<T> oclass) {
+        // CraftBukkit end
+        return world.getEntitiesOfClass(oclass, axisalignedbb, EntitySelector.NO_SPECTATORS.and((entity) -> {
+            return !entity.isIgnoringBlockTriggers();
+        })); // CraftBukkit
+    }
+
     protected abstract int getSignalStrength(Level level, BlockPos pos);
 
-    protected abstract int getSignalForState(BlockState state);
+    protected abstract int getSignalForState(IBlockData state);
 
-    protected abstract BlockState setSignalForState(BlockState state, int signal);
+    protected abstract IBlockData setSignalForState(IBlockData state, int signal);
 }
