--- a/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/LootItemRandomChanceWithLootingCondition.java
@@ -37,6 +34,11 @@
         if (entity instanceof LivingEntity) {
             i = EnchantmentHelper.getMobLooting((LivingEntity) entity);
         }
+        // CraftBukkit start - only use lootingModifier if set by Bukkit
+        if (context.hasParam(LootContextParams.LOOTING_MOD)) {
+            i = context.getParamOrNull(LootContextParams.LOOTING_MOD);
+        }
+        // CraftBukkit end
 
         return lootcontext.getRandom().nextFloat() < this.percent + (float) i * this.lootingMultiplier;
     }
