--- a/net/minecraft/world/inventory/LecternMenu.java
+++ b/net/minecraft/world/inventory/LecternMenu.java
@@ -4,8 +4,31 @@
 import net.minecraft.world.SimpleContainer;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.block.entity.LecternBlockEntity.LecternInventory;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.inventory.CraftInventoryLectern;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerTakeLecternBookEvent;
+// CraftBukkit end
 
 public class LecternMenu extends AbstractContainerMenu {
+
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private Player player;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventoryLectern inventory = new CraftInventoryLectern(this.lectern);
+        bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+        return bukkitEntity;
+    }
+    // CraftBukkit end
     private static final int DATA_COUNT = 1;
     private static final int SLOT_COUNT = 1;
     public static final int BUTTON_PREV_PAGE = 1;
@@ -15,24 +38,27 @@
     private final Container lectern;
     private final ContainerData lecternData;
 
-    public LecternMenu(int containerId) {
-        this(containerId, new SimpleContainer(1), new SimpleContainerData(1));
+    // CraftBukkit start - add player
+    public LecternMenu(int i, Inventory playerinventory) {
+        this(i, new SimpleContainer(1), new SimpleContainerData(1), playerinventory);
     }
 
-    public LecternMenu(int containerId, Container lectern, ContainerData lecternData) {
-        super(MenuType.LECTERN, containerId);
-        checkContainerSize(lectern, 1);
-        checkContainerDataCount(lecternData, 1);
-        this.lectern = lectern;
-        this.lecternData = lecternData;
-        this.addSlot(new Slot(lectern, 0, 0, 0) {
+    public LecternMenu(int i, Container iinventory, ContainerData icontainerproperties, Inventory playerinventory) {
+        // CraftBukkit end
+        super(MenuType.LECTERN, i);
+        checkContainerSize(iinventory, 1);
+        checkContainerDataCount(icontainerproperties, 1);
+        this.lectern = iinventory;
+        this.lecternData = icontainerproperties;
+        this.addSlot(new Slot(iinventory, 0, 0, 0) {
             @Override
             public void setChanged() {
                 super.setChanged();
                 LecternMenu.this.slotsChanged(this.container);
             }
         });
-        this.addDataSlots(lecternData);
+        this.addDataSlots(icontainerproperties);
+        player = (Player) playerinventory.player.getBukkitEntity(); // CraftBukkit
     }
 
     @Override
@@ -58,7 +84,15 @@
                         return false;
                     }
 
-                    ItemStack itemStack = this.lectern.removeItemNoUpdate(0);
+                    // CraftBukkit start - Event for taking the book
+                    PlayerTakeLecternBookEvent event = new PlayerTakeLecternBookEvent(player, ((CraftInventoryLectern) getBukkitView().getTopInventory()).getHolder());
+                    Bukkit.getServer().getPluginManager().callEvent(event);
+                    if (event.isCancelled()) {
+                        return false;
+                    }
+                    // CraftBukkit end
+                    ItemStack itemstack = this.lectern.removeItemNoUpdate(0);
+
                     this.lectern.setChanged();
                     if (!player.getInventory().add(itemStack)) {
                         player.drop(itemStack, false);
@@ -83,7 +117,9 @@
     }
 
     @Override
-    public boolean stillValid(Player player) {
+    public boolean stillValid(net.minecraft.world.entity.player.Player player) {
+        if (lectern instanceof LecternInventory && !((LecternInventory) lectern).getLectern().hasBook()) return false; // CraftBukkit
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.lectern.stillValid(player);
     }
 
