--- a/net/minecraft/commands/CommandSource.java
+++ b/net/minecraft/commands/CommandSource.java
@@ -6,26 +6,29 @@
 
     CommandSource NULL = new CommandSource() {
         @Override
-        @Override
         public void sendSystemMessage(Component component) {}
 
         @Override
-        @Override
         public boolean acceptsSuccess() {
             return false;
         }
 
         @Override
-        @Override
         public boolean acceptsFailure() {
             return false;
         }
 
         @Override
-        @Override
         public boolean shouldInformAdmins() {
             return false;
         }
+
+        // CraftBukkit start
+        @Override
+        public org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper) {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // CraftBukkit end
     };
 
     void sendSystemMessage(Component component);
@@ -39,4 +42,6 @@
     default boolean alwaysAccepts() {
         return false;
     }
+
+    org.bukkit.command.CommandSender getBukkitSender(CommandSourceStack wrapper); // CraftBukkit
 }
