--- a/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
@@ -27,11 +26,14 @@
 
     @Override
     public boolean test(LootContext context) {
-        Float _float = context.getParamOrNull(LootContextParams.EXPLOSION_RADIUS);
-        if (_float != null) {
-            RandomSource random = context.getRandom();
-            float f = 1.0F / _float;
-            return random.nextFloat() <= f;
+        Float ofloat = (Float) context.getParamOrNull(LootContextParams.EXPLOSION_RADIUS);
+
+        if (ofloat != null) {
+            RandomSource randomsource = context.getRandom();
+            float f = 1.0F / ofloat;
+
+            // CraftBukkit - <= to < to allow for plugins to completely disable block drops from explosions
+            return randomsource.nextFloat() < f;
         } else {
             return true;
         }
