--- a/net/minecraft/world/level/border/WorldBorder.java
+++ b/net/minecraft/world/level/border/WorldBorder.java
@@ -29,6 +29,7 @@
     int absoluteMaxSize = 29999984;
     private WorldBorder.BorderExtent extent = new WorldBorder.StaticBorderExtent(5.9999968E7D);
     public static final WorldBorder.Settings DEFAULT_SETTINGS = new WorldBorder.Settings(0.0D, 0.0D, 0.2D, 5.0D, 5, 15, 5.9999968E7D, 0L, 0.0D);
+    public net.minecraft.server.level.ServerLevel world; // CraftBukkit
 
     public WorldBorder() {}
 
@@ -163,8 +164,9 @@
         return Lists.newArrayList(this.listeners);
     }
 
-    public void addListener(BorderChangeListener borderchangelistener) {
-        this.listeners.add(borderchangelistener);
+    public void addListener(BorderChangeListener listener) {
+        if (listeners.contains(listener)) return; // CraftBukkit
+        this.listeners.add(listener);
     }
 
     public void removeListener(BorderChangeListener borderchangelistener) {
