--- a/net/minecraft/world/item/enchantment/DamageEnchantment.java
+++ b/net/minecraft/world/item/enchantment/DamageEnchantment.java
@@ -70,7 +63,7 @@
             if (this.type == 2 && i > 0 && livingentity1.getMobType() == MobType.ARTHROPOD) {
                 int j = 20 + livingentity.getRandom().nextInt(10 * i);
 
-                livingentity1.addEffect(new MobEffectInstance(MobEffects.MOVEMENT_SLOWDOWN, j, 3));
+                entityliving1.addEffect(new MobEffectInstance(MobEffects.MOVEMENT_SLOWDOWN, j, 3), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
         }
 
