--- a/net/minecraft/world/effect/HealOrHarmMobEffect.java
+++ b/net/minecraft/world/effect/HealOrHarmMobEffect.java
@@ -8,37 +8,35 @@
 
     private final boolean isHarm;
 
-    public HealOrHarmMobEffect(MobEffectCategory mobeffectcategory, int i, boolean flag) {
-        super(mobeffectcategory, i);
+    public HealOrHarmMobEffect(MobEffectCategory mobeffectinfo, int i, boolean flag) {
+        super(mobeffectinfo, i);
         this.isHarm = flag;
     }
 
     @Override
-    @Override
-    public void applyEffectTick(LivingEntity livingentity, int i) {
-        super.applyEffectTick(livingentity, i);
-        if (this.isHarm == livingentity.isInvertedHealAndHarm()) {
-            livingentity.heal((float) Math.max(4 << i, 0));
+    public void applyEffectTick(LivingEntity livingEntity, int amplifier) {
+        super.applyEffectTick(livingEntity, amplifier);
+        if (this.isHarm == livingEntity.isInvertedHealAndHarm()) {
+            livingEntity.heal((float) Math.max(4 << amplifier, 0), org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
-            livingentity.hurt(livingentity.damageSources().magic(), (float) (6 << i));
+            livingEntity.hurt(livingEntity.damageSources().magic(), (float) (6 << amplifier));
         }
 
     }
 
     @Override
-    @Override
-    public void applyInstantenousEffect(@Nullable Entity entity, @Nullable Entity entity1, LivingEntity livingentity, int i, double d0) {
+    public void applyInstantenousEffect(@Nullable Entity source, @Nullable Entity indirectSource, LivingEntity livingEntity, int amplifier, double health) {
         int j;
 
-        if (this.isHarm == livingentity.isInvertedHealAndHarm()) {
-            j = (int) (d0 * (double) (4 << i) + 0.5D);
-            livingentity.heal((float) j);
+        if (this.isHarm == livingEntity.isInvertedHealAndHarm()) {
+            j = (int) (health * (double) (4 << amplifier) + 0.5D);
+            livingEntity.heal((float) j, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.MAGIC); // CraftBukkit
         } else {
-            j = (int) (d0 * (double) (6 << i) + 0.5D);
-            if (entity == null) {
-                livingentity.hurt(livingentity.damageSources().magic(), (float) j);
+            j = (int) (health * (double) (6 << amplifier) + 0.5D);
+            if (source == null) {
+                livingEntity.hurt(livingEntity.damageSources().magic(), (float) j);
             } else {
-                livingentity.hurt(livingentity.damageSources().indirectMagic(entity, entity1), (float) j);
+                livingEntity.hurt(livingEntity.damageSources().indirectMagic(source, indirectSource), (float) j);
             }
         }
 
