--- a/net/minecraft/world/effect/HungerMobEffect.java
+++ b/net/minecraft/world/effect/HungerMobEffect.java
@@ -5,24 +5,22 @@
 
 class HungerMobEffect extends MobEffect {
 
-    protected HungerMobEffect(MobEffectCategory mobeffectcategory, int i) {
-        super(mobeffectcategory, i);
+    protected HungerMobEffect(MobEffectCategory category, int color) {
+        super(category, color);
     }
 
     @Override
-    @Override
-    public void applyEffectTick(LivingEntity livingentity, int i) {
-        super.applyEffectTick(livingentity, i);
-        if (livingentity instanceof Player) {
-            Player player = (Player) livingentity;
+    public void applyEffectTick(LivingEntity livingEntity, int amplifier) {
+        super.applyEffectTick(livingEntity, amplifier);
+        if (livingEntity instanceof Player) {
+            Player entityhuman = (Player) livingEntity;
 
-            player.causeFoodExhaustion(0.005F * (float) (i + 1));
+            entityhuman.causeFoodExhaustion(0.005F * (float) (amplifier + 1), org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.HUNGER_EFFECT); // CraftBukkit - EntityExhaustionEvent
         }
 
     }
 
     @Override
-    @Override
     public boolean shouldApplyEffectTickThisTick(int i, int j) {
         return true;
     }
