--- a/net/minecraft/world/entity/npc/InventoryCarrier.java
+++ b/net/minecraft/world/entity/npc/InventoryCarrier.java
@@ -12,24 +12,31 @@
 
     SimpleContainer getInventory();
 
-    static void pickUpItem(Mob mob, InventoryCarrier inventorycarrier, ItemEntity itementity) {
-        ItemStack itemstack = itementity.getItem();
+    static void pickUpItem(Mob mob, InventoryCarrier carrier, ItemEntity itemEntity) {
+        ItemStack itemstack = itemEntity.getItem();
 
         if (mob.wantsToPickUp(itemstack)) {
-            SimpleContainer simplecontainer = inventorycarrier.getInventory();
-            boolean flag = simplecontainer.canAddItem(itemstack);
+            SimpleContainer inventorysubcontainer = carrier.getInventory();
+            boolean flag = inventorysubcontainer.canAddItem(itemstack);
 
             if (!flag) {
                 return;
             }
 
-            mob.onItemPickup(itementity);
+            // CraftBukkit start
+            ItemStack remaining = new SimpleContainer(inventorysubcontainer).addItem(itemstack);
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(mob, itemEntity, remaining.getCount(), false).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+
+            mob.onItemPickup(itemEntity);
             int i = itemstack.getCount();
-            ItemStack itemstack1 = simplecontainer.addItem(itemstack);
+            ItemStack itemstack1 = inventorysubcontainer.addItem(itemstack);
 
-            mob.take(itementity, i - itemstack1.getCount());
+            mob.take(itemEntity, i - itemstack1.getCount());
             if (itemstack1.isEmpty()) {
-                itementity.discard();
+                itemEntity.discard();
             } else {
                 itemstack.setCount(itemstack1.getCount());
             }
@@ -37,14 +44,14 @@
 
     }
 
-    default void readInventoryFromTag(CompoundTag compoundtag) {
-        if (compoundtag.contains("Inventory", 9)) {
-            this.getInventory().fromTag(compoundtag.getList("Inventory", 10));
+    default void readInventoryFromTag(CompoundTag tag) {
+        if (tag.contains("Inventory", 9)) {
+            this.getInventory().fromTag(tag.getList("Inventory", 10));
         }
 
     }
 
-    default void writeInventoryToTag(CompoundTag compoundtag) {
-        compoundtag.put("Inventory", this.getInventory().createTag());
+    default void writeInventoryToTag(CompoundTag tag) {
+        tag.put("Inventory", this.getInventory().createTag());
     }
 }
