--- a/net/minecraft/world/level/LevelWriter.java
+++ b/net/minecraft/world/level/LevelWriter.java
@@ -3,24 +3,24 @@
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.Entity;
-import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.IBlockData;
 
 public interface LevelWriter {
 
-    boolean setBlock(BlockPos pos, BlockState state, int flags, int recursionLeft);
+    boolean setBlock(BlockPos pos, IBlockData state, int flags, int recursionLeft);
 
-    default boolean setBlock(BlockPos blockpos, BlockState blockstate, int i) {
-        return this.setBlock(blockpos, blockstate, i, 512);
+    default boolean setBlock(BlockPos pos, IBlockData newState, int flags) {
+        return this.setBlock(pos, newState, flags, 512);
     }
 
     boolean removeBlock(BlockPos pos, boolean isMoving);
 
-    default boolean destroyBlock(BlockPos blockpos, boolean flag) {
-        return this.destroyBlock(blockpos, flag, (Entity) null);
+    default boolean destroyBlock(BlockPos pos, boolean dropBlock) {
+        return this.destroyBlock(pos, dropBlock, (Entity) null);
     }
 
-    default boolean destroyBlock(BlockPos blockpos, boolean flag, @Nullable Entity entity) {
-        return this.destroyBlock(blockpos, flag, entity, 512);
+    default boolean destroyBlock(BlockPos pos, boolean dropBlock, @Nullable Entity entity) {
+        return this.destroyBlock(pos, dropBlock, entity, 512);
     }
 
     boolean destroyBlock(BlockPos pos, boolean dropBlock, @Nullable Entity entity, int recursionLeft);
@@ -28,4 +28,10 @@
     default boolean addFreshEntity(Entity entity) {
         return false;
     }
+
+    // CraftBukkit start
+    default boolean addFreshEntity(Entity entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason reason) {
+        return false;
+    }
+    // CraftBukkit end
 }
