--- a/net/minecraft/world/level/block/entity/BannerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BannerBlockEntity.java
@@ -105,7 +101,12 @@
             this.name = Component.Serializer.fromJson(compoundtag.getString("CustomName"));
         }
 
-        this.itemPatterns = compoundtag.getList("Patterns", 10);
+        this.itemPatterns = tag.getList("Patterns", 10);
+        // CraftBukkit start
+        while (this.itemPatterns.size() > 20) {
+            this.itemPatterns.remove(20);
+        }
+        // CraftBukkit end
         this.patterns = null;
     }
 
