--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -39,6 +39,10 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.phys.AABB;
+// CraftBukkit start
+import org.bukkit.craftbukkit.potion.CraftPotionUtil;
+import org.bukkit.potion.PotionEffect;
+// CraftBukkit end
 
 public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Nameable {
 
@@ -65,7 +69,16 @@
     private Component name;
     private LockCode lockKey;
     private final ContainerData dataAccess;
+    // CraftBukkit start - add fields and methods
+    public PotionEffect getPrimaryEffect() {
+        return (this.primaryPower != null) ? CraftPotionUtil.toBukkit(new MobEffectInstance(this.primaryPower, getLevel(this.levels), getAmplification(levels, primaryPower, secondaryPower), true, true)) : null;
+    }
 
+    public PotionEffect getSecondaryEffect() {
+        return (hasSecondaryEffect(levels, primaryPower, secondaryPower)) ? CraftPotionUtil.toBukkit(new MobEffectInstance(this.secondaryPower, getLevel(this.levels), getAmplification(levels, primaryPower, secondaryPower), true, true)) : null;
+    }
+    // CraftBukkit end
+
     @Nullable
     static MobEffect filterEffect(@Nullable MobEffect mobeffect) {
         return BeaconBlockEntity.VALID_EFFECTS.contains(mobeffect) ? mobeffect : null;
@@ -249,9 +258,9 @@
         super.setRemoved();
     }
 
-    private static void applyEffects(Level level, BlockPos blockpos, int i, @Nullable MobEffect mobeffect, @Nullable MobEffect mobeffect1) {
-        if (!level.isClientSide && mobeffect != null) {
-            double d0 = (double) (i * 10 + 10);
+    // CraftBukkit start - split into components
+    private static byte getAmplification(int i, @Nullable MobEffect mobeffectlist, @Nullable MobEffect mobeffectlist1) {
+        {
             byte b0 = 0;
 
             if (i >= 4 && mobeffect == mobeffect1) {
@@ -285,6 +328,7 @@
     public static void playSound(Level level, BlockPos blockpos, SoundEvent soundevent) {
         level.playSound((Player) null, blockpos, soundevent, SoundSource.BLOCKS, 1.0F, 1.0F);
     }
+    // CraftBukkit end
 
     public List<BeaconBlockEntity.BeaconBeamSection> getBeamSections() {
         return (List) (this.levels == 0 ? ImmutableList.of() : this.beamSections);
@@ -318,20 +364,20 @@
         if (compoundtag.contains(s, 8)) {
             ResourceLocation resourcelocation = ResourceLocation.tryParse(compoundtag.getString(s));
 
-            return filterEffect((MobEffect) BuiltInRegistries.MOB_EFFECT.get(resourcelocation));
+            return (MobEffect) BuiltInRegistries.MOB_EFFECT.get(minecraftkey); // CraftBukkit - persist manually set non-default beacon effects (SPIGOT-3598)
         } else {
             return null;
         }
     }
 
     @Override
-    @Override
-    public void load(CompoundTag compoundtag) {
-        super.load(compoundtag);
-        this.primaryPower = loadEffect(compoundtag, "primary_effect");
-        this.secondaryPower = loadEffect(compoundtag, "secondary_effect");
-        if (compoundtag.contains("CustomName", 8)) {
-            this.name = Component.Serializer.fromJson(compoundtag.getString("CustomName"));
+    public void load(CompoundTag tag) {
+        super.load(tag);
+        this.primaryPower = loadEffect(tag, "primary_effect");
+        this.secondaryPower = loadEffect(tag, "secondary_effect");
+        this.levels = tag.getInt("Levels"); // CraftBukkit - SPIGOT-5053, use where available
+        if (tag.contains("CustomName", 8)) {
+            this.name = Component.Serializer.fromJson(tag.getString("CustomName"));
         }
 
         this.lockKey = LockCode.fromTag(compoundtag);
