--- a/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
@@ -28,6 +28,12 @@
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.util.Arrays;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
+
 public class BrushableBlockEntity extends BlockEntity {
 
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -152,8 +158,11 @@
                 double d5 = (double) blockpos.getZ() + 0.5D * d1 + d2;
                 ItemEntity itementity = new ItemEntity(this.level, d3, d4, d5, this.item.split(this.level.random.nextInt(21) + 10));
 
-                itementity.setDeltaMovement(Vec3.ZERO);
-                this.level.addFreshEntity(itementity);
+                entityitem.setDeltaMovement(Vec3.ZERO);
+                // CraftBukkit start
+                org.bukkit.block.Block bblock = CraftBlock.at(this.level, this.worldPosition);
+                CraftEventFactory.handleBlockDropItemEvent(bblock, bblock.getState(), (ServerPlayer) player, Arrays.asList(entityitem));
+                // CraftBukkit end
                 this.item = ItemStack.EMPTY;
             }
 
@@ -231,10 +238,10 @@
     }
 
     @Override
-    @Override
-    public void load(CompoundTag compoundtag) {
-        if (!this.tryLoadLootTable(compoundtag) && compoundtag.contains("item")) {
-            this.item = ItemStack.of(compoundtag.getCompound("item"));
+    public void load(CompoundTag tag) {
+        super.load(tag); // CraftBukkit - SPIGOT-7393: Load super Bukkit data
+        if (!this.tryLoadLootTable(tag) && tag.contains("item")) {
+            this.item = ItemStack.of(tag.getCompound("item"));
         }
 
         if (compoundtag.contains("hit_direction")) {
