--- a/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ConduitBlockEntity.java
@@ -26,6 +26,10 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
+// CraftBukkit start
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
 
 public class ConduitBlockEntity extends BlockEntity {
 
@@ -204,8 +204,8 @@
             while (iterator.hasNext()) {
                 Player player = (Player) iterator.next();
 
-                if (blockpos.closerThan(player.blockPosition(), (double) j) && player.isInWaterOrRain()) {
-                    player.addEffect(new MobEffectInstance(MobEffects.CONDUIT_POWER, 260, 0, true, true));
+                if (pos.closerThan(entityhuman.blockPosition(), (double) j) && entityhuman.isInWaterOrRain()) {
+                    entityhuman.addEffect(new MobEffectInstance(MobEffects.CONDUIT_POWER, 260, 0, true, true), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONDUIT); // CraftBukkit
                 }
             }
 
@@ -233,9 +233,14 @@
             conduitblockentity.destroyTarget = null;
         }
 
-        if (conduitblockentity.destroyTarget != null) {
-            level.playSound((Player) null, conduitblockentity.destroyTarget.getX(), conduitblockentity.destroyTarget.getY(), conduitblockentity.destroyTarget.getZ(), SoundEvents.CONDUIT_ATTACK_TARGET, SoundSource.BLOCKS, 1.0F, 1.0F);
-            conduitblockentity.destroyTarget.hurt(level.damageSources().magic(), 4.0F);
+        if (blockEntity.destroyTarget != null) {
+            // CraftBukkit start
+            CraftEventFactory.blockDamage = CraftBlock.at(level, pos);
+            if (blockEntity.destroyTarget.hurt(level.damageSources().magic(), 4.0F)) {
+                level.playSound((Player) null, blockEntity.destroyTarget.getX(), blockEntity.destroyTarget.getY(), blockEntity.destroyTarget.getZ(), SoundEvents.CONDUIT_ATTACK_TARGET, SoundSource.BLOCKS, 1.0F, 1.0F);
+            }
+            CraftEventFactory.blockDamage = null;
+            // CraftBukkit end
         }
 
         if (livingentity != conduitblockentity.destroyTarget) {
