--- a/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
+++ b/net/minecraft/world/level/block/entity/ContainerOpenersCounter.java
@@ -13,6 +13,7 @@
 
     private static final int CHECK_TICK_DELAY = 5;
     private int openCount;
+    public boolean opened; // CraftBukkit
 
     public ContainerOpenersCounter() {}
 
@@ -22,11 +23,36 @@
 
     protected abstract void openerCountChanged(Level level, BlockPos pos, BlockState state, int count, int openCount);
 
+    // CraftBukkit start
+    public void onAPIOpen(Level world, BlockPos blockposition, IBlockData iblockdata) {
+        onOpen(world, blockposition, iblockdata);
+    }
+
+    public void onAPIClose(Level world, BlockPos blockposition, IBlockData iblockdata) {
+        onClose(world, blockposition, iblockdata);
+    }
+
+    public void openerAPICountChanged(Level world, BlockPos blockposition, IBlockData iblockdata, int i, int j) {
+        openerCountChanged(world, blockposition, iblockdata, i, j);
+    }
+    // CraftBukkit end
+
     protected abstract boolean isOwnContainer(Player player);
 
-    public void incrementOpeners(Player player, Level level, BlockPos blockpos, BlockState blockstate) {
+    public void incrementOpeners(Player player, Level level, BlockPos pos, IBlockData state) {
+        int oldPower = Math.max(0, Math.min(15, this.openCount)); // CraftBukkit - Get power before new viewer is added
         int i = this.openCount++;
 
+        // CraftBukkit start - Call redstone event
+        if (level.getBlockState(pos).is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
+            int newPower = Math.max(0, Math.min(15, this.openCount));
+
+            if (oldPower != newPower) {
+                org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, pos, oldPower, newPower);
+            }
+        }
+        // CraftBukkit end
+
         if (i == 0) {
             this.onOpen(level, blockpos, blockstate);
             level.gameEvent((Entity) player, GameEvent.CONTAINER_OPEN, blockpos);
@@ -36,9 +62,20 @@
         this.openerCountChanged(level, blockpos, blockstate, i, this.openCount);
     }
 
-    public void decrementOpeners(Player player, Level level, BlockPos blockpos, BlockState blockstate) {
+    public void decrementOpeners(Player player, Level level, BlockPos pos, IBlockData state) {
+        int oldPower = Math.max(0, Math.min(15, this.openCount)); // CraftBukkit - Get power before new viewer is added
         int i = this.openCount--;
 
+        // CraftBukkit start - Call redstone event
+        if (level.getBlockState(pos).is(net.minecraft.world.level.block.Blocks.TRAPPED_CHEST)) {
+            int newPower = Math.max(0, Math.min(15, this.openCount));
+
+            if (oldPower != newPower) {
+                org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, pos, oldPower, newPower);
+            }
+        }
+        // CraftBukkit end
+
         if (this.openCount == 0) {
             this.onClose(level, blockpos, blockstate);
             level.gameEvent((Entity) player, GameEvent.CONTAINER_CLOSE, blockpos);
@@ -57,8 +94,9 @@
         return level.getEntities(EntityTypeTest.forClass(Player.class), aabb, this::isOwnContainer).size();
     }
 
-    public void recheckOpeners(Level level, BlockPos blockpos, BlockState blockstate) {
-        int i = this.getOpenCount(level, blockpos);
+    public void recheckOpeners(Level level, BlockPos pos, IBlockData state) {
+        int i = this.getOpenCount(level, pos);
+        if (opened) i++; // CraftBukkit - add dummy count from API
         int j = this.openCount;
 
         if (j != i) {
