--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -32,6 +33,11 @@
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.slf4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.util.CraftLocation;
+import org.bukkit.event.player.PlayerTeleportEvent;
+// CraftBukkit end
 
 public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
 
@@ -174,12 +175,12 @@
             theendgatewayblockentity.teleportCooldown = 100;
             BlockPos blockpos1;
 
-            if (theendgatewayblockentity.exitPortal == null && level.dimension() == Level.END) {
-                blockpos1 = findOrCreateValidTeleportPos(serverlevel, blockpos);
-                blockpos1 = blockpos1.above(10);
-                TheEndGatewayBlockEntity.LOGGER.debug("Creating portal at {}", blockpos1);
-                spawnGatewayPortal(serverlevel, blockpos1, EndGatewayConfiguration.knownExit(blockpos, false));
-                theendgatewayblockentity.exitPortal = blockpos1;
+            if (blockEntity.exitPortal == null && level.getTypeKey() == LevelStem.END) { // CraftBukkit - work in alternate worlds
+                blockposition1 = findOrCreateValidTeleportPos(worldserver, pos);
+                blockposition1 = blockposition1.above(10);
+                TheEndGatewayBlockEntity.LOGGER.debug("Creating portal at {}", blockposition1);
+                spawnGatewayPortal(worldserver, blockposition1, EndGatewayConfiguration.knownExit(pos, false));
+                blockEntity.exitPortal = blockposition1;
             }
 
             if (theendgatewayblockentity.exitPortal != null) {
@@ -203,8 +204,34 @@
                     entity1 = entity.getRootVehicle();
                 }
 
+                // CraftBukkit start - Fire PlayerTeleportEvent/EntityTeleportEvent
+                if (entity1 instanceof ServerPlayer) {
+                    org.bukkit.craftbukkit.entity.CraftPlayer player = (CraftPlayer) entity1.getBukkitEntity();
+                    org.bukkit.Location location = CraftLocation.toBukkit(blockposition1, level.getWorld()).add(0.5D, 0.5D, 0.5D);
+                    location.setPitch(player.getLocation().getPitch());
+                    location.setYaw(player.getLocation().getYaw());
+
+                    PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), location, PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+                    Bukkit.getPluginManager().callEvent(teleEvent);
+                    if (teleEvent.isCancelled()) {
+                        return;
+                    }
+
+                    entity1.setPortalCooldown();
+                    ((ServerPlayer) entity1).connection.teleport(teleEvent.getTo());
+                    triggerCooldown(level, pos, state, blockEntity); // CraftBukkit - call at end of method
+                    return;
+
+                }
+
+                org.bukkit.event.entity.EntityTeleportEvent teleEvent = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTeleportEvent(entity1, blockposition1.getX() + 0.5, blockposition1.getY() + 0.5, blockposition1.getZ() + 0.5);
+                if (teleEvent.isCancelled()) {
+                    return;
+                }
+
                 entity1.setPortalCooldown();
-                entity1.teleportToWithTicket((double) blockpos1.getX() + 0.5D, (double) blockpos1.getY(), (double) blockpos1.getZ() + 0.5D);
+                entity1.teleportToWithTicket(teleEvent.getTo().getX(), teleEvent.getTo().getY(), teleEvent.getTo().getZ());
+                // CraftBukkit end
             }
 
             triggerCooldown(level, blockpos, blockstate, theendgatewayblockentity);
