--- a/net/minecraft/world/effect/PoisonMobEffect.java
+++ b/net/minecraft/world/effect/PoisonMobEffect.java
@@ -9,11 +9,10 @@
     }
 
     @Override
-    @Override
-    public void applyEffectTick(LivingEntity livingentity, int i) {
-        super.applyEffectTick(livingentity, i);
-        if (livingentity.getHealth() > 1.0F) {
-            livingentity.hurt(livingentity.damageSources().magic(), 1.0F);
+    public void applyEffectTick(LivingEntity livingEntity, int amplifier) {
+        super.applyEffectTick(livingEntity, amplifier);
+        if (livingEntity.getHealth() > 1.0F) {
+            livingEntity.hurt(livingEntity.damageSources().poison, 1.0F);  // CraftBukkit - DamageSource.MAGIC -> CraftEventFactory.POISON
         }
 
     }
