--- a/net/minecraft/world/level/levelgen/structure/structures/EndCityPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/EndCityPieces.java
@@ -387,9 +281,15 @@
         @Override
         protected void handleDataMarker(String name, BlockPos pos, ServerLevelAccessor level, RandomSource random, BoundingBox box) {
             if (name.startsWith("Chest")) {
-                BlockPos blockPos = pos.below();
-                if (box.isInside(blockPos)) {
-                    RandomizableContainer.setBlockEntityLootTable(level, random, blockPos, BuiltInLootTables.END_CITY_TREASURE);
+                BlockPos blockposition1 = pos.below();
+
+                if (box.isInside(blockposition1)) {
+                    // CraftBukkit start - ensure block transformation
+                    /*
+                    RandomizableContainer.setBlockEntityLootTable(worldaccess, randomsource, blockposition1, LootTables.END_CITY_TREASURE);
+                    */
+                    setCraftLootTable(level, blockposition1, random, BuiltInLootTables.END_CITY_TREASURE);
+                    // CraftBukkit end
                 }
             } else if (box.isInside(pos) && Level.isInSpawnableBounds(pos)) {
                 if (name.startsWith("Sentry")) {
