--- a/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
@@ -92,12 +94,13 @@
                 BlockPos worldPos = this.getWorldPos(2, 2, 5);
                 if (box.isInside(worldPos)) {
                     this.spawnedWitch = true;
-                    Witch witch = EntityType.WITCH.create(level.getLevel());
-                    if (witch != null) {
-                        witch.setPersistenceRequired();
-                        witch.moveTo((double)worldPos.getX() + 0.5, (double)worldPos.getY(), (double)worldPos.getZ() + 0.5, 0.0F, 0.0F);
-                        witch.finalizeSpawn(level, level.getCurrentDifficultyAt(worldPos), MobSpawnType.STRUCTURE, null, null);
-                        level.addFreshEntityWithPassengers(witch);
+                    Witch entitywitch = (Witch) EntityType.WITCH.create(level.getLevel());
+
+                    if (entitywitch != null) {
+                        entitywitch.setPersistenceRequired();
+                        entitywitch.moveTo((double) blockposition_mutableblockposition.getX() + 0.5D, (double) blockposition_mutableblockposition.getY(), (double) blockposition_mutableblockposition.getZ() + 0.5D, 0.0F, 0.0F);
+                        entitywitch.finalizeSpawn(level, level.getCurrentDifficultyAt(blockposition_mutableblockposition), EnumMobSpawn.STRUCTURE, (GroupDataEntity) null, (CompoundTag) null);
+                        level.addFreshEntityWithPassengers(entitywitch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit - add SpawnReason
                     }
                 }
             }
@@ -111,12 +115,13 @@
             BlockPos worldPos = this.getWorldPos(2, 2, 5);
             if (box.isInside(worldPos)) {
                 this.spawnedCat = true;
-                Cat cat = EntityType.CAT.create(level.getLevel());
-                if (cat != null) {
-                    cat.setPersistenceRequired();
-                    cat.moveTo((double)worldPos.getX() + 0.5, (double)worldPos.getY(), (double)worldPos.getZ() + 0.5, 0.0F, 0.0F);
-                    cat.finalizeSpawn(level, level.getCurrentDifficultyAt(worldPos), MobSpawnType.STRUCTURE, null, null);
-                    level.addFreshEntityWithPassengers(cat);
+                Cat entitycat = (Cat) EntityType.CAT.create(level.getLevel());
+
+                if (entitycat != null) {
+                    entitycat.setPersistenceRequired();
+                    entitycat.moveTo((double) blockposition_mutableblockposition.getX() + 0.5D, (double) blockposition_mutableblockposition.getY(), (double) blockposition_mutableblockposition.getZ() + 0.5D, 0.0F, 0.0F);
+                    entitycat.finalizeSpawn(level, level.getCurrentDifficultyAt(blockposition_mutableblockposition), EnumMobSpawn.STRUCTURE, (GroupDataEntity) null, (CompoundTag) null);
+                    level.addFreshEntityWithPassengers(entitycat, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit - add SpawnReason
                 }
             }
         }
