--- a/net/minecraft/network/protocol/game/ClientboundSetBorderCenterPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSetBorderCenterPacket.java
@@ -9,26 +9,26 @@
     private final double newCenterX;
     private final double newCenterZ;
 
-    public ClientboundSetBorderCenterPacket(WorldBorder worldborder) {
-        this.newCenterX = worldborder.getCenterX();
-        this.newCenterZ = worldborder.getCenterZ();
+    public ClientboundSetBorderCenterPacket(WorldBorder worldBorder) {
+        // CraftBukkit start - multiply out nether border
+        this.newCenterX = worldBorder.getCenterX() * (worldBorder.world != null ? worldBorder.world.dimensionType().coordinateScale() : 1.0);
+        this.newCenterZ = worldBorder.getCenterZ() * (worldBorder.world != null ? worldBorder.world.dimensionType().coordinateScale() : 1.0);
+        // CraftBukkit end
     }
 
-    public ClientboundSetBorderCenterPacket(FriendlyByteBuf friendlybytebuf) {
-        this.newCenterX = friendlybytebuf.readDouble();
-        this.newCenterZ = friendlybytebuf.readDouble();
+    public ClientboundSetBorderCenterPacket(FriendlyByteBuf buffer) {
+        this.newCenterX = buffer.readDouble();
+        this.newCenterZ = buffer.readDouble();
     }
 
     @Override
-    @Override
-    public void write(FriendlyByteBuf friendlybytebuf) {
-        friendlybytebuf.writeDouble(this.newCenterX);
-        friendlybytebuf.writeDouble(this.newCenterZ);
+    public void write(FriendlyByteBuf buffer) {
+        buffer.writeDouble(this.newCenterX);
+        buffer.writeDouble(this.newCenterZ);
     }
 
-    @Override
-    public void handle(ClientGamePacketListener clientgamepacketlistener) {
-        clientgamepacketlistener.handleSetBorderCenter(this);
+    public void handle(ClientGamePacketListener handler) {
+        handler.handleSetBorderCenter(this);
     }
 
     public double getNewCenterZ() {
