--- a/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundInitializeBorderPacket.java
@@ -15,44 +15,44 @@
     private final int warningBlocks;
     private final int warningTime;
 
-    public ClientboundInitializeBorderPacket(FriendlyByteBuf friendlybytebuf) {
-        this.newCenterX = friendlybytebuf.readDouble();
-        this.newCenterZ = friendlybytebuf.readDouble();
-        this.oldSize = friendlybytebuf.readDouble();
-        this.newSize = friendlybytebuf.readDouble();
-        this.lerpTime = friendlybytebuf.readVarLong();
-        this.newAbsoluteMaxSize = friendlybytebuf.readVarInt();
-        this.warningBlocks = friendlybytebuf.readVarInt();
-        this.warningTime = friendlybytebuf.readVarInt();
+    public ClientboundInitializeBorderPacket(FriendlyByteBuf buffer) {
+        this.newCenterX = buffer.readDouble();
+        this.newCenterZ = buffer.readDouble();
+        this.oldSize = buffer.readDouble();
+        this.newSize = buffer.readDouble();
+        this.lerpTime = buffer.readVarLong();
+        this.newAbsoluteMaxSize = buffer.readVarInt();
+        this.warningBlocks = buffer.readVarInt();
+        this.warningTime = buffer.readVarInt();
     }
 
-    public ClientboundInitializeBorderPacket(WorldBorder worldborder) {
-        this.newCenterX = worldborder.getCenterX();
-        this.newCenterZ = worldborder.getCenterZ();
-        this.oldSize = worldborder.getSize();
-        this.newSize = worldborder.getLerpTarget();
-        this.lerpTime = worldborder.getLerpRemainingTime();
-        this.newAbsoluteMaxSize = worldborder.getAbsoluteMaxSize();
-        this.warningBlocks = worldborder.getWarningBlocks();
-        this.warningTime = worldborder.getWarningTime();
+    public ClientboundInitializeBorderPacket(WorldBorder worldBorder) {
+        // CraftBukkit start - multiply out nether border
+        this.newCenterX = worldBorder.getCenterX() * worldBorder.world.dimensionType().coordinateScale();
+        this.newCenterZ = worldBorder.getCenterZ() * worldBorder.world.dimensionType().coordinateScale();
+        // CraftBukkit end
+        this.oldSize = worldBorder.getSize();
+        this.newSize = worldBorder.getLerpTarget();
+        this.lerpTime = worldBorder.getLerpRemainingTime();
+        this.newAbsoluteMaxSize = worldBorder.getAbsoluteMaxSize();
+        this.warningBlocks = worldBorder.getWarningBlocks();
+        this.warningTime = worldBorder.getWarningTime();
     }
 
     @Override
-    @Override
-    public void write(FriendlyByteBuf friendlybytebuf) {
-        friendlybytebuf.writeDouble(this.newCenterX);
-        friendlybytebuf.writeDouble(this.newCenterZ);
-        friendlybytebuf.writeDouble(this.oldSize);
-        friendlybytebuf.writeDouble(this.newSize);
-        friendlybytebuf.writeVarLong(this.lerpTime);
-        friendlybytebuf.writeVarInt(this.newAbsoluteMaxSize);
-        friendlybytebuf.writeVarInt(this.warningBlocks);
-        friendlybytebuf.writeVarInt(this.warningTime);
+    public void write(FriendlyByteBuf buffer) {
+        buffer.writeDouble(this.newCenterX);
+        buffer.writeDouble(this.newCenterZ);
+        buffer.writeDouble(this.oldSize);
+        buffer.writeDouble(this.newSize);
+        buffer.writeVarLong(this.lerpTime);
+        buffer.writeVarInt(this.newAbsoluteMaxSize);
+        buffer.writeVarInt(this.warningBlocks);
+        buffer.writeVarInt(this.warningTime);
     }
 
-    @Override
-    public void handle(ClientGamePacketListener clientgamepacketlistener) {
-        clientgamepacketlistener.handleInitializeBorder(this);
+    public void handle(ClientGamePacketListener handler) {
+        handler.handleInitializeBorder(this);
     }
 
     public double getNewCenterX() {
