--- a/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
@@ -1,3 +1,4 @@
+// mc-dev import
 package net.minecraft.network.protocol.game;
 
 import net.minecraft.network.FriendlyByteBuf;
@@ -6,24 +7,21 @@
 
 public record ClientboundSystemChatPacket(Component content, boolean overlay) implements Packet<ClientGamePacketListener> {
 
-    public ClientboundSystemChatPacket(FriendlyByteBuf friendlybytebuf) {
-        this(friendlybytebuf.readComponentTrusted(), friendlybytebuf.readBoolean());
+    public ClientboundSystemChatPacket(FriendlyByteBuf buffer) {
+        this(buffer.readComponentTrusted(), buffer.readBoolean());
     }
 
     @Override
-    @Override
-    public void write(FriendlyByteBuf friendlybytebuf) {
-        friendlybytebuf.writeComponent(this.content);
-        friendlybytebuf.writeBoolean(this.overlay);
+    public void write(FriendlyByteBuf buffer) {
+        buffer.writeComponent(this.content);
+        buffer.writeBoolean(this.overlay);
     }
 
-    @Override
-    public void handle(ClientGamePacketListener clientgamepacketlistener) {
-        clientgamepacketlistener.handleSystemChat(this);
+    public void handle(ClientGamePacketListener handler) {
+        handler.handleSystemChat(this);
     }
 
     @Override
-    @Override
     public boolean isSkippable() {
         return true;
     }
