--- a/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundUseItemPacket.java
@@ -1,37 +1,36 @@
+// mc-dev import
 package net.minecraft.network.protocol.game;
 
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
-import net.minecraft.world.InteractionHand;
+import net.minecraft.world.EnumHand;
 
 public class ServerboundUseItemPacket implements Packet<ServerGamePacketListener> {
 
-    private final InteractionHand hand;
+    private final EnumHand hand;
     private final int sequence;
 
-    public ServerboundUseItemPacket(InteractionHand interactionhand, int i) {
-        this.hand = interactionhand;
-        this.sequence = i;
+    public ServerboundUseItemPacket(EnumHand hand, int sequence) {
+        this.hand = hand;
+        this.sequence = sequence;
     }
 
-    public ServerboundUseItemPacket(FriendlyByteBuf friendlybytebuf) {
-        this.hand = (InteractionHand) friendlybytebuf.readEnum(InteractionHand.class);
-        this.sequence = friendlybytebuf.readVarInt();
+    public ServerboundUseItemPacket(FriendlyByteBuf buffer) {
+        this.hand = (EnumHand) buffer.readEnum(EnumHand.class);
+        this.sequence = buffer.readVarInt();
     }
 
     @Override
-    @Override
-    public void write(FriendlyByteBuf friendlybytebuf) {
-        friendlybytebuf.writeEnum(this.hand);
-        friendlybytebuf.writeVarInt(this.sequence);
+    public void write(FriendlyByteBuf buffer) {
+        buffer.writeEnum(this.hand);
+        buffer.writeVarInt(this.sequence);
     }
 
-    @Override
-    public void handle(ServerGamePacketListener servergamepacketlistener) {
-        servergamepacketlistener.handleUseItem(this);
+    public void handle(ServerGamePacketListener handler) {
+        handler.handleUseItem(this);
     }
 
-    public InteractionHand getHand() {
+    public EnumHand getHand() {
         return this.hand;
     }
 
