--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -151,6 +166,7 @@
         this.finalizers.forEach(Runnable::run);
     }
 
+    private static final java.util.regex.Pattern ANSI = java.util.regex.Pattern.compile("\\x1B\\[([0-9]{1,2}(;[0-9]{1,2})*)?[m|K]"); // CraftBukkit
     public void print(JTextArea textArea, JScrollPane scrollPane, String line) {
         if (!SwingUtilities.isEventDispatchThread()) {
             SwingUtilities.invokeLater(() -> this.print(textArea, scrollPane, line));
@@ -164,8 +182,9 @@
             }
 
             try {
-                document.insertString(document.getLength(), line, null);
-            } catch (BadLocationException var8) {
+                document.insertString(document.getLength(), ANSI.matcher(line).replaceAll(""), (AttributeSet) null); // CraftBukkit
+            } catch (BadLocationException badlocationexception) {
+                ;
             }
 
             if (flag) {
