--- a/net/minecraft/world/entity/projectile/Arrow.java
+++ b/net/minecraft/world/entity/projectile/Arrow.java
@@ -177,22 +207,21 @@
         super.doPostHurtEffects(living);
         Entity effectSource = this.getEffectSource();
 
-        for (MobEffectInstance mobEffectInstance : this.potion.getEffects()) {
-            living.addEffect(
-                new MobEffectInstance(
-                    mobEffectInstance.getEffect(),
-                    Math.max(mobEffectInstance.mapDuration(i -> i / 8), 1),
-                    mobEffectInstance.getAmplifier(),
-                    mobEffectInstance.isAmbient(),
-                    mobEffectInstance.isVisible()
-                ),
-                effectSource
-            );
+        MobEffectInstance mobeffect;
+
+        while (iterator.hasNext()) {
+            mobeffect = (MobEffectInstance) iterator.next();
+            living.addEffect(new MobEffectInstance(mobeffect.getEffect(), Math.max(mobeffect.mapDuration((i) -> {
+                return i / 8;
+            }), 1), mobeffect.getAmplifier(), mobeffect.isAmbient(), mobeffect.isVisible()), entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
         }
 
         if (!this.effects.isEmpty()) {
-            for (MobEffectInstance mobEffectInstance : this.effects) {
-                living.addEffect(mobEffectInstance, effectSource);
+            iterator = this.effects.iterator();
+
+            while (iterator.hasNext()) {
+                mobeffect = (MobEffectInstance) iterator.next();
+                living.addEffect(mobeffect, entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
             }
         }
     }
