--- a/net/minecraft/world/entity/projectile/ThrowableProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableProjectile.java
@@ -60,8 +65,8 @@
             }
         }
 
-        if (hitResultOnMoveVector.getType() != HitResult.Type.MISS && !flag) {
-            this.onHit(hitResultOnMoveVector);
+        if (movingobjectposition.getType() != HitResult.EnumMovingObjectType.MISS && !flag) {
+            this.preOnHit(movingobjectposition); // CraftBukkit - projectile hit event
         }
 
         this.checkInsideBlocks();
