--- a/net/minecraft/world/entity/projectile/ThrowableItemProjectile.java
+++ b/net/minecraft/world/entity/projectile/ThrowableItemProjectile.java
@@ -35,9 +35,11 @@
 
     protected abstract Item getDefaultItem();
 
-    protected ItemStack getItemRaw() {
-        return this.getEntityData().get(DATA_ITEM_STACK);
+    // CraftBukkit start
+    public Item getDefaultItemPublic() {
+        return getDefaultItem();
     }
+    // CraftBukkit end
 
     @Override
     public ItemStack getItem() {
