--- a/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundSystemChatPacket.java
@@ -1,3 +1,4 @@
+// mc-dev import
 package net.minecraft.network.protocol.game;
 
 import net.minecraft.network.FriendlyByteBuf;
@@ -5,6 +6,7 @@
 import net.minecraft.network.protocol.Packet;
 
 public record ClientboundSystemChatPacket(Component content, boolean overlay) implements Packet<ClientGamePacketListener> {
+
     public ClientboundSystemChatPacket(FriendlyByteBuf buffer) {
         this(buffer.readComponentTrusted(), buffer.readBoolean());
     }
@@ -15,7 +17,6 @@
         buffer.writeBoolean(this.overlay);
     }
 
-    @Override
     public void handle(ClientGamePacketListener handler) {
         handler.handleSystemChat(this);
     }
