--- a/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundUseItemOnPacket.java
@@ -1,23 +1,25 @@
+// mc-dev import
 package net.minecraft.network.protocol.game;
 
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.protocol.Packet;
-import net.minecraft.world.InteractionHand;
+import net.minecraft.world.EnumHand;
 import net.minecraft.world.phys.BlockHitResult;
 
 public class ServerboundUseItemOnPacket implements Packet<ServerGamePacketListener> {
+
     private final BlockHitResult blockHit;
-    private final InteractionHand hand;
+    private final EnumHand hand;
     private final int sequence;
 
-    public ServerboundUseItemOnPacket(InteractionHand hand, BlockHitResult blockHit, int sequence) {
+    public ServerboundUseItemOnPacket(EnumHand hand, BlockHitResult blockHit, int sequence) {
         this.hand = hand;
         this.blockHit = blockHit;
         this.sequence = sequence;
     }
 
     public ServerboundUseItemOnPacket(FriendlyByteBuf buffer) {
-        this.hand = buffer.readEnum(InteractionHand.class);
+        this.hand = (EnumHand) buffer.readEnum(EnumHand.class);
         this.blockHit = buffer.readBlockHitResult();
         this.sequence = buffer.readVarInt();
     }
@@ -29,12 +31,11 @@
         buffer.writeVarInt(this.sequence);
     }
 
-    @Override
     public void handle(ServerGamePacketListener handler) {
         handler.handleUseItemOn(this);
     }
 
-    public InteractionHand getHand() {
+    public EnumHand getHand() {
         return this.hand;
     }
 
