--- a/net/minecraft/world/level/storage/PlayerDataStorage.java
+++ b/net/minecraft/world/level/storage/PlayerDataStorage.java
@@ -15,6 +16,12 @@
 import net.minecraft.world.entity.player.Player;
 import org.slf4j.Logger;
 
+// CraftBukkit start
+import java.io.FileInputStream;
+import java.io.InputStream;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+// CraftBukkit end
+
 public class PlayerDataStorage {
     private static final Logger LOGGER = LogUtils.getLogger();
     private final File playerDir;
@@ -53,15 +65,42 @@
             LOGGER.warn("Failed to load player data for {}", player.getName().getString());
         }
 
-        if (compoundTag != null) {
-            int dataVersion = NbtUtils.getDataVersion(compoundTag, -1);
-            compoundTag = DataFixTypes.PLAYER.updateToCurrentVersion(this.fixerUpper, compoundTag, dataVersion);
-            player.load(compoundTag);
+        if (nbttagcompound != null) {
+            // CraftBukkit start
+            if (player instanceof ServerPlayer) {
+                CraftPlayer player = (CraftPlayer) player.getBukkitEntity();
+                // Only update first played if it is older than the one we have
+                long modified = new File(this.playerDir, player.getUUID().toString() + ".dat").lastModified();
+                if (modified < player.getFirstPlayed()) {
+                    player.setFirstPlayed(modified);
+                }
+            }
+            // CraftBukkit end
+            int i = NbtUtils.getDataVersion(nbttagcompound, -1);
+
+            nbttagcompound = DataFixTypes.PLAYER.updateToCurrentVersion(this.fixerUpper, nbttagcompound, i);
+            player.load(nbttagcompound);
         }
 
         return compoundTag;
     }
 
+    // CraftBukkit start
+    public CompoundTag getPlayerData(String s) {
+        try {
+            File file1 = new File(this.playerDir, s + ".dat");
+
+            if (file1.exists()) {
+                return NbtIo.readCompressed(file1.toPath(), NbtAccounter.unlimitedHeap());
+            }
+        } catch (Exception exception) {
+            LOGGER.warn("Failed to load player data for " + s);
+        }
+
+        return null;
+    }
+    // CraftBukkit end
+
     public String[] getSeenPlayers() {
         String[] strings = this.playerDir.list();
         if (strings == null) {
@@ -76,4 +116,10 @@
 
         return strings;
     }
+
+    // CraftBukkit start
+    public File getPlayerDir() {
+        return playerDir;
+    }
+    // CraftBukkit end
 }
