--- a/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
+++ b/net/minecraft/world/level/storage/loot/predicates/ExplosionCondition.java
@@ -16,26 +16,24 @@
     private ExplosionCondition() {}
 
     @Override
-    @Override
     public LootItemConditionType getType() {
         return LootItemConditions.SURVIVES_EXPLOSION;
     }
 
     @Override
-    @Override
     public Set<LootContextParam<?>> getReferencedContextParams() {
         return ImmutableSet.of(LootContextParams.EXPLOSION_RADIUS);
     }
 
-    @Override
-    public boolean test(LootContext lootcontext) {
-        Float ofloat = (Float) lootcontext.getParamOrNull(LootContextParams.EXPLOSION_RADIUS);
+    public boolean test(LootContext context) {
+        Float ofloat = (Float) context.getParamOrNull(LootContextParams.EXPLOSION_RADIUS);
 
         if (ofloat != null) {
-            RandomSource randomsource = lootcontext.getRandom();
+            RandomSource randomsource = context.getRandom();
             float f = 1.0F / ofloat;
 
-            return randomsource.nextFloat() <= f;
+            // CraftBukkit - <= to < to allow for plugins to completely disable block drops from explosions
+            return randomsource.nextFloat() < f;
         } else {
             return true;
         }
