--- a/net/minecraft/world/entity/vehicle/MinecartTNT.java
+++ b/net/minecraft/world/entity/vehicle/MinecartTNT.java
@@ -19,6 +22,9 @@
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.FluidState;
+// CraftBukkit start
+import org.bukkit.event.entity.ExplosionPrimeEvent;
+// CraftBukkit end
 
 public class MinecartTNT extends AbstractMinecart {
     private static final byte EVENT_PRIME = 10;
@@ -99,18 +118,15 @@
                 squareRoot = 5.0;
             }
 
-            this.level()
-                .explode(
-                    this,
-                    damageSource,
-                    null,
-                    this.getX(),
-                    this.getY(),
-                    this.getZ(),
-                    (float)(4.0 + this.random.nextDouble() * 1.5 * squareRoot),
-                    false,
-                    Level.ExplosionInteraction.TNT
-                );
+            // CraftBukkit start
+            ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), (float) (4.0D + this.random.nextDouble() * 1.5D * d1), false);
+            this.level().getCraftServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+                fuse = -1;
+                return;
+            }
+            this.level().explode(this, damageSource, (ExplosionDamageCalculator) null, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), Level.a.TNT);
+            // CraftBukkit end
             this.discard();
         }
     }
