--- a/net/minecraft/world/item/crafting/CustomRecipe.java
+++ b/net/minecraft/world/item/crafting/CustomRecipe.java
@@ -3,8 +3,13 @@
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.world.item.ItemStack;
 
-public abstract class CustomRecipe implements CraftingRecipe {
+// CraftBukkit start
+import org.bukkit.NamespacedKey;
+import org.bukkit.inventory.Recipe;
+// CraftBukkit end
 
+public abstract class CustomRecipe implements RecipeCrafting {
+
     private final CraftingBookCategory category;
 
     public CustomRecipe(CraftingBookCategory craftingbookcategory) {
@@ -12,20 +17,24 @@
     }
 
     @Override
-    @Override
     public boolean isSpecial() {
         return true;
     }
 
     @Override
-    @Override
-    public ItemStack getResultItem(RegistryAccess registryaccess) {
+    public ItemStack getResultItem(RegistryAccess registryAccess) {
         return ItemStack.EMPTY;
     }
 
     @Override
-    @Override
     public CraftingBookCategory category() {
         return this.category;
     }
+
+    // CraftBukkit start
+    @Override
+    public Recipe toBukkitRecipe(NamespacedKey id) {
+        return new org.bukkit.craftbukkit.inventory.CraftComplexRecipe(id, this);
+    }
+    // CraftBukkit end
 }
