--- a/net/minecraft/server/players/UserBanListEntry.java
+++ b/net/minecraft/server/players/UserBanListEntry.java
@@ -1,3 +1,4 @@
+// mc-dev import
 package net.minecraft.server.players;
 
 import com.google.gson.JsonObject;
@@ -9,30 +10,28 @@
 
 public class UserBanListEntry extends BanListEntry<GameProfile> {
 
-    public UserBanListEntry(@Nullable GameProfile gameprofile) {
-        this(gameprofile, (Date) null, (String) null, (Date) null, (String) null);
+    public UserBanListEntry(@Nullable GameProfile user) {
+        this(user, (Date) null, (String) null, (Date) null, (String) null);
     }
 
-    public UserBanListEntry(@Nullable GameProfile gameprofile, @Nullable Date date, @Nullable String s, @Nullable Date date1, @Nullable String s1) {
-        super(gameprofile, date, s, date1, s1);
+    public UserBanListEntry(@Nullable GameProfile profile, @Nullable Date created, @Nullable String source, @Nullable Date expires, @Nullable String reason) {
+        super(profile, created, source, expires, reason);
     }
 
-    public UserBanListEntry(JsonObject jsonobject) {
-        super(createGameProfile(jsonobject), jsonobject);
+    public UserBanListEntry(JsonObject entryData) {
+        super(createGameProfile(entryData), entryData);
     }
 
     @Override
-    @Override
-    protected void serialize(JsonObject jsonobject) {
+    protected void serialize(JsonObject data) {
         if (this.getUser() != null) {
-            jsonobject.addProperty("uuid", ((GameProfile) this.getUser()).getId().toString());
-            jsonobject.addProperty("name", ((GameProfile) this.getUser()).getName());
-            super.serialize(jsonobject);
+            data.addProperty("uuid", ((GameProfile) this.getUser()).getId().toString());
+            data.addProperty("name", ((GameProfile) this.getUser()).getName());
+            super.serialize(data);
         }
     }
 
     @Override
-    @Override
     public Component getDisplayName() {
         GameProfile gameprofile = (GameProfile) this.getUser();
 
@@ -40,9 +39,9 @@
     }
 
     @Nullable
-    private static GameProfile createGameProfile(JsonObject jsonobject) {
-        if (jsonobject.has("uuid") && jsonobject.has("name")) {
-            String s = jsonobject.get("uuid").getAsString();
+    private static GameProfile createGameProfile(JsonObject json) {
+        if (json.has("uuid") && json.has("name")) {
+            String s = json.get("uuid").getAsString();
 
             UUID uuid;
 
@@ -52,7 +51,7 @@
                 return null;
             }
 
-            return new GameProfile(uuid, jsonobject.get("name").getAsString());
+            return new GameProfile(uuid, json.get("name").getAsString());
         } else {
             return null;
         }
