--- a/net/minecraft/world/entity/animal/horse/Llama.java
+++ b/net/minecraft/world/entity/animal/horse/Llama.java
@@ -84,9 +84,14 @@
         return false;
     }
 
-    private void setStrength(int i) {
-        this.entityData.set(Llama.DATA_STRENGTH_ID, Math.max(1, Math.min(5, i)));
+    // CraftBukkit start
+    public void setStrengthPublic(int i) {
+        this.setStrength(i);
     }
+    // CraftBukkit end
+    private void setStrength(int strength) {
+        this.entityData.set(Llama.DATA_STRENGTH_ID, Math.max(1, Math.min(5, strength)));
+    }
 
     private void setRandomStrength(RandomSource randomsource) {
         int i = randomsource.nextFloat() < 0.04F ? 5 : 3;
