--- a/net/minecraft/world/entity/animal/Pig.java
+++ b/net/minecraft/world/entity/animal/Pig.java
@@ -51,7 +51,9 @@
 import net.minecraft.world.phys.Vec3;
 import org.joml.Vector3f;
 
-public class Pig extends Animal implements ItemSteerable, Saddleable {
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
 
     private static final EntityDataAccessor<Boolean> DATA_SADDLE_ID = SynchedEntityData.defineId(Pig.class, EntityDataSerializers.BOOLEAN);
     private static final EntityDataAccessor<Integer> DATA_BOOST_TIME = SynchedEntityData.defineId(Pig.class, EntityDataSerializers.INT);
@@ -268,8 +255,14 @@
                     zombifiedpiglin.setCustomNameVisible(this.isCustomNameVisible());
                 }
 
-                zombifiedpiglin.setPersistenceRequired();
-                serverlevel.addFreshEntity(zombifiedpiglin);
+                entitypigzombie.setPersistenceRequired();
+                // CraftBukkit start
+                if (CraftEventFactory.callPigZapEvent(this, lightning, entitypigzombie).isCancelled()) {
+                    return;
+                }
+                // CraftBukkit - added a reason for spawning this creature
+                level.addFreshEntity(entitypigzombie, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+                // CraftBukkit end
                 this.discard();
             } else {
                 super.thunderHit(serverlevel, lightningbolt);
