--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -476,7 +510,7 @@
         return this.previouslyKilled;
     }
 
-    public void tryRespawn() {
+    public boolean tryRespawn() { // CraftBukkit - return boolean
         if (this.dragonKilled && this.respawnStage == null) {
             BlockPos blockPos = this.portalLocation;
             if (blockPos == null) {
@@ -495,21 +532,24 @@
             List<EndCrystal> list = Lists.newArrayList();
             BlockPos blockPos1 = blockPos.above(1);
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
-                List<EndCrystal> entitiesOfClass = this.level.getEntitiesOfClass(EndCrystal.class, new AABB(blockPos1.relative(direction, 2)));
-                if (entitiesOfClass.isEmpty()) {
-                    return;
+            while (iterator.hasNext()) {
+                Direction enumdirection = (Direction) iterator.next();
+                List<EndCrystal> list1 = this.level.getEntitiesOfClass(EndCrystal.class, new AABB(blockposition1.relative(enumdirection, 2)));
+
+                if (list1.isEmpty()) {
+                    return false; // CraftBukkit - return value
                 }
 
                 list.addAll(entitiesOfClass);
             }
 
-            LOGGER.debug("Found all crystals, respawning dragon.");
-            this.respawnDragon(list);
+            EndDragonFight.LOGGER.debug("Found all crystals, respawning dragon.");
+            return this.respawnDragon(list); // CraftBukkit - return value
         }
+        return false; // CraftBukkit - return value
     }
 
-    private void respawnDragon(List<EndCrystal> crystals) {
+    public boolean respawnDragon(List<EndCrystal> list) { // CraftBukkit - return boolean
         if (this.dragonKilled && this.respawnStage == null) {
             for (BlockPattern.BlockPatternMatch blockPatternMatch = this.findExitPortal(); blockPatternMatch != null; blockPatternMatch = this.findExitPortal()) {
                 for (int i = 0; i < this.exitPortalPattern.getWidth(); i++) {
@@ -527,8 +568,10 @@
             this.respawnStage = DragonRespawnAnimation.START;
             this.respawnTime = 0;
             this.spawnExitPortal(false);
-            this.respawnCrystals = crystals;
+            this.respawnCrystals = list;
+            return true; // CraftBukkit - return value
         }
+        return false; // CraftBukkit - return value
     }
 
     public void resetSpikeCrystals() {
