--- a/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BaseContainerBlockEntity.java
@@ -7,21 +7,23 @@
 import net.minecraft.sounds.SoundEvents;
 import net.minecraft.sounds.SoundSource;
 import net.minecraft.world.Container;
+import net.minecraft.world.INamableTileEntity;
+import net.minecraft.world.ITileInventory;
 import net.minecraft.world.LockCode;
-import net.minecraft.world.MenuProvider;
-import net.minecraft.world.Nameable;
 import net.minecraft.world.entity.player.Inventory;
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.inventory.AbstractContainerMenu;
-import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.IBlockData;
 
-public abstract class BaseContainerBlockEntity extends BlockEntity implements Container, MenuProvider, Nameable {
-    private LockCode lockKey = LockCode.NO_LOCK;
+public abstract class BaseContainerBlockEntity extends BlockEntity implements Container, ITileInventory, INamableTileEntity {
+
+    public LockCode lockKey;
     @Nullable
-    private Component name;
+    public Component name;
 
-    protected BaseContainerBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState blockState) {
+    protected BaseContainerBlockEntity(BlockEntityType<?> type, BlockPos pos, IBlockData blockState) {
         super(type, pos, blockState);
+        this.lockKey = LockCode.NO_LOCK;
     }
 
     @Override
@@ -31,6 +33,7 @@
         if (tag.contains("CustomName", 8)) {
             this.name = Component.Serializer.fromJson(tag.getString("CustomName"));
         }
+
     }
 
     @Override
@@ -40,6 +43,7 @@
         if (this.name != null) {
             tag.putString("CustomName", Component.Serializer.toJson(this.name));
         }
+
     }
 
     public void setCustomName(Component name) {
@@ -85,4 +89,12 @@
     }
 
     protected abstract AbstractContainerMenu createMenu(int containerId, Inventory inventory);
+
+    // CraftBukkit start
+    @Override
+    public org.bukkit.Location getLocation() {
+        if (level == null) return null;
+        return new org.bukkit.Location(level.getWorld(), worldPosition.getX(), worldPosition.getY(), worldPosition.getZ());
+    }
+    // CraftBukkit end
 }
