--- a/net/minecraft/world/entity/animal/MushroomCow.java
+++ b/net/minecraft/world/entity/animal/MushroomCow.java
@@ -8,7 +8,6 @@
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtOps;
-import net.minecraft.nbt.Tag;
 import net.minecraft.network.syncher.EntityDataAccessor;
 import net.minecraft.network.syncher.EntityDataSerializers;
 import net.minecraft.network.syncher.SynchedEntityData;
@@ -20,12 +19,13 @@
 import net.minecraft.tags.ItemTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.util.StringRepresentable;
-import net.minecraft.world.InteractionHand;
+import net.minecraft.world.EnumHand;
 import net.minecraft.world.InteractionResult;
 import net.minecraft.world.entity.AgeableMob;
+import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.EnumMobSpawn;
 import net.minecraft.world.entity.LightningBolt;
-import net.minecraft.world.entity.MobSpawnType;
 import net.minecraft.world.entity.Shearable;
 import net.minecraft.world.entity.VariantHolder;
 import net.minecraft.world.entity.item.ItemEntity;
@@ -39,15 +39,23 @@
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.SuspiciousEffectHolder;
-import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.IBlockData;
 import net.minecraft.world.level.gameevent.GameEvent;
 
+// CraftBukkit start
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityDropItemEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
+// CraftBukkit end
+
 public class MushroomCow extends Cow implements Shearable, VariantHolder<MushroomCow.MushroomType> {
+
     private static final EntityDataAccessor<String> DATA_TYPE = SynchedEntityData.defineId(MushroomCow.class, EntityDataSerializers.STRING);
     private static final int MUTATE_CHANCE = 1024;
     private static final String TAG_STEW_EFFECTS = "stew_effects";
     @Nullable
-    private List<SuspiciousEffectHolder.EffectEntry> stewEffects;
+    private List<SuspiciousEffectHolder.a> stewEffects;
     @Nullable
     private UUID lastLightningBoltUUID;
 
@@ -60,100 +68,94 @@
         return level.getBlockState(pos.below()).is(Blocks.MYCELIUM) ? 10.0F : level.getPathfindingCostFromLightLevels(pos);
     }
 
-    public static boolean checkMushroomSpawnRules(
-        EntityType<MushroomCow> mushroomCow, LevelAccessor level, MobSpawnType spawnType, BlockPos pos, RandomSource randomSource
-    ) {
+    public static boolean checkMushroomSpawnRules(EntityType<MushroomCow> mushroomCow, LevelAccessor level, EnumMobSpawn spawnType, BlockPos pos, RandomSource randomSource) {
         return level.getBlockState(pos.below()).is(BlockTags.MOOSHROOMS_SPAWNABLE_ON) && isBrightEnoughToSpawn(level, pos);
     }
 
     @Override
     public void thunderHit(ServerLevel level, LightningBolt lightning) {
-        UUID uUID = lightning.getUUID();
-        if (!uUID.equals(this.lastLightningBoltUUID)) {
+        UUID uuid = lightning.getUUID();
+
+        if (!uuid.equals(this.lastLightningBoltUUID)) {
             this.setVariant(this.getVariant() == MushroomCow.MushroomType.RED ? MushroomCow.MushroomType.BROWN : MushroomCow.MushroomType.RED);
-            this.lastLightningBoltUUID = uUID;
+            this.lastLightningBoltUUID = uuid;
             this.playSound(SoundEvents.MOOSHROOM_CONVERT, 2.0F, 1.0F);
         }
+
     }
 
     @Override
     protected void defineSynchedData() {
         super.defineSynchedData();
-        this.entityData.define(DATA_TYPE, MushroomCow.MushroomType.RED.type);
+        this.entityData.define(MushroomCow.DATA_TYPE, MushroomCow.MushroomType.RED.type);
     }
 
     @Override
-    public InteractionResult mobInteract(Player player, InteractionHand hand) {
-        ItemStack itemInHand = player.getItemInHand(hand);
-        if (itemInHand.is(Items.BOWL) && !this.isBaby()) {
+    public InteractionResult mobInteract(Player player, EnumHand hand) {
+        ItemStack itemstack = player.getItemInHand(hand);
+
+        if (itemstack.is(Items.BOWL) && !this.isBaby()) {
             boolean flag = false;
-            ItemStack itemStack;
+            ItemStack itemstack1;
+
             if (this.stewEffects != null) {
                 flag = true;
-                itemStack = new ItemStack(Items.SUSPICIOUS_STEW);
-                SuspiciousStewItem.saveMobEffects(itemStack, this.stewEffects);
+                itemstack1 = new ItemStack(Items.SUSPICIOUS_STEW);
+                SuspiciousStewItem.saveMobEffects(itemstack1, this.stewEffects);
                 this.stewEffects = null;
             } else {
-                itemStack = new ItemStack(Items.MUSHROOM_STEW);
+                itemstack1 = new ItemStack(Items.MUSHROOM_STEW);
             }
 
-            ItemStack itemStack1 = ItemUtils.createFilledResult(itemInHand, player, itemStack, false);
-            player.setItemInHand(hand, itemStack1);
-            SoundEvent soundEvent;
+            ItemStack itemstack2 = ItemUtils.createFilledResult(itemstack, player, itemstack1, false);
+
+            player.setItemInHand(hand, itemstack2);
+            SoundEvent soundeffect;
+
             if (flag) {
-                soundEvent = SoundEvents.MOOSHROOM_MILK_SUSPICIOUSLY;
+                soundeffect = SoundEvents.MOOSHROOM_MILK_SUSPICIOUSLY;
             } else {
-                soundEvent = SoundEvents.MOOSHROOM_MILK;
+                soundeffect = SoundEvents.MOOSHROOM_MILK;
             }
 
-            this.playSound(soundEvent, 1.0F, 1.0F);
+            this.playSound(soundeffect, 1.0F, 1.0F);
             return InteractionResult.sidedSuccess(this.level().isClientSide);
-        } else if (itemInHand.is(Items.SHEARS) && this.readyForShearing()) {
+        } else if (itemstack.is(Items.SHEARS) && this.readyForShearing()) {
+            // CraftBukkit start
+            if (!CraftEventFactory.handlePlayerShearEntityEvent(player, this, itemstack, hand)) {
+                return InteractionResult.PASS;
+            }
+            // CraftBukkit end
             this.shear(SoundSource.PLAYERS);
             this.gameEvent(GameEvent.SHEAR, player);
             if (!this.level().isClientSide) {
-                itemInHand.hurtAndBreak(1, player, entity -> entity.broadcastBreakEvent(hand));
+                itemstack.hurtAndBreak(1, player, (entityhuman1) -> {
+                    entityhuman1.broadcastBreakEvent(hand);
+                });
             }
 
             return InteractionResult.sidedSuccess(this.level().isClientSide);
-        } else if (this.getVariant() == MushroomCow.MushroomType.BROWN && itemInHand.is(ItemTags.SMALL_FLOWERS)) {
+        } else if (this.getVariant() == MushroomCow.MushroomType.BROWN && itemstack.is(ItemTags.SMALL_FLOWERS)) {
             if (this.stewEffects != null) {
-                for (int i = 0; i < 2; i++) {
-                    this.level()
-                        .addParticle(
-                            ParticleTypes.SMOKE,
-                            this.getX() + this.random.nextDouble() / 2.0,
-                            this.getY(0.5),
-                            this.getZ() + this.random.nextDouble() / 2.0,
-                            0.0,
-                            this.random.nextDouble() / 5.0,
-                            0.0
-                        );
+                for (int i = 0; i < 2; ++i) {
+                    this.level().addParticle(ParticleTypes.SMOKE, this.getX() + this.random.nextDouble() / 2.0D, this.getY(0.5D), this.getZ() + this.random.nextDouble() / 2.0D, 0.0D, this.random.nextDouble() / 5.0D, 0.0D);
                 }
             } else {
-                Optional<List<SuspiciousEffectHolder.EffectEntry>> effectsFromItemStack = this.getEffectsFromItemStack(itemInHand);
-                if (effectsFromItemStack.isEmpty()) {
+                Optional<List<SuspiciousEffectHolder.a>> optional = this.getEffectsFromItemStack(itemstack);
+
+                if (optional.isEmpty()) {
                     return InteractionResult.PASS;
                 }
 
                 if (!player.getAbilities().instabuild) {
-                    itemInHand.shrink(1);
+                    itemstack.shrink(1);
                 }
 
-                for (int i1 = 0; i1 < 4; i1++) {
-                    this.level()
-                        .addParticle(
-                            ParticleTypes.EFFECT,
-                            this.getX() + this.random.nextDouble() / 2.0,
-                            this.getY(0.5),
-                            this.getZ() + this.random.nextDouble() / 2.0,
-                            0.0,
-                            this.random.nextDouble() / 5.0,
-                            0.0
-                        );
+                for (int j = 0; j < 4; ++j) {
+                    this.level().addParticle(ParticleTypes.EFFECT, this.getX() + this.random.nextDouble() / 2.0D, this.getY(0.5D), this.getZ() + this.random.nextDouble() / 2.0D, 0.0D, this.random.nextDouble() / 5.0D, 0.0D);
                 }
 
-                this.stewEffects = effectsFromItemStack.get();
+                this.stewEffects = (List) optional.get();
                 this.playSound(SoundEvents.MOOSHROOM_EAT, 2.0F, 1.0F);
             }
 
@@ -165,35 +167,49 @@
 
     @Override
     public void shear(SoundSource category) {
-        this.level().playSound(null, this, SoundEvents.MOOSHROOM_SHEAR, category, 1.0F, 1.0F);
+        this.level().playSound((Player) null, (Entity) this, SoundEvents.MOOSHROOM_SHEAR, category, 1.0F, 1.0F);
         if (!this.level().isClientSide()) {
-            Cow cow = EntityType.COW.create(this.level());
-            if (cow != null) {
-                ((ServerLevel)this.level()).sendParticles(ParticleTypes.EXPLOSION, this.getX(), this.getY(0.5), this.getZ(), 1, 0.0, 0.0, 0.0, 0.0);
-                this.discard();
-                cow.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
-                cow.setHealth(this.getHealth());
-                cow.yBodyRot = this.yBodyRot;
+            Cow entitycow = (Cow) EntityType.COW.create(this.level());
+
+            if (entitycow != null) {
+                ((ServerLevel) this.level()).sendParticles(ParticleTypes.EXPLOSION, this.getX(), this.getY(0.5D), this.getZ(), 1, 0.0D, 0.0D, 0.0D, 0.0D);
+                // this.discard(); // CraftBukkit - moved down
+                entitycow.moveTo(this.getX(), this.getY(), this.getZ(), this.getYRot(), this.getXRot());
+                entitycow.setHealth(this.getHealth());
+                entitycow.yBodyRot = this.yBodyRot;
                 if (this.hasCustomName()) {
-                    cow.setCustomName(this.getCustomName());
-                    cow.setCustomNameVisible(this.isCustomNameVisible());
+                    entitycow.setCustomName(this.getCustomName());
+                    entitycow.setCustomNameVisible(this.isCustomNameVisible());
                 }
 
                 if (this.isPersistenceRequired()) {
-                    cow.setPersistenceRequired();
+                    entitycow.setPersistenceRequired();
                 }
 
-                cow.setInvulnerable(this.isInvulnerable());
-                this.level().addFreshEntity(cow);
+                entitycow.setInvulnerable(this.isInvulnerable());
+                // CraftBukkit start
+                if (CraftEventFactory.callEntityTransformEvent(this, entitycow, EntityTransformEvent.TransformReason.SHEARED).isCancelled()) {
+                    return;
+                }
+                this.level().addFreshEntity(entitycow, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SHEARED);
 
-                for (int i = 0; i < 5; i++) {
-                    this.level()
-                        .addFreshEntity(
-                            new ItemEntity(this.level(), this.getX(), this.getY(1.0), this.getZ(), new ItemStack(this.getVariant().blockState.getBlock()))
-                        );
+                this.discard(); // CraftBukkit - from above
+                // CraftBukkit end
+
+                for (int i = 0; i < 5; ++i) {
+                    // CraftBukkit start
+                    ItemEntity entityitem = new ItemEntity(this.level(), this.getX(), this.getY(1.0D), this.getZ(), new ItemStack(this.getVariant().blockState.getBlock()));
+                    EntityDropItemEvent event = new EntityDropItemEvent(this.getBukkitEntity(), (org.bukkit.entity.Item) entityitem.getBukkitEntity());
+                    Bukkit.getPluginManager().callEvent(event);
+                    if (event.isCancelled()) {
+                        continue;
+                    }
+                    this.level().addFreshEntity(entityitem);
+                    // CraftBukkit end
                 }
             }
         }
+
     }
 
     @Override
@@ -206,11 +222,11 @@
         super.addAdditionalSaveData(compound);
         compound.putString("Type", this.getVariant().getSerializedName());
         if (this.stewEffects != null) {
-            SuspiciousEffectHolder.EffectEntry.LIST_CODEC
-                .encodeStart(NbtOps.INSTANCE, this.stewEffects)
-                .result()
-                .ifPresent(tag -> compound.put("stew_effects", tag));
+            SuspiciousEffectHolder.a.LIST_CODEC.encodeStart(NbtOps.INSTANCE, this.stewEffects).result().ifPresent((nbtbase) -> {
+                compound.put("stew_effects", nbtbase);
+            });
         }
+
     }
 
     @Override
@@ -218,66 +234,68 @@
         super.readAdditionalSaveData(compound);
         this.setVariant(MushroomCow.MushroomType.byType(compound.getString("Type")));
         if (compound.contains("stew_effects", 9)) {
-            SuspiciousEffectHolder.EffectEntry.LIST_CODEC
-                .parse(NbtOps.INSTANCE, compound.get("stew_effects"))
-                .result()
-                .ifPresent(list -> this.stewEffects = (List<SuspiciousEffectHolder.EffectEntry>)list);
+            SuspiciousEffectHolder.a.LIST_CODEC.parse(NbtOps.INSTANCE, compound.get("stew_effects")).result().ifPresent((list) -> {
+                this.stewEffects = list;
+            });
         }
+
     }
 
-    private Optional<List<SuspiciousEffectHolder.EffectEntry>> getEffectsFromItemStack(ItemStack itemStack) {
-        SuspiciousEffectHolder suspiciousEffectHolder = SuspiciousEffectHolder.tryGet(itemStack.getItem());
-        return suspiciousEffectHolder != null ? Optional.of(suspiciousEffectHolder.getSuspiciousEffects()) : Optional.empty();
+    private Optional<List<SuspiciousEffectHolder.a>> getEffectsFromItemStack(ItemStack itemstack) {
+        SuspiciousEffectHolder suspiciouseffectholder = SuspiciousEffectHolder.tryGet(itemstack.getItem());
+
+        return suspiciouseffectholder != null ? Optional.of(suspiciouseffectholder.getSuspiciousEffects()) : Optional.empty();
     }
 
-    @Override
     public void setVariant(MushroomCow.MushroomType variant) {
-        this.entityData.set(DATA_TYPE, variant.type);
+        this.entityData.set(MushroomCow.DATA_TYPE, variant.type);
     }
 
     @Override
     public MushroomCow.MushroomType getVariant() {
-        return MushroomCow.MushroomType.byType(this.entityData.get(DATA_TYPE));
+        return MushroomCow.MushroomType.byType((String) this.entityData.get(MushroomCow.DATA_TYPE));
     }
 
     @Nullable
     @Override
     public MushroomCow getBreedOffspring(ServerLevel level, AgeableMob otherParent) {
-        MushroomCow mushroomCow = EntityType.MOOSHROOM.create(level);
-        if (mushroomCow != null) {
-            mushroomCow.setVariant(this.getOffspringType((MushroomCow)otherParent));
+        MushroomCow entitymushroomcow = (MushroomCow) EntityType.MOOSHROOM.create(level);
+
+        if (entitymushroomcow != null) {
+            entitymushroomcow.setVariant(this.getOffspringType((MushroomCow) otherParent));
         }
 
-        return mushroomCow;
+        return entitymushroomcow;
     }
 
     private MushroomCow.MushroomType getOffspringType(MushroomCow mate) {
-        MushroomCow.MushroomType variant = this.getVariant();
-        MushroomCow.MushroomType variant1 = mate.getVariant();
-        MushroomCow.MushroomType mushroomType;
-        if (variant == variant1 && this.random.nextInt(1024) == 0) {
-            mushroomType = variant == MushroomCow.MushroomType.BROWN ? MushroomCow.MushroomType.RED : MushroomCow.MushroomType.BROWN;
+        MushroomCow.MushroomType entitymushroomcow_type = this.getVariant();
+        MushroomCow.MushroomType entitymushroomcow_type1 = mate.getVariant();
+        MushroomCow.MushroomType entitymushroomcow_type2;
+
+        if (entitymushroomcow_type == entitymushroomcow_type1 && this.random.nextInt(1024) == 0) {
+            entitymushroomcow_type2 = entitymushroomcow_type == MushroomCow.MushroomType.BROWN ? MushroomCow.MushroomType.RED : MushroomCow.MushroomType.BROWN;
         } else {
-            mushroomType = this.random.nextBoolean() ? variant : variant1;
+            entitymushroomcow_type2 = this.random.nextBoolean() ? entitymushroomcow_type : entitymushroomcow_type1;
         }
 
-        return mushroomType;
+        return entitymushroomcow_type2;
     }
 
     public static enum MushroomType implements StringRepresentable {
-        RED("red", Blocks.RED_MUSHROOM.defaultBlockState()),
-        BROWN("brown", Blocks.BROWN_MUSHROOM.defaultBlockState());
 
+        RED("red", Blocks.RED_MUSHROOM.defaultBlockState()), BROWN("brown", Blocks.BROWN_MUSHROOM.defaultBlockState());
+
         public static final StringRepresentable.EnumCodec<MushroomCow.MushroomType> CODEC = StringRepresentable.fromEnum(MushroomCow.MushroomType::values);
         final String type;
-        final BlockState blockState;
+        final IBlockData blockState;
 
-        private MushroomType(String type, BlockState blockState) {
-            this.type = type;
-            this.blockState = blockState;
+        private MushroomType(String s, IBlockData iblockdata) {
+            this.type = s;
+            this.blockState = iblockdata;
         }
 
-        public BlockState getBlockState() {
+        public IBlockData getBlockState() {
             return this.blockState;
         }
 
@@ -287,7 +305,7 @@
         }
 
         static MushroomCow.MushroomType byType(String name) {
-            return CODEC.byName(name, RED);
+            return (MushroomCow.MushroomType) MushroomCow.MushroomType.CODEC.byName(name, MushroomCow.MushroomType.RED);
         }
     }
 }
