--- a/net/minecraft/world/entity/ai/behavior/AssignProfessionFromJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/AssignProfessionFromJobSite.java
@@ -9,36 +9,49 @@
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.entity.npc.VillagerProfession;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftVillager;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.VillagerCareerChangeEvent;
+// CraftBukkit end
+
 public class AssignProfessionFromJobSite {
 
     public AssignProfessionFromJobSite() {}
 
     public static BehaviorControl<Villager> create() {
-        return BehaviorBuilder.create((behaviorbuilder_instance) -> {
-            return behaviorbuilder_instance.group(behaviorbuilder_instance.present(MemoryModuleType.POTENTIAL_JOB_SITE), behaviorbuilder_instance.registered(MemoryModuleType.JOB_SITE)).apply(behaviorbuilder_instance, (memoryaccessor, memoryaccessor1) -> {
-                return (serverlevel, villager, i) -> {
-                    GlobalPos globalpos = (GlobalPos) behaviorbuilder_instance.get(memoryaccessor);
+        return BehaviorBuilder.create((behaviorbuilder_b) -> {
+            return behaviorbuilder_b.group(behaviorbuilder_b.present(MemoryModuleType.POTENTIAL_JOB_SITE), behaviorbuilder_b.registered(MemoryModuleType.JOB_SITE)).apply(behaviorbuilder_b, (memoryaccessor, memoryaccessor1) -> {
+                return (worldserver, entityvillager, i) -> {
+                    GlobalPos globalpos = (GlobalPos) behaviorbuilder_b.get(memoryaccessor);
 
-                    if (!globalpos.pos().closerToCenterThan(villager.position(), 2.0D) && !villager.assignProfessionWhenSpawned()) {
+                    if (!globalpos.pos().closerToCenterThan(entityvillager.position(), 2.0D) && !entityvillager.assignProfessionWhenSpawned()) {
                         return false;
                     } else {
                         memoryaccessor.erase();
                         memoryaccessor1.set(globalpos);
-                        serverlevel.broadcastEntityEvent(villager, (byte) 14);
-                        if (villager.getVillagerData().getProfession() != VillagerProfession.NONE) {
+                        worldserver.broadcastEntityEvent(entityvillager, (byte) 14);
+                        if (entityvillager.getVillagerData().getProfession() != VillagerProfession.NONE) {
                             return true;
                         } else {
-                            MinecraftServer minecraftserver = serverlevel.getServer();
+                            MinecraftServer minecraftserver = worldserver.getServer();
 
-                            Optional.ofNullable(minecraftserver.getLevel(globalpos.dimension())).flatMap((serverlevel1) -> {
-                                return serverlevel1.getPoiManager().getType(globalpos.pos());
+                            Optional.ofNullable(minecraftserver.getLevel(globalpos.dimension())).flatMap((worldserver1) -> {
+                                return worldserver1.getPoiManager().getType(globalpos.pos());
                             }).flatMap((holder) -> {
                                 return BuiltInRegistries.VILLAGER_PROFESSION.stream().filter((villagerprofession) -> {
                                     return villagerprofession.heldJobSite().test(holder);
                                 }).findFirst();
                             }).ifPresent((villagerprofession) -> {
-                                villager.setVillagerData(villager.getVillagerData().setProfession(villagerprofession));
-                                villager.refreshBrain(serverlevel);
+                                // CraftBukkit start - Fire VillagerCareerChangeEvent where Villager gets employed
+                                VillagerCareerChangeEvent event = CraftEventFactory.callVillagerCareerChangeEvent(entityvillager, CraftVillager.CraftProfession.minecraftToBukkit(villagerprofession), VillagerCareerChangeEvent.ChangeReason.EMPLOYED);
+                                if (event.isCancelled()) {
+                                    return;
+                                }
+
+                                entityvillager.setVillagerData(entityvillager.getVillagerData().setProfession(CraftVillager.CraftProfession.bukkitToMinecraft(event.getProfession())));
+                                // CraftBukkit end
+                                entityvillager.refreshBrain(worldserver);
                             });
                             return true;
                         }
