--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -39,6 +40,13 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
+// CraftBukkit start
+import com.google.common.collect.ImmutableList;
+import java.util.AbstractList;
+import org.bukkit.craftbukkit.block.CraftBlock;
+import org.bukkit.event.block.BlockPistonRetractEvent;
+import org.bukkit.event.block.BlockPistonExtendEvent;
+// CraftBukkit end
 
 public class PistonBaseBlock extends DirectionalBlock {
     public static final MapCodec<PistonBaseBlock> CODEC = RecordCodecBuilder.mapCodec(
@@ -143,7 +158,19 @@
                 i = 2;
             }
 
-            level.blockEvent(pos, this, i, direction.get3DDataValue());
+            // CraftBukkit start
+            if (!this.isSticky) {
+                org.bukkit.block.Block block = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
+                BlockPistonRetractEvent event = new BlockPistonRetractEvent(block, ImmutableList.<org.bukkit.block.Block>of(), CraftBlock.notchToBlockFace(enumdirection));
+                level.getCraftServer().getPluginManager().callEvent(event);
+
+                if (event.isCancelled()) {
+                    return;
+                }
+            }
+            // PAIL: checkME - what happened to setTypeAndData?
+            // CraftBukkit end
+            level.blockEvent(pos, this, b0, enumdirection.get3DDataValue());
         }
     }
 
@@ -303,6 +350,8 @@
             BlockState[] blockStates = new BlockState[toPush.size() + toDestroy.size()];
             Direction direction = extending ? facing : facing.getOpposite();
             int i = 0;
+            // CraftBukkit start
+            final org.bukkit.block.Block bblock = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
 
             for (int i1 = toDestroy.size() - 1; i1 >= 0; i1--) {
                 BlockPos blockPos2 = toDestroy.get(i1);
@@ -328,6 +391,7 @@
                 level.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(blockPos2, blockState2, list.get(i1), facing, extending, false));
                 blockStates[i++] = blockState1;
             }
+            // CraftBukkit end
 
             if (extending) {
                 PistonType pistonType = this.isSticky ? PistonType.STICKY : PistonType.DEFAULT;
