--- a/net/minecraft/world/inventory/BrewingStandMenu.java
+++ b/net/minecraft/world/inventory/BrewingStandMenu.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.inventory;
 
 import net.minecraft.advancements.CriteriaTriggers;
+import net.minecraft.core.Holder;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.world.Container;
 import net.minecraft.world.SimpleContainer;
@@ -11,8 +12,13 @@
 import net.minecraft.world.item.alchemy.Potion;
 import net.minecraft.world.item.alchemy.PotionBrewing;
 import net.minecraft.world.item.alchemy.PotionUtils;
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftInventoryBrewer;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+// CraftBukkit end
 
 public class BrewingStandMenu extends AbstractContainerMenu {
+
     private static final int BOTTLE_SLOT_START = 0;
     private static final int BOTTLE_SLOT_END = 2;
     private static final int INGREDIENT_SLOT = 3;
@@ -27,12 +33,18 @@
     private final ContainerData brewingStandData;
     private final Slot ingredientSlot;
 
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private Inventory player;
+    // CraftBukkit end
+
     public BrewingStandMenu(int containerId, Inventory playerInventory) {
         this(containerId, playerInventory, new SimpleContainer(5), new SimpleContainerData(2));
     }
 
     public BrewingStandMenu(int containerId, Inventory playerInventory, Container brewingStandContainer, ContainerData brewingStandData) {
         super(MenuType.BREWING_STAND, containerId);
+        player = playerInventory; // CraftBukkit
         checkContainerSize(brewingStandContainer, 5);
         checkContainerDataCount(brewingStandData, 2);
         this.brewingStand = brewingStandContainer;
@@ -44,75 +56,81 @@
         this.addSlot(new BrewingStandMenu.FuelSlot(brewingStandContainer, 4, 17, 17));
         this.addDataSlots(brewingStandData);
 
-        for (int i = 0; i < 3; i++) {
-            for (int i1 = 0; i1 < 9; i1++) {
-                this.addSlot(new Slot(playerInventory, i1 + i * 9 + 9, 8 + i1 * 18, 84 + i * 18));
+        int j;
+
+        for (j = 0; j < 3; ++j) {
+            for (int k = 0; k < 9; ++k) {
+                this.addSlot(new Slot(playerInventory, k + j * 9 + 9, 8 + k * 18, 84 + j * 18));
             }
         }
 
-        for (int i = 0; i < 9; i++) {
-            this.addSlot(new Slot(playerInventory, i, 8 + i * 18, 142));
+        for (j = 0; j < 9; ++j) {
+            this.addSlot(new Slot(playerInventory, j, 8 + j * 18, 142));
         }
+
     }
 
     @Override
     public boolean stillValid(Player player) {
+        if (!this.checkReachable) return true; // CraftBukkit
         return this.brewingStand.stillValid(player);
     }
 
     @Override
     public ItemStack quickMoveStack(Player player, int index) {
-        ItemStack itemStack = ItemStack.EMPTY;
-        Slot slot = this.slots.get(index);
+        ItemStack itemstack = ItemStack.EMPTY;
+        Slot slot = (Slot) this.slots.get(index);
+
         if (slot != null && slot.hasItem()) {
-            ItemStack item = slot.getItem();
-            itemStack = item.copy();
+            ItemStack itemstack1 = slot.getItem();
+
+            itemstack = itemstack1.copy();
             if ((index < 0 || index > 2) && index != 3 && index != 4) {
-                if (BrewingStandMenu.FuelSlot.mayPlaceItem(itemStack)) {
-                    if (this.moveItemStackTo(item, 4, 5, false) || this.ingredientSlot.mayPlace(item) && !this.moveItemStackTo(item, 3, 4, false)) {
+                if (BrewingStandMenu.FuelSlot.mayPlaceItem(itemstack)) {
+                    if (this.moveItemStackTo(itemstack1, 4, 5, false) || this.ingredientSlot.mayPlace(itemstack1) && !this.moveItemStackTo(itemstack1, 3, 4, false)) {
                         return ItemStack.EMPTY;
                     }
-                } else if (this.ingredientSlot.mayPlace(item)) {
-                    if (!this.moveItemStackTo(item, 3, 4, false)) {
+                } else if (this.ingredientSlot.mayPlace(itemstack1)) {
+                    if (!this.moveItemStackTo(itemstack1, 3, 4, false)) {
                         return ItemStack.EMPTY;
                     }
-                } else if (BrewingStandMenu.PotionSlot.mayPlaceItem(itemStack) && itemStack.getCount() == 1) {
-                    if (!this.moveItemStackTo(item, 0, 3, false)) {
+                } else if (BrewingStandMenu.PotionSlot.mayPlaceItem(itemstack) && itemstack.getCount() == 1) {
+                    if (!this.moveItemStackTo(itemstack1, 0, 3, false)) {
                         return ItemStack.EMPTY;
                     }
                 } else if (index >= 5 && index < 32) {
-                    if (!this.moveItemStackTo(item, 32, 41, false)) {
+                    if (!this.moveItemStackTo(itemstack1, 32, 41, false)) {
                         return ItemStack.EMPTY;
                     }
                 } else if (index >= 32 && index < 41) {
-                    if (!this.moveItemStackTo(item, 5, 32, false)) {
+                    if (!this.moveItemStackTo(itemstack1, 5, 32, false)) {
                         return ItemStack.EMPTY;
                     }
-                } else if (!this.moveItemStackTo(item, 5, 41, false)) {
+                } else if (!this.moveItemStackTo(itemstack1, 5, 41, false)) {
                     return ItemStack.EMPTY;
                 }
             } else {
-                if (!this.moveItemStackTo(item, 5, 41, true)) {
+                if (!this.moveItemStackTo(itemstack1, 5, 41, true)) {
                     return ItemStack.EMPTY;
                 }
 
-                slot.onQuickCraft(item, itemStack);
+                slot.onQuickCraft(itemstack1, itemstack);
             }
 
-            if (item.isEmpty()) {
+            if (itemstack1.isEmpty()) {
                 slot.setByPlayer(ItemStack.EMPTY);
             } else {
                 slot.setChanged();
             }
 
-            if (item.getCount() == itemStack.getCount()) {
+            if (itemstack1.getCount() == itemstack.getCount()) {
                 return ItemStack.EMPTY;
             }
 
-            slot.onTake(player, item);
+            slot.onTake(player, itemstack1);
         }
 
-        return itemStack;
+        return itemstack;
     }
 
     public int getFuel() {
@@ -123,8 +141,9 @@
         return this.brewingStandData.get(0);
     }
 
-    static class FuelSlot extends Slot {
-        public FuelSlot(Container container, int slot, int x, int y) {
+    private static class PotionSlot extends Slot {
+
+        public PotionSlot(Container container, int slot, int x, int y) {
             super(container, slot, x, y);
         }
 
@@ -133,17 +152,29 @@
             return mayPlaceItem(stack);
         }
 
-        public static boolean mayPlaceItem(ItemStack itemStack) {
-            return itemStack.is(Items.BLAZE_POWDER);
+        @Override
+        public int getMaxStackSize() {
+            return 1;
         }
 
         @Override
-        public int getMaxStackSize() {
-            return 64;
+        public void onTake(Player player, ItemStack stack) {
+            Potion potionregistry = PotionUtils.getPotion(stack);
+
+            if (player instanceof ServerPlayer) {
+                CriteriaTriggers.BREWED_POTION.trigger((ServerPlayer) player, (Holder) potionregistry.builtInRegistryHolder());
+            }
+
+            super.onTake(player, stack);
         }
+
+        public static boolean mayPlaceItem(ItemStack stack) {
+            return stack.is(Items.POTION) || stack.is(Items.SPLASH_POTION) || stack.is(Items.LINGERING_POTION) || stack.is(Items.GLASS_BOTTLE);
+        }
     }
 
-    static class IngredientsSlot extends Slot {
+    private static class IngredientsSlot extends Slot {
+
         public IngredientsSlot(Container container, int slot, int x, int y) {
             super(container, slot, x, y);
         }
@@ -159,8 +190,9 @@
         }
     }
 
-    static class PotionSlot extends Slot {
-        public PotionSlot(Container container, int slot, int x, int y) {
+    private static class FuelSlot extends Slot {
+
+        public FuelSlot(Container container, int slot, int x, int y) {
             super(container, slot, x, y);
         }
 
@@ -169,23 +201,26 @@
             return mayPlaceItem(stack);
         }
 
-        @Override
-        public int getMaxStackSize() {
-            return 1;
+        public static boolean mayPlaceItem(ItemStack itemStack) {
+            return itemStack.is(Items.BLAZE_POWDER);
         }
 
         @Override
-        public void onTake(Player player, ItemStack stack) {
-            Potion potion = PotionUtils.getPotion(stack);
-            if (player instanceof ServerPlayer) {
-                CriteriaTriggers.BREWED_POTION.trigger((ServerPlayer)player, potion.builtInRegistryHolder());
-            }
-
-            super.onTake(player, stack);
+        public int getMaxStackSize() {
+            return 64;
         }
+    }
 
-        public static boolean mayPlaceItem(ItemStack stack) {
-            return stack.is(Items.POTION) || stack.is(Items.SPLASH_POTION) || stack.is(Items.LINGERING_POTION) || stack.is(Items.GLASS_BOTTLE);
+    // CraftBukkit start
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
         }
+
+        CraftInventoryBrewer inventory = new CraftInventoryBrewer(this.brewingStand);
+        bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+        return bukkitEntity;
     }
+    // CraftBukkit end
 }
