--- a/net/minecraft/world/inventory/MenuType.java
+++ b/net/minecraft/world/inventory/MenuType.java
@@ -9,6 +9,7 @@
 import net.minecraft.world.flag.FeatureFlags;
 
 public class MenuType<T extends AbstractContainerMenu> implements FeatureElement {
+
     public static final MenuType<ChestMenu> GENERIC_9x1 = register("generic_9x1", ChestMenu::oneRow);
     public static final MenuType<ChestMenu> GENERIC_9x2 = register("generic_9x2", ChestMenu::twoRows);
     public static final MenuType<ChestMenu> GENERIC_9x3 = register("generic_9x3", ChestMenu::threeRows);
@@ -26,7 +27,9 @@
     public static final MenuType<FurnaceMenu> FURNACE = register("furnace", FurnaceMenu::new);
     public static final MenuType<GrindstoneMenu> GRINDSTONE = register("grindstone", GrindstoneMenu::new);
     public static final MenuType<HopperMenu> HOPPER = register("hopper", HopperMenu::new);
-    public static final MenuType<LecternMenu> LECTERN = register("lectern", (containerId, playerInventory) -> new LecternMenu(containerId));
+    public static final MenuType<LecternMenu> LECTERN = register("lectern", (i, playerinventory) -> {
+        return new LecternMenu(i, playerinventory); // CraftBukkit
+    });
     public static final MenuType<LoomMenu> LOOM = register("loom", LoomMenu::new);
     public static final MenuType<MerchantMenu> MERCHANT = register("merchant", MerchantMenu::new);
     public static final MenuType<ShulkerBoxMenu> SHULKER_BOX = register("shulker_box", ShulkerBoxMenu::new);
@@ -38,11 +41,11 @@
     private final MenuType.MenuSupplier<T> constructor;
 
     private static <T extends AbstractContainerMenu> MenuType<T> register(String key, MenuType.MenuSupplier<T> factory) {
-        return Registry.register(BuiltInRegistries.MENU, key, new MenuType<>(factory, FeatureFlags.VANILLA_SET));
+        return (MenuType) Registry.register(BuiltInRegistries.MENU, key, new MenuType<>(factory, FeatureFlags.VANILLA_SET));
     }
 
     private static <T extends AbstractContainerMenu> MenuType<T> register(String key, MenuType.MenuSupplier<T> factory, FeatureFlag... requiredFeatures) {
-        return Registry.register(BuiltInRegistries.MENU, key, new MenuType<>(factory, FeatureFlags.REGISTRY.subset(requiredFeatures)));
+        return (MenuType) Registry.register(BuiltInRegistries.MENU, key, new MenuType<>(factory, FeatureFlags.REGISTRY.subset(requiredFeatures)));
     }
 
     private MenuType(MenuType.MenuSupplier<T> constructor, FeatureFlagSet requiredFeatures) {
@@ -59,7 +62,8 @@
         return this.requiredFeatures;
     }
 
-    interface MenuSupplier<T extends AbstractContainerMenu> {
+    private interface MenuSupplier<T extends AbstractContainerMenu> {
+
         T create(int containerId, Inventory playerInventory);
     }
 }
