--- a/net/minecraft/world/inventory/PlayerEnderChestContainer.java
+++ b/net/minecraft/world/inventory/PlayerEnderChestContainer.java
@@ -7,13 +7,32 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.block.entity.EnderChestBlockEntity;
+// CraftBukkit start
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.util.CraftLocation;
+import org.bukkit.inventory.InventoryHolder;
+// CraftBukkit end
 
 public class PlayerEnderChestContainer extends SimpleContainer {
+
     @Nullable
     private EnderChestBlockEntity activeChest;
+    // CraftBukkit start
+    private final Player owner;
 
-    public PlayerEnderChestContainer() {
+    public InventoryHolder getBukkitOwner() {
+        return owner.getBukkitEntity();
+    }
+
+    @Override
+    public Location getLocation() {
+        return this.activeChest != null ? CraftLocation.toBukkit(this.activeChest.getBlockPos(), this.activeChest.getLevel().getWorld()) : null;
+    }
+
+    public PlayerEnderChestContainer(Player owner) {
         super(27);
+        this.owner = owner;
+        // CraftBukkit end
     }
 
     public void setActiveChest(EnderChestBlockEntity enderChestBlockEntity) {
@@ -26,39 +45,45 @@
 
     @Override
     public void fromTag(ListTag containerNbt) {
-        for (int i = 0; i < this.getContainerSize(); i++) {
+        int i;
+
+        for (i = 0; i < this.getContainerSize(); ++i) {
             this.setItem(i, ItemStack.EMPTY);
         }
 
-        for (int i = 0; i < containerNbt.size(); i++) {
-            CompoundTag compound = containerNbt.getCompound(i);
-            int i1 = compound.getByte("Slot") & 255;
-            if (i1 >= 0 && i1 < this.getContainerSize()) {
-                this.setItem(i1, ItemStack.of(compound));
+        for (i = 0; i < containerNbt.size(); ++i) {
+            CompoundTag nbttagcompound = containerNbt.getCompound(i);
+            int j = nbttagcompound.getByte("Slot") & 255;
+
+            if (j >= 0 && j < this.getContainerSize()) {
+                this.setItem(j, ItemStack.of(nbttagcompound));
             }
         }
+
     }
 
     @Override
     public ListTag createTag() {
-        ListTag list = new ListTag();
+        ListTag nbttaglist = new ListTag();
 
-        for (int i = 0; i < this.getContainerSize(); i++) {
-            ItemStack item = this.getItem(i);
-            if (!item.isEmpty()) {
-                CompoundTag compoundTag = new CompoundTag();
-                compoundTag.putByte("Slot", (byte)i);
-                item.save(compoundTag);
-                list.add(compoundTag);
+        for (int i = 0; i < this.getContainerSize(); ++i) {
+            ItemStack itemstack = this.getItem(i);
+
+            if (!itemstack.isEmpty()) {
+                CompoundTag nbttagcompound = new CompoundTag();
+
+                nbttagcompound.putByte("Slot", (byte) i);
+                itemstack.save(nbttagcompound);
+                nbttaglist.add(nbttagcompound);
             }
         }
 
-        return list;
+        return nbttaglist;
     }
 
     @Override
     public boolean stillValid(Player player) {
-        return (this.activeChest == null || this.activeChest.stillValid(player)) && super.stillValid(player);
+        return this.activeChest != null && !this.activeChest.stillValid(player) ? false : super.stillValid(player);
     }
 
     @Override
