--- a/net/minecraft/world/item/BoatItem.java
+++ b/net/minecraft/world/item/BoatItem.java
@@ -51,16 +59,32 @@
                 }
             }
 
-            if (playerPOVHitResult.getType() == HitResult.Type.BLOCK) {
-                Boat boat = this.getBoat(level, playerPOVHitResult, itemInHand, player);
-                boat.setVariant(this.type);
-                boat.setYRot(player.getYRot());
-                if (!level.noCollision(boat, boat.getBoundingBox())) {
-                    return InteractionResultHolder.fail(itemInHand);
+            if (movingobjectpositionblock.getType() == HitResult.EnumMovingObjectType.BLOCK) {
+                // CraftBukkit start - Boat placement
+                org.bukkit.event.player.PlayerInteractEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerInteractEvent(player, org.bukkit.event.block.Action.RIGHT_CLICK_BLOCK, movingobjectpositionblock.getBlockPos(), movingobjectpositionblock.getDirection(), itemstack, false, hand, movingobjectpositionblock.getLocation());
+
+                if (event.isCancelled()) {
+                    return InteractionResultHolder.pass(itemstack);
+                }
+                // CraftBukkit end
+                Boat entityboat = this.getBoat(level, movingobjectpositionblock, itemstack, player);
+
+                entityboat.setVariant(this.type);
+                entityboat.setYRot(player.getYRot());
+                if (!level.noCollision(entityboat, entityboat.getBoundingBox())) {
+                    return InteractionResultHolder.fail(itemstack);
                 } else {
                     if (!level.isClientSide) {
-                        level.addFreshEntity(boat);
-                        level.gameEvent(player, GameEvent.ENTITY_PLACE, playerPOVHitResult.getLocation());
+                        // CraftBukkit start
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPlaceEvent(level, movingobjectpositionblock.getBlockPos(), movingobjectpositionblock.getDirection(), player, entityboat, hand).isCancelled()) {
+                            return InteractionResultHolder.fail(itemstack);
+                        }
+
+                        if (!level.addFreshEntity(entityboat)) {
+                            return InteractionResultHolder.pass(itemstack);
+                        }
+                        // CraftBukkit end
+                        level.gameEvent((Entity) player, GameEvent.ENTITY_PLACE, movingobjectpositionblock.getLocation());
                         if (!player.getAbilities().instabuild) {
                             itemInHand.shrink(1);
                         }
