--- a/net/minecraft/world/item/DyeItem.java
+++ b/net/minecraft/world/item/DyeItem.java
@@ -11,7 +12,7 @@
 import net.minecraft.world.entity.player.Player;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.SignBlockEntity;
-import net.minecraft.world.level.block.entity.SignText;
+import org.bukkit.event.entity.SheepDyeWoolEvent; // CraftBukkit
 
 public class DyeItem extends Item implements SignApplicator {
     private static final Map<DyeColor, DyeItem> ITEM_BY_COLOR = Maps.newEnumMap(DyeColor.class);
@@ -32,7 +30,25 @@
                 stack.shrink(1);
             }
 
-            return InteractionResult.sidedSuccess(player.level().isClientSide);
+            if (entitysheep.isAlive() && !entitysheep.isSheared() && entitysheep.getColor() != this.dyeColor) {
+                entitysheep.level().playSound(player, (Entity) entitysheep, SoundEvents.DYE_USE, SoundSource.PLAYERS, 1.0F, 1.0F);
+                if (!player.level().isClientSide) {
+                    // CraftBukkit start
+                    byte bColor = (byte) this.dyeColor.getId();
+                    SheepDyeWoolEvent event = new SheepDyeWoolEvent((org.bukkit.entity.Sheep) entitysheep.getBukkitEntity(), org.bukkit.DyeColor.getByWoolData(bColor), (org.bukkit.entity.Player) player.getBukkitEntity());
+                    entitysheep.level().getCraftServer().getPluginManager().callEvent(event);
+
+                    if (event.isCancelled()) {
+                        return InteractionResult.PASS;
+                    }
+
+                    entitysheep.setColor(DyeColor.byId((byte) event.getColor().getWoolData()));
+                    // CraftBukkit end
+                    stack.shrink(1);
+                }
+
+                return InteractionResult.sidedSuccess(player.level().isClientSide);
+            }
         }
 
         return InteractionResult.PASS;
