--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -64,8 +67,9 @@
 
     @Nullable
     public static Integer getMapId(ItemStack stack) {
-        CompoundTag tag = stack.getTag();
-        return tag != null && tag.contains("map", 99) ? tag.getInt("map") : null;
+        CompoundTag nbttagcompound = stack.getTag();
+
+        return nbttagcompound != null && nbttagcompound.contains("map", 99) ? nbttagcompound.getInt("map") : -1; // CraftBukkit - make new maps for no tag
     }
 
     private static int createNewSavedData(
