--- a/net/minecraft/world/item/MilkBucketItem.java
+++ b/net/minecraft/world/item/MilkBucketItem.java
@@ -28,7 +31,7 @@
         }
 
         if (!level.isClientSide) {
-            entityLiving.removeAllEffects();
+            entityLiving.removeAllEffects(org.bukkit.event.entity.EntityPotionEffectEvent.Cause.MILK); // CraftBukkit
         }
 
         return stack.isEmpty() ? new ItemStack(Items.BUCKET) : stack;
