--- a/net/minecraft/world/item/RecordItem.java
+++ b/net/minecraft/world/item/RecordItem.java
@@ -36,16 +38,23 @@
 
     @Override
     public InteractionResult useOn(UseOnContext context) {
-        Level level = context.getLevel();
-        BlockPos clickedPos = context.getClickedPos();
-        BlockState blockState = level.getBlockState(clickedPos);
-        if (blockState.is(Blocks.JUKEBOX) && !blockState.getValue(JukeboxBlock.HAS_RECORD)) {
-            ItemStack itemInHand = context.getItemInHand();
-            if (!level.isClientSide) {
-                Player player = context.getPlayer();
-                if (level.getBlockEntity(clickedPos) instanceof JukeboxBlockEntity jukeboxBlockEntity) {
-                    jukeboxBlockEntity.setTheItem(itemInHand.copy());
-                    level.gameEvent(GameEvent.BLOCK_CHANGE, clickedPos, GameEvent.Context.of(player, blockState));
+        Level world = context.getLevel();
+        BlockPos blockposition = context.getClickedPos();
+        IBlockData iblockdata = world.getBlockState(blockposition);
+
+        if (iblockdata.is(Blocks.JUKEBOX) && !(Boolean) iblockdata.getValue(JukeboxBlock.HAS_RECORD)) {
+            ItemStack itemstack = context.getItemInHand();
+
+            if (!world.isClientSide) {
+                if (true) return InteractionResult.SUCCESS; // CraftBukkit - handled in ItemStack
+                Player entityhuman = context.getPlayer();
+                BlockEntity tileentity = world.getBlockEntity(blockposition);
+
+                if (tileentity instanceof JukeboxBlockEntity) {
+                    JukeboxBlockEntity tileentityjukebox = (JukeboxBlockEntity) tileentity;
+
+                    tileentityjukebox.setTheItem(itemstack.copy());
+                    world.gameEvent(GameEvent.BLOCK_CHANGE, blockposition, GameEvent.Context.of(entityhuman, iblockdata));
                 }
 
                 itemInHand.shrink(1);
