--- a/net/minecraft/world/item/SignItem.java
+++ b/net/minecraft/world/item/SignItem.java
@@ -11,6 +12,9 @@
 import net.minecraft.world.level.block.state.BlockState;
 
 public class SignItem extends StandingAndWallBlockItem {
+
+    public static BlockPos openSign; // CraftBukkit
+
     public SignItem(Item.Properties properties, Block standingBlock, Block wallBlock) {
         super(standingBlock, wallBlock, properties, Direction.DOWN);
     }
@@ -22,12 +26,23 @@
     @Override
     protected boolean updateCustomBlockEntityTag(BlockPos pos, Level level, @Nullable Player player, ItemStack stack, BlockState state) {
         boolean flag = super.updateCustomBlockEntityTag(pos, level, player, stack, state);
-        if (!level.isClientSide
-            && !flag
-            && player != null
-            && level.getBlockEntity(pos) instanceof SignBlockEntity signBlockEntity
-            && level.getBlockState(pos).getBlock() instanceof SignBlock signBlock) {
-            signBlock.openTextEdit(player, signBlockEntity, true);
+
+        if (!level.isClientSide && !flag && player != null) {
+            BlockEntity tileentity = level.getBlockEntity(pos);
+
+            if (tileentity instanceof SignBlockEntity) {
+                SignBlockEntity tileentitysign = (SignBlockEntity) tileentity;
+                Block block = level.getBlockState(pos).getBlock();
+
+                if (block instanceof SignBlock) {
+                    SignBlock blocksign = (SignBlock) block;
+
+                    // CraftBukkit start - SPIGOT-4678
+                    // blocksign.openTextEdit(entityhuman, tileentitysign, true);
+                    SignItem.openSign = pos;
+                    // CraftBukkit end
+                }
+            }
         }
 
         return flag;
