--- a/net/minecraft/stats/StatsCounter.java
+++ b/net/minecraft/stats/StatsCounter.java
@@ -13,18 +13,24 @@
         this.stats.defaultReturnValue(0);
     }
 
-    public void increment(Player player, Stat<?> stat, int i) {
-        int j = (int) Math.min((long) this.getValue(stat) + (long) i, 2147483647L);
+    public void increment(Player player, Stat<?> stat, int amount) {
+        int j = (int) Math.min((long) this.getValue(stat) + (long) amount, 2147483647L);
 
+        // CraftBukkit start - fire Statistic events
+        org.bukkit.event.Cancellable cancellable = org.bukkit.craftbukkit.event.CraftEventFactory.handleStatisticsIncrease(player, stat, this.getValue(stat), j);
+        if (cancellable != null && cancellable.isCancelled()) {
+            return;
+        }
+        // CraftBukkit end
         this.setValue(player, stat, j);
     }
 
-    public void setValue(Player player, Stat<?> stat, int i) {
-        this.stats.put(stat, i);
+    public void setValue(Player player, Stat<?> stat, int value) {
+        this.stats.put(stat, value);
     }
 
-    public <T> int getValue(StatType<T> stattype, T t0) {
-        return stattype.contains(t0) ? this.getValue(stattype.get(t0)) : 0;
+    public <T> int getValue(StatType<T> type, T value) {
+        return type.contains(value) ? this.getValue(type.get(value)) : 0;
     }
 
     public int getValue(Stat<?> stat) {
