--- a/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
@@ -8,54 +8,57 @@
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.level.block.state.BlockBehaviour;
-import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.IBlockData;
 import net.minecraft.world.level.lighting.LightEngine;
 
 public abstract class SpreadingSnowyDirtBlock extends SnowyDirtBlock {
 
-    protected SpreadingSnowyDirtBlock(BlockBehaviour.Properties blockbehaviour_properties) {
-        super(blockbehaviour_properties);
+    protected SpreadingSnowyDirtBlock(BlockBehaviour.Properties properties) {
+        super(properties);
     }
 
-    private static boolean canBeGrass(BlockState blockstate, LevelReader levelreader, BlockPos blockpos) {
-        BlockPos blockpos1 = blockpos.above();
-        BlockState blockstate1 = levelreader.getBlockState(blockpos1);
+    private static boolean canBeGrass(IBlockData state, LevelReader levelReader, BlockPos pos) {
+        BlockPos blockposition1 = pos.above();
+        IBlockData iblockdata1 = levelReader.getBlockState(blockposition1);
 
-        if (blockstate1.is(Blocks.SNOW) && (Integer) blockstate1.getValue(SnowLayerBlock.LAYERS) == 1) {
+        if (iblockdata1.is(Blocks.SNOW) && (Integer) iblockdata1.getValue(SnowLayerBlock.LAYERS) == 1) {
             return true;
-        } else if (blockstate1.getFluidState().getAmount() == 8) {
+        } else if (iblockdata1.getFluidState().getAmount() == 8) {
             return false;
         } else {
-            int i = LightEngine.getLightBlockInto(levelreader, blockstate, blockpos, blockstate1, blockpos1, Direction.UP, blockstate1.getLightBlock(levelreader, blockpos1));
+            int i = LightEngine.getLightBlockInto(levelReader, state, pos, iblockdata1, blockposition1, Direction.UP, iblockdata1.getLightBlock(levelReader, blockposition1));
 
-            return i < levelreader.getMaxLightLevel();
+            return i < levelReader.getMaxLightLevel();
         }
     }
 
     @Override
-    @Override
     protected abstract MapCodec<? extends SpreadingSnowyDirtBlock> codec();
 
-    private static boolean canPropagate(BlockState blockstate, LevelReader levelreader, BlockPos blockpos) {
-        BlockPos blockpos1 = blockpos.above();
+    private static boolean canPropagate(IBlockData state, LevelReader level, BlockPos pos) {
+        BlockPos blockposition1 = pos.above();
 
-        return canBeGrass(blockstate, levelreader, blockpos) && !levelreader.getFluidState(blockpos1).is(FluidTags.WATER);
+        return canBeGrass(state, level, pos) && !level.getFluidState(blockposition1).is(FluidTags.WATER);
     }
 
     @Override
-    @Override
-    public void randomTick(BlockState blockstate, ServerLevel serverlevel, BlockPos blockpos, RandomSource randomsource) {
-        if (!canBeGrass(blockstate, serverlevel, blockpos)) {
-            serverlevel.setBlockAndUpdate(blockpos, Blocks.DIRT.defaultBlockState());
+    public void randomTick(IBlockData state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (!canBeGrass(state, level, pos)) {
+            // CraftBukkit start
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(level, pos, Blocks.DIRT.defaultBlockState()).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+            level.setBlockAndUpdate(pos, Blocks.DIRT.defaultBlockState());
         } else {
-            if (serverlevel.getMaxLocalRawBrightness(blockpos.above()) >= 9) {
-                BlockState blockstate1 = this.defaultBlockState();
+            if (level.getMaxLocalRawBrightness(pos.above()) >= 9) {
+                IBlockData iblockdata1 = this.defaultBlockState();
 
                 for (int i = 0; i < 4; ++i) {
-                    BlockPos blockpos1 = blockpos.offset(randomsource.nextInt(3) - 1, randomsource.nextInt(5) - 3, randomsource.nextInt(3) - 1);
+                    BlockPos blockposition1 = pos.offset(random.nextInt(3) - 1, random.nextInt(5) - 3, random.nextInt(3) - 1);
 
-                    if (serverlevel.getBlockState(blockpos1).is(Blocks.DIRT) && canPropagate(blockstate1, serverlevel, blockpos1)) {
-                        serverlevel.setBlockAndUpdate(blockpos1, (BlockState) blockstate1.setValue(SpreadingSnowyDirtBlock.SNOWY, serverlevel.getBlockState(blockpos1.above()).is(Blocks.SNOW)));
+                    if (level.getBlockState(blockposition1).is(Blocks.DIRT) && canPropagate(iblockdata1, level, blockposition1)) {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, blockposition1, (IBlockData) iblockdata1.setValue(SpreadingSnowyDirtBlock.SNOWY, level.getBlockState(blockposition1.above()).is(Blocks.SNOW))); // CraftBukkit
                     }
                 }
             }
