--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -9,63 +9,66 @@
 import net.minecraft.world.item.context.UseOnContext;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.Blocks;
-import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.block.state.IBlockData;
 import net.minecraft.world.level.dimension.end.EndDragonFight;
 import net.minecraft.world.level.gameevent.GameEvent;
 import net.minecraft.world.phys.AABB;
 
 public class EndCrystalItem extends Item {
 
-    public EndCrystalItem(Item.Properties item_properties) {
-        super(item_properties);
+    public EndCrystalItem(Item.Properties properties) {
+        super(properties);
     }
 
     @Override
-    @Override
-    public InteractionResult useOn(UseOnContext useoncontext) {
-        Level level = useoncontext.getLevel();
-        BlockPos blockpos = useoncontext.getClickedPos();
-        BlockState blockstate = level.getBlockState(blockpos);
+    public InteractionResult useOn(UseOnContext context) {
+        Level world = context.getLevel();
+        BlockPos blockposition = context.getClickedPos();
+        IBlockData iblockdata = world.getBlockState(blockposition);
 
-        if (!blockstate.is(Blocks.OBSIDIAN) && !blockstate.is(Blocks.BEDROCK)) {
+        if (!iblockdata.is(Blocks.OBSIDIAN) && !iblockdata.is(Blocks.BEDROCK)) {
             return InteractionResult.FAIL;
         } else {
-            BlockPos blockpos1 = blockpos.above();
+            BlockPos blockposition1 = blockposition.above();
 
-            if (!level.isEmptyBlock(blockpos1)) {
+            if (!world.isEmptyBlock(blockposition1)) {
                 return InteractionResult.FAIL;
             } else {
-                double d0 = (double) blockpos1.getX();
-                double d1 = (double) blockpos1.getY();
-                double d2 = (double) blockpos1.getZ();
-                List<Entity> list = level.getEntities((Entity) null, new AABB(d0, d1, d2, d0 + 1.0D, d1 + 2.0D, d2 + 1.0D));
+                double d0 = (double) blockposition1.getX();
+                double d1 = (double) blockposition1.getY();
+                double d2 = (double) blockposition1.getZ();
+                List<Entity> list = world.getEntities((Entity) null, new AABB(d0, d1, d2, d0 + 1.0D, d1 + 2.0D, d2 + 1.0D));
 
                 if (!list.isEmpty()) {
                     return InteractionResult.FAIL;
                 } else {
-                    if (level instanceof ServerLevel) {
-                        EndCrystal endcrystal = new EndCrystal(level, d0 + 0.5D, d1, d2 + 0.5D);
+                    if (world instanceof ServerLevel) {
+                        EndCrystal entityendercrystal = new EndCrystal(world, d0 + 0.5D, d1, d2 + 0.5D);
 
-                        endcrystal.setShowBottom(false);
-                        level.addFreshEntity(endcrystal);
-                        level.gameEvent((Entity) useoncontext.getPlayer(), GameEvent.ENTITY_PLACE, blockpos1);
-                        EndDragonFight enddragonfight = ((ServerLevel) level).getDragonFight();
+                        entityendercrystal.setShowBottom(false);
+                        // CraftBukkit start
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPlaceEvent(context, entityendercrystal).isCancelled()) {
+                            return InteractionResult.FAIL;
+                        }
+                        // CraftBukkit end
+                        world.addFreshEntity(entityendercrystal);
+                        world.gameEvent((Entity) context.getPlayer(), GameEvent.ENTITY_PLACE, blockposition1);
+                        EndDragonFight enderdragonbattle = ((ServerLevel) world).getDragonFight();
 
-                        if (enddragonfight != null) {
-                            enddragonfight.tryRespawn();
+                        if (enderdragonbattle != null) {
+                            enderdragonbattle.tryRespawn();
                         }
                     }
 
-                    useoncontext.getItemInHand().shrink(1);
-                    return InteractionResult.sidedSuccess(level.isClientSide);
+                    context.getItemInHand().shrink(1);
+                    return InteractionResult.sidedSuccess(world.isClientSide);
                 }
             }
         }
     }
 
     @Override
-    @Override
-    public boolean isFoil(ItemStack itemstack) {
+    public boolean isFoil(ItemStack stack) {
         return true;
     }
 }
