--- a/net/minecraft/world/entity/animal/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/Pufferfish.java
@@ -133,9 +141,10 @@
     }
 
     private void touch(Mob mob) {
-        int puffState = this.getPuffState();
-        if (mob.hurt(this.damageSources().mobAttack(this), (float)(1 + puffState))) {
-            mob.addEffect(new MobEffectInstance(MobEffects.POISON, 60 * puffState, 0), this);
+        int i = this.getPuffState();
+
+        if (mob.hurt(this.damageSources().mobAttack(this), (float) (1 + i))) {
+            mob.addEffect(new MobEffectInstance(MobEffects.POISON, 60 * i, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             this.playSound(SoundEvents.PUFFER_FISH_STING, 1.0F, 1.0F);
         }
     }
@@ -148,7 +159,7 @@
                 ((ServerPlayer)entity).connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.PUFFER_FISH_STING, 0.0F));
             }
 
-            entity.addEffect(new MobEffectInstance(MobEffects.POISON, 60 * puffState, 0), this);
+            entity.addEffect(new MobEffectInstance(MobEffects.POISON, 60 * i, 0), this, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
         }
     }
 
