--- a/net/minecraft/world/inventory/FurnaceResultSlot.java
+++ b/net/minecraft/world/inventory/FurnaceResultSlot.java
@@ -59,7 +54,7 @@
             if (container instanceof AbstractFurnaceBlockEntity) {
                 AbstractFurnaceBlockEntity abstractfurnaceblockentity = (AbstractFurnaceBlockEntity) container;
 
-                abstractfurnaceblockentity.awardUsedRecipesAndPopExperience(serverplayer);
+                tileentityfurnace.awardUsedRecipesAndPopExperience(entityplayer, stack, this.removeCount); // CraftBukkit
             }
         }
 
