--- a/net/minecraft/world/inventory/MerchantMenu.java
+++ b/net/minecraft/world/inventory/MerchantMenu.java
@@ -11,6 +11,7 @@
 import net.minecraft.world.item.trading.Merchant;
 import net.minecraft.world.item.trading.MerchantOffer;
 import net.minecraft.world.item.trading.MerchantOffers;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView; // CraftBukkit
 
 public class MerchantMenu extends AbstractContainerMenu {
 
@@ -31,9 +32,18 @@
     private boolean showProgressBar;
     private boolean canRestock;
 
-    public MerchantMenu(int i, Inventory inventory) {
-        this(i, inventory, new ClientSideMerchant(inventory.player));
+    // CraftBukkit start
+    private CraftInventoryView bukkitEntity = null;
+    private Inventory player;
+
+    @Override
+    public CraftInventoryView getBukkitView() {
+        if (bukkitEntity == null) {
+            bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), new org.bukkit.craftbukkit.inventory.CraftInventoryMerchant(trader, tradeContainer), this);
+        }
+        return bukkitEntity;
     }
+    // CraftBukkit end
 
     public MerchantMenu(int i, Inventory inventory, Merchant merchant) {
         super(MenuType.MERCHANT, i);
@@ -41,7 +55,8 @@
         this.tradeContainer = new MerchantContainer(merchant);
         this.addSlot(new Slot(this.tradeContainer, 0, 136, 37));
         this.addSlot(new Slot(this.tradeContainer, 1, 162, 37));
-        this.addSlot(new MerchantResultSlot(inventory.player, merchant, this.tradeContainer, 2, 220, 37));
+        this.addSlot(new MerchantResultSlot(playerInventory.player, trader, this.tradeContainer, 2, 220, 37));
+        this.player = playerInventory; // CraftBukkit - save player
 
         int j;
 
@@ -158,7 +169,7 @@
     }
 
     private void playTradeSound() {
-        if (!this.trader.isClientSide()) {
+        if (!this.trader.isClientSide() && this.trader instanceof Entity) { // CraftBukkit - SPIGOT-5035
             Entity entity = (Entity) this.trader;
 
             entity.level().playLocalSound(entity.getX(), entity.getY(), entity.getZ(), this.trader.getNotifyTradeSound(), SoundSource.NEUTRAL, 1.0F, 1.0F, false);
