--- a/net/minecraft/world/entity/npc/InventoryCarrier.java
+++ b/net/minecraft/world/entity/npc/InventoryCarrier.java
@@ -23,7 +23,14 @@
                 return;
             }
 
-            mob.onItemPickup(itementity);
+            // CraftBukkit start
+            ItemStack remaining = new SimpleContainer(inventorysubcontainer).addItem(itemstack);
+            if (org.bukkit.craftbukkit.event.CraftEventFactory.callEntityPickupItemEvent(mob, itemEntity, remaining.getCount(), false).isCancelled()) {
+                return;
+            }
+            // CraftBukkit end
+
+            mob.onItemPickup(itemEntity);
             int i = itemstack.getCount();
             ItemStack itemstack1 = simplecontainer.addItem(itemstack);
 
