--- a/net/minecraft/server/players/UserBanListEntry.java
+++ b/net/minecraft/server/players/UserBanListEntry.java
@@ -1,3 +1,4 @@
+// mc-dev import
 package net.minecraft.server.players;
 
 import com.google.gson.JsonObject;
@@ -8,8 +9,9 @@
 import net.minecraft.network.chat.Component;
 
 public class UserBanListEntry extends BanListEntry<GameProfile> {
+
     public UserBanListEntry(@Nullable GameProfile user) {
-        this(user, null, null, null, null);
+        this(user, (Date) null, (String) null, (Date) null, (String) null);
     }
 
     public UserBanListEntry(@Nullable GameProfile profile, @Nullable Date created, @Nullable String source, @Nullable Date expires, @Nullable String reason) {
@@ -23,31 +25,33 @@
     @Override
     protected void serialize(JsonObject data) {
         if (this.getUser() != null) {
-            data.addProperty("uuid", this.getUser().getId().toString());
-            data.addProperty("name", this.getUser().getName());
+            data.addProperty("uuid", ((GameProfile) this.getUser()).getId().toString());
+            data.addProperty("name", ((GameProfile) this.getUser()).getName());
             super.serialize(data);
         }
     }
 
     @Override
     public Component getDisplayName() {
-        GameProfile gameProfile = this.getUser();
-        return gameProfile != null ? Component.literal(gameProfile.getName()) : Component.translatable("commands.banlist.entry.unknown");
+        GameProfile gameprofile = (GameProfile) this.getUser();
+
+        return gameprofile != null ? Component.literal(gameprofile.getName()) : Component.translatable("commands.banlist.entry.unknown");
     }
 
     @Nullable
     private static GameProfile createGameProfile(JsonObject json) {
         if (json.has("uuid") && json.has("name")) {
-            String asString = json.get("uuid").getAsString();
+            String s = json.get("uuid").getAsString();
 
-            UUID uUID;
+            UUID uuid;
+
             try {
-                uUID = UUID.fromString(asString);
-            } catch (Throwable var4) {
+                uuid = UUID.fromString(s);
+            } catch (Throwable throwable) {
                 return null;
             }
 
-            return new GameProfile(uUID, json.get("name").getAsString());
+            return new GameProfile(uuid, json.get("name").getAsString());
         } else {
             return null;
         }
