From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Wed, 14 Feb 2024 19:33:27 +0100
Subject: [PATCH] Code generation marker stub


diff --git a/src/main/java/io/papermc/paper/enchantments/EnchantmentRarity.java b/src/main/java/io/papermc/paper/enchantments/EnchantmentRarity.java
index e6a40c1fcea761bd66743b50e3da3d14797d05b0..0ef8a64ca3fb93b878a0f67bb85629e7b85b07e2 100644
--- a/src/main/java/io/papermc/paper/enchantments/EnchantmentRarity.java
+++ b/src/main/java/io/papermc/paper/enchantments/EnchantmentRarity.java
@@ -2,10 +2,12 @@ package io.papermc.paper.enchantments;
 
 public enum EnchantmentRarity {
 
+    // Paper start - Generated/EnchantmentRarity
     COMMON(10),
     UNCOMMON(5),
     RARE(2),
     VERY_RARE(1);
+    // Paper end - Generated/EnchantmentRarity
 
     private final int weight;
 
diff --git a/src/main/java/io/papermc/paper/inventory/ItemRarity.java b/src/main/java/io/papermc/paper/inventory/ItemRarity.java
index 74ef8395cc040ce488c2acaa416db20272cc2734..73d3b072368ff9b0aa480659857c456c8c6d0438 100644
--- a/src/main/java/io/papermc/paper/inventory/ItemRarity.java
+++ b/src/main/java/io/papermc/paper/inventory/ItemRarity.java
@@ -6,10 +6,12 @@ import org.jetbrains.annotations.NotNull;
 
 public enum ItemRarity {
 
+    // Paper start - Generated/ItemRarity
     COMMON(NamedTextColor.WHITE),
     UNCOMMON(NamedTextColor.YELLOW),
     RARE(NamedTextColor.AQUA),
     EPIC(NamedTextColor.LIGHT_PURPLE);
+    // Paper end - Generated/ItemRarity
 
     TextColor color;
 
diff --git a/src/main/java/org/bukkit/Art.java b/src/main/java/org/bukkit/Art.java
index 24daba85ce4129fb0babe67570059ca8119360c0..7aca317825492157e78ef04470914c2ec5a9d45d 100644
--- a/src/main/java/org/bukkit/Art.java
+++ b/src/main/java/org/bukkit/Art.java
@@ -10,6 +10,7 @@ import org.jetbrains.annotations.Nullable;
  * Represents the art on a painting
  */
 public enum Art implements Keyed {
+    // Paper start - Generated/Art
     KEBAB(0, 1, 1),
     AZTEC(1, 1, 1),
     ALBAN(2, 1, 1),
@@ -40,6 +41,7 @@ public enum Art implements Keyed {
     WIND(27, 2, 2),
     WATER(28, 2, 2),
     FIRE(29, 2, 2);
+    // Paper end - Generated/Art
 
     private final int id, width, height;
     private final NamespacedKey key;
diff --git a/src/main/java/org/bukkit/Fluid.java b/src/main/java/org/bukkit/Fluid.java
index 7202b44a65e8699cf64329e441e83d4dd1779c19..ebcef7cdbcd402010af13d9fb055b3a3f6ce665e 100644
--- a/src/main/java/org/bukkit/Fluid.java
+++ b/src/main/java/org/bukkit/Fluid.java
@@ -8,6 +8,7 @@ import org.jetbrains.annotations.NotNull;
  */
 public enum Fluid implements Keyed {
 
+    // Paper start - Generated/Fluid
     // Paper start
     /**
      * No fluid.
@@ -30,6 +31,7 @@ public enum Fluid implements Keyed {
      * Flowing lava.
      */
     FLOWING_LAVA;
+    // Paper end - Generated/Fluid
 
     private final NamespacedKey key;
 
diff --git a/src/main/java/org/bukkit/GameEvent.java b/src/main/java/org/bukkit/GameEvent.java
index 6c9689baca1763e2ef79495d38618d587e792434..3073fe5929650076f99d2ea04ebd226534bbf93f 100644
--- a/src/main/java/org/bukkit/GameEvent.java
+++ b/src/main/java/org/bukkit/GameEvent.java
@@ -12,6 +12,7 @@ import org.jetbrains.annotations.Nullable;
  */
 public abstract class GameEvent implements Keyed {
 
+    // Paper start - Generated/GameEvent
     public static final GameEvent BLOCK_ACTIVATE = getEvent("block_activate");
     public static final GameEvent BLOCK_ATTACH = getEvent("block_attach");
     public static final GameEvent BLOCK_CHANGE = getEvent("block_change");
@@ -21,42 +22,18 @@ public abstract class GameEvent implements Keyed {
     public static final GameEvent BLOCK_DETACH = getEvent("block_detach");
     public static final GameEvent BLOCK_OPEN = getEvent("block_open");
     public static final GameEvent BLOCK_PLACE = getEvent("block_place");
-    @Deprecated
-    public static final GameEvent BLOCK_PRESS = getEvent("block_activate");
-    @Deprecated
-    public static final GameEvent BLOCK_SWITCH = getEvent("block_activate");
-    @Deprecated
-    public static final GameEvent BLOCK_UNPRESS = getEvent("block_deactivate");
-    @Deprecated
-    public static final GameEvent BLOCK_UNSWITCH = getEvent("block_deactivate");
     public static final GameEvent CONTAINER_CLOSE = getEvent("container_close");
     public static final GameEvent CONTAINER_OPEN = getEvent("container_open");
-    @Deprecated
-    public static final GameEvent DISPENSE_FAIL = getEvent("block_activate");
     public static final GameEvent DRINK = getEvent("drink");
-    @Deprecated
-    public static final GameEvent DRINKING_FINISH = getEvent("drink");
     public static final GameEvent EAT = getEvent("eat");
-    @Deprecated
-    public static final GameEvent ELYTRA_FREE_FALL = getEvent("elytra_glide");
     public static final GameEvent ELYTRA_GLIDE = getEvent("elytra_glide");
     public static final GameEvent ENTITY_DAMAGE = getEvent("entity_damage");
-    @Deprecated
-    public static final GameEvent ENTITY_DAMAGED = getEvent("entity_damage");
     public static final GameEvent ENTITY_DIE = getEvent("entity_die");
     public static final GameEvent ENTITY_DISMOUNT = getEvent("entity_dismount");
-    @Deprecated
-    public static final GameEvent ENTITY_DYING = getEvent("entity_die");
     public static final GameEvent ENTITY_INTERACT = getEvent("entity_interact");
     public static final GameEvent ENTITY_MOUNT = getEvent("entity_mount");
-    @Deprecated
-    public static final GameEvent ENTITY_KILLED = getEvent("entity_die");
     public static final GameEvent ENTITY_PLACE = getEvent("entity_place");
     public static final GameEvent ENTITY_ACTION = getEvent("entity_action");
-    @Deprecated
-    public static final GameEvent ENTITY_ROAR = getEvent("entity_action");
-    @Deprecated
-    public static final GameEvent ENTITY_SHAKE = getEvent("entity_action");
     public static final GameEvent EQUIP = getEvent("equip");
     public static final GameEvent EXPLODE = getEvent("explode");
     public static final GameEvent FLAP = getEvent("flap");
@@ -69,34 +46,18 @@ public abstract class GameEvent implements Keyed {
     public static final GameEvent JUKEBOX_PLAY = getEvent("jukebox_play");
     public static final GameEvent JUKEBOX_STOP_PLAY = getEvent("jukebox_stop_play");
     public static final GameEvent LIGHTNING_STRIKE = getEvent("lightning_strike");
-    @Deprecated
-    public static final GameEvent MOB_INTERACT = getEvent("entity_interact");
     public static final GameEvent NOTE_BLOCK_PLAY = getEvent("note_block_play");
-    @Deprecated
-    public static final GameEvent PISTON_CONTRACT = getEvent("block_deactivate");
-    @Deprecated
-    public static final GameEvent PISTON_EXTEND = getEvent("block_activate");
     public static final GameEvent PRIME_FUSE = getEvent("prime_fuse");
     public static final GameEvent PROJECTILE_LAND = getEvent("projectile_land");
     public static final GameEvent PROJECTILE_SHOOT = getEvent("projectile_shoot");
-    @Deprecated
-    public static final GameEvent RAVAGER_ROAR = getEvent("entity_action");
-    @Deprecated
-    public static final GameEvent RING_BELL = getEvent("block_change");
     public static final GameEvent SCULK_SENSOR_TENDRILS_CLICKING = getEvent("sculk_sensor_tendrils_clicking");
     public static final GameEvent SHEAR = getEvent("shear");
     public static final GameEvent SHRIEK = getEvent("shriek");
-    @Deprecated
-    public static final GameEvent SHULKER_CLOSE = getEvent("container_close");
-    @Deprecated
-    public static final GameEvent SHULKER_OPEN = getEvent("container_open");
     public static final GameEvent SPLASH = getEvent("splash");
     public static final GameEvent STEP = getEvent("step");
     public static final GameEvent SWIM = getEvent("swim");
     public static final GameEvent TELEPORT = getEvent("teleport");
     public static final GameEvent UNEQUIP = getEvent("unequip");
-    @Deprecated
-    public static final GameEvent WOLF_SHAKING = getEvent("entity_action");
     public static final GameEvent RESONATE_1 = getEvent("resonate_1");
     public static final GameEvent RESONATE_2 = getEvent("resonate_2");
     public static final GameEvent RESONATE_3 = getEvent("resonate_3");
@@ -112,6 +73,47 @@ public abstract class GameEvent implements Keyed {
     public static final GameEvent RESONATE_13 = getEvent("resonate_13");
     public static final GameEvent RESONATE_14 = getEvent("resonate_14");
     public static final GameEvent RESONATE_15 = getEvent("resonate_15");
+    // Paper end - Generated/GameEvent
+    @Deprecated
+    public static final GameEvent BLOCK_PRESS = getEvent("block_activate");
+    @Deprecated
+    public static final GameEvent BLOCK_SWITCH = getEvent("block_activate");
+    @Deprecated
+    public static final GameEvent BLOCK_UNPRESS = getEvent("block_deactivate");
+    @Deprecated
+    public static final GameEvent BLOCK_UNSWITCH = getEvent("block_deactivate");
+    @Deprecated
+    public static final GameEvent DISPENSE_FAIL = getEvent("block_activate");
+    @Deprecated
+    public static final GameEvent DRINKING_FINISH = getEvent("drink");
+    @Deprecated
+    public static final GameEvent ELYTRA_FREE_FALL = getEvent("elytra_glide");
+    @Deprecated
+    public static final GameEvent ENTITY_DAMAGED = getEvent("entity_damage");
+    @Deprecated
+    public static final GameEvent ENTITY_DYING = getEvent("entity_die");
+    @Deprecated
+    public static final GameEvent ENTITY_KILLED = getEvent("entity_die");
+    @Deprecated
+    public static final GameEvent ENTITY_ROAR = getEvent("entity_action");
+    @Deprecated
+    public static final GameEvent ENTITY_SHAKE = getEvent("entity_action");
+    @Deprecated
+    public static final GameEvent MOB_INTERACT = getEvent("entity_interact");
+    @Deprecated
+    public static final GameEvent PISTON_CONTRACT = getEvent("block_deactivate");
+    @Deprecated
+    public static final GameEvent PISTON_EXTEND = getEvent("block_activate");
+    @Deprecated
+    public static final GameEvent RAVAGER_ROAR = getEvent("entity_action");
+    @Deprecated
+    public static final GameEvent RING_BELL = getEvent("block_change");
+    @Deprecated
+    public static final GameEvent SHULKER_CLOSE = getEvent("container_close");
+    @Deprecated
+    public static final GameEvent SHULKER_OPEN = getEvent("container_open");
+    @Deprecated
+    public static final GameEvent WOLF_SHAKING = getEvent("entity_action");
 
     /**
      * Returns a {@link GameEvent} by a {@link NamespacedKey}.
diff --git a/src/main/java/org/bukkit/Material.java b/src/main/java/org/bukkit/Material.java
index f9af60356da4668cec8b24e73f5747ab82e35a91..64923077d86a7fcd3d81af23c390fe0a1d8d36f2 100644
--- a/src/main/java/org/bukkit/Material.java
+++ b/src/main/java/org/bukkit/Material.java
@@ -4942,6 +4942,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     // Paper end
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isBlock">
+            // Paper start - Generated/Material#isBlock
             case ACACIA_BUTTON:
             case ACACIA_DOOR:
             case ACACIA_FENCE:
@@ -6000,6 +6001,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case YELLOW_WOOL:
             case ZOMBIE_HEAD:
             case ZOMBIE_WALL_HEAD:
+            // Paper end - Generated/Material#isBlock
             //</editor-fold>
                 return true;
             default:
@@ -6015,6 +6017,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     public boolean isEdible() {
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isEdible">
+            // Paper start - Generated/Material#isEdible
             case APPLE:
             case BAKED_POTATO:
             case BEEF:
@@ -6055,6 +6058,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case SUSPICIOUS_STEW:
             case SWEET_BERRIES:
             case TROPICAL_FISH:
+            // Paper end - Generated/Material#isEdible
             // ----- Legacy Separator -----
             case LEGACY_BREAD:
             case LEGACY_CARROT_ITEM:
@@ -6191,6 +6195,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     public boolean isRecord() {
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isRecord">
+            // Paper start - Generated/Material#isRecord
             case MUSIC_DISC_5:
             case MUSIC_DISC_11:
             case MUSIC_DISC_13:
@@ -6207,6 +6212,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case MUSIC_DISC_STRAD:
             case MUSIC_DISC_WAIT:
             case MUSIC_DISC_WARD:
+            // Paper end - Generated/Material#isRecord
             //</editor-fold>
                 return true;
             default:
@@ -6225,6 +6231,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         }
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isSolid">
+            // Paper start - Generated/Material#isSolid
             case ACACIA_DOOR:
             case ACACIA_FENCE:
             case ACACIA_FENCE_GATE:
@@ -7058,6 +7065,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case YELLOW_TERRACOTTA:
             case YELLOW_WALL_BANNER:
             case YELLOW_WOOL:
+            // Paper end - Generated/Material#isSolid
             // ----- Legacy Separator -----
             case LEGACY_STONE:
             case LEGACY_GRASS:
@@ -7273,9 +7281,11 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     public boolean isAir() {
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isAir">
+            // Paper start - Generated/Material#isAir
             case AIR:
             case CAVE_AIR:
             case VOID_AIR:
+            // Paper end - Generated/Material#isAir
             // ----- Legacy Separator -----
             case LEGACY_AIR:
                 //</editor-fold>
@@ -7299,6 +7309,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         }
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isTransparent">
+            // Paper start - Generated/Material#isTransparent
             case ACACIA_BUTTON:
             case ACACIA_SAPLING:
             case ACTIVATOR_RAIL:
@@ -7426,6 +7437,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case YELLOW_CARPET:
             case ZOMBIE_HEAD:
             case ZOMBIE_WALL_HEAD:
+            // Paper end - Generated/Material#isTransparent
             // ----- Legacy Separator -----
             case LEGACY_AIR:
             case LEGACY_SAPLING:
@@ -7495,6 +7507,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         }
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isFlammable">
+            // Paper start - Generated/Material#isFlammable
             case ACACIA_DOOR:
             case ACACIA_FENCE:
             case ACACIA_FENCE_GATE:
@@ -7769,6 +7782,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case YELLOW_CARPET:
             case YELLOW_WALL_BANNER:
             case YELLOW_WOOL:
+            // Paper end - Generated/Material#isFlammable
             // ----- Legacy Separator -----
             case LEGACY_WOOD:
             case LEGACY_LOG:
@@ -7843,6 +7857,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         }
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isBurnable">
+            // Paper start - Generated/Material#isBurnable
             case ACACIA_FENCE:
             case ACACIA_FENCE_GATE:
             case ACACIA_LEAVES:
@@ -8017,6 +8032,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case WITHER_ROSE:
             case YELLOW_CARPET:
             case YELLOW_WOOL:
+            // Paper end - Generated/Material#isBurnable
             // ----- Legacy Separator -----
             case LEGACY_WOOD:
             case LEGACY_LOG:
@@ -8070,6 +8086,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     public boolean isFuel() {
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isFuel">
+            // Paper start - Generated/Material#isFuel
             case ACACIA_BOAT:
             case ACACIA_BUTTON:
             case ACACIA_CHEST_BOAT:
@@ -8318,6 +8335,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case YELLOW_BANNER:
             case YELLOW_CARPET:
             case YELLOW_WOOL:
+            // Paper end - Generated/Material#isFuel
             // ----- Legacy Separator -----
             case LEGACY_LAVA_BUCKET:
             case LEGACY_COAL_BLOCK:
@@ -8413,6 +8431,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         }
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isOccluding">
+            // Paper start - Generated/Material#isOccluding
             case ACACIA_LOG:
             case ACACIA_PLANKS:
             case ACACIA_WOOD:
@@ -8783,6 +8802,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case YELLOW_SHULKER_BOX:
             case YELLOW_TERRACOTTA:
             case YELLOW_WOOL:
+            // Paper end - Generated/Material#isOccluding
             // ----- Legacy Separator -----
             case LEGACY_STONE:
             case LEGACY_GRASS:
@@ -8898,6 +8918,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         }
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="hasGravity">
+            // Paper start - Generated/Material#hasGravity
             case ANVIL:
             case BLACK_CONCRETE_POWDER:
             case BLUE_CONCRETE_POWDER:
@@ -8921,6 +8942,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case SAND:
             case WHITE_CONCRETE_POWDER:
             case YELLOW_CONCRETE_POWDER:
+            // Paper end - Generated/Material#hasGravity
             // ----- Legacy Separator -----
             case LEGACY_SAND:
             case LEGACY_GRAVEL:
@@ -8941,6 +8963,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     public boolean isItem() {
         switch (this) {
             //<editor-fold defaultstate="collapsed" desc="isItem">
+            // Paper start - Generated/Material#isItem
             case ACACIA_WALL_HANGING_SIGN:
             case ACACIA_WALL_SIGN:
             case ATTACHED_MELON_STEM:
@@ -9088,6 +9111,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case YELLOW_CANDLE_CAKE:
             case YELLOW_WALL_BANNER:
             case ZOMBIE_WALL_HEAD:
+            // Paper end - Generated/Material#isItem
             // ----- Legacy Separator -----
             case LEGACY_ACACIA_DOOR:
             case LEGACY_BED_BLOCK:
@@ -9171,6 +9195,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
     public boolean isInteractable() {
         switch (this) {
             // <editor-fold defaultstate="collapsed" desc="isInteractable">
+            // Paper start - Generated/Material#isInteractable
             case ACACIA_BUTTON:
             case ACACIA_DOOR:
             case ACACIA_FENCE:
@@ -9454,6 +9479,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
             case YELLOW_CANDLE:
             case YELLOW_CANDLE_CAKE:
             case YELLOW_SHULKER_BOX:
+            // Paper end - Generated/Material#isInteractable
                 // </editor-fold>
                 return true;
             default:
@@ -9474,6 +9500,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         Preconditions.checkArgument(isBlock(), "The Material is not a block!");
         switch (this) {
             // <editor-fold defaultstate="collapsed" desc="getBlockHardness">
+            // Paper start - Generated/Material#getHardness
             case BARRIER:
             case BEDROCK:
             case CHAIN_COMMAND_BLOCK:
@@ -10404,6 +10431,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
                 return 100.0F;
             default:
                 return 0F;
+            // Paper end - Generated/Material#getHardness
             // </editor-fold>
         }
     }
@@ -10422,6 +10450,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         Preconditions.checkArgument(isBlock(), "The Material is not a block!");
         switch (this) {
             // <editor-fold defaultstate="collapsed" desc="getBlastResistance">
+            // Paper start - Generated/Material#getBlastResistance
             case BIG_DRIPLEAF:
             case BIG_DRIPLEAF_STEM:
             case BLACK_CANDLE:
@@ -11352,6 +11381,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
                 return 3600000.8F;
             default:
                 return 0;
+            // Paper end - Generated/Material#getBlastResistance
             // </editor-fold>
         }
     }
@@ -11372,6 +11402,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         Preconditions.checkArgument(isBlock(), "The Material is not a block!");
         switch (this) {
             // <editor-fold defaultstate="collapsed" desc="getSlipperiness">
+            // Paper start - Generated/Material#getSlipperiness
             default:
                 return 0.6F;
             case SLIME_BLOCK:
@@ -11382,6 +11413,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
                 return 0.98F;
             case BLUE_ICE:
                 return 0.989F;
+            // Paper end - Generated/Material#getSlipperiness
             // </editor-fold>
         }
     }
@@ -11399,6 +11431,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         Preconditions.checkArgument(isItem(), "The Material is not an item!");
         switch (this) {
             // <editor-fold defaultstate="collapsed" desc="getCraftingRemainingItem">
+            // Paper start - Generated/Material#getCraftingRemainingItem
             case WATER_BUCKET:
             case LAVA_BUCKET:
             case MILK_BUCKET:
@@ -11408,6 +11441,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
                 return GLASS_BOTTLE;
             default:
                 return null;
+            // Paper end - Generated/Material#getCraftingRemainingItem
             // </editor-fold>
         }
     }
@@ -11424,6 +11458,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
         Preconditions.checkArgument(isItem(), "The Material is not an item!");
         switch (this) {
             // <editor-fold defaultstate="collapsed" desc="getEquipmentSlot">
+            // Paper start - Generated/Material#getEquipmentSlot
             case CARVED_PUMPKIN:
             case CHAINMAIL_HELMET:
             case CREEPER_HEAD:
@@ -11466,6 +11501,7 @@ public enum Material implements Keyed, Translatable, net.kyori.adventure.transla
                 return EquipmentSlot.OFF_HAND;
             default:
                 return EquipmentSlot.HAND;
+            // Paper end - Generated/Material#getEquipmentSlot
             // </editor-fold>
         }
     }
diff --git a/src/main/java/org/bukkit/MusicInstrument.java b/src/main/java/org/bukkit/MusicInstrument.java
index ee5368372e136541eafe1d7ffb395de670fe4843..da849da92617fe123bf768ca2cdde4ac15f74959 100644
--- a/src/main/java/org/bukkit/MusicInstrument.java
+++ b/src/main/java/org/bukkit/MusicInstrument.java
@@ -9,6 +9,7 @@ import org.jetbrains.annotations.Nullable;
 
 public abstract class MusicInstrument implements Keyed, net.kyori.adventure.translation.Translatable { // Paper - translation keys
 
+    // Paper start - Generated/MusicInstrument
     public static final MusicInstrument PONDER = getInstrument("ponder_goat_horn");
     public static final MusicInstrument SING = getInstrument("sing_goat_horn");
     public static final MusicInstrument SEEK = getInstrument("seek_goat_horn");
@@ -17,6 +18,7 @@ public abstract class MusicInstrument implements Keyed, net.kyori.adventure.tran
     public static final MusicInstrument CALL = getInstrument("call_goat_horn");
     public static final MusicInstrument YEARN = getInstrument("yearn_goat_horn");
     public static final MusicInstrument DREAM = getInstrument("dream_goat_horn");
+    // Paper end - Generated/MusicInstrument
 
     /**
      * Returns a {@link MusicInstrument} by a {@link NamespacedKey}.
diff --git a/src/main/java/org/bukkit/Sound.java b/src/main/java/org/bukkit/Sound.java
index bb1b6bc9349af6dec369420035e860921da52589..98bdf175bea2f559655ccb96ceb900669752fbb5 100644
--- a/src/main/java/org/bukkit/Sound.java
+++ b/src/main/java/org/bukkit/Sound.java
@@ -12,6 +12,7 @@ import org.jetbrains.annotations.NotNull;
  */
 public enum Sound implements Keyed, net.kyori.adventure.sound.Sound.Type { // Paper - implement Sound.Type
 
+    // Paper start - Generated/Sound
     AMBIENT_BASALT_DELTAS_ADDITIONS("ambient.basalt_deltas.additions"),
     AMBIENT_BASALT_DELTAS_LOOP("ambient.basalt_deltas.loop"),
     AMBIENT_BASALT_DELTAS_MOOD("ambient.basalt_deltas.mood"),
@@ -1653,6 +1654,7 @@ public enum Sound implements Keyed, net.kyori.adventure.sound.Sound.Type { // Pa
     UI_TOAST_OUT("ui.toast.out"),
     WEATHER_RAIN("weather.rain"),
     WEATHER_RAIN_ABOVE("weather.rain.above");
+    // Paper end - Generated/Sound
 
     private final NamespacedKey key;
 
diff --git a/src/main/java/org/bukkit/Tag.java b/src/main/java/org/bukkit/Tag.java
index 1137a21afa0283fa3d391415e98171e3c160aa1f..97d39786e5026dafd35ee285d2b8cc2928346fce 100644
--- a/src/main/java/org/bukkit/Tag.java
+++ b/src/main/java/org/bukkit/Tag.java
@@ -19,6 +19,7 @@ import org.jetbrains.annotations.NotNull;
  */
 public interface Tag<T extends Keyed> extends Keyed {
 
+    // Paper start - Generated/Tag
     /**
      * Key for the built in block registry.
      */
@@ -53,11 +54,6 @@ public interface Tag<T extends Keyed> extends Keyed {
      * Vanilla block tag representing all colors of carpet.
      */
     Tag<Material> WOOL_CARPETS = Bukkit.getTag(REGISTRY_BLOCKS, NamespacedKey.minecraft("wool_carpets"), Material.class);
-    /**
-     * @deprecated {@link #WOOL_CARPETS}.
-     */
-    @Deprecated
-    Tag<Material> CARPETS = WOOL_CARPETS;
     /**
      * Vanilla block tag representing all wooden doors.
      */
@@ -1006,6 +1002,12 @@ public interface Tag<T extends Keyed> extends Keyed {
      */
     Tag<GameEvent> GAME_EVENT_ALLAY_CAN_LISTEN = Bukkit.getTag(REGISTRY_GAME_EVENTS, NamespacedKey.minecraft("allay_can_listen"), GameEvent.class);
     // Paper end
+    // Paper end - Generated/Tag
+    /**
+     * @deprecated {@link #WOOL_CARPETS}.
+     */
+    @Deprecated
+    Tag<Material> CARPETS = WOOL_CARPETS;
 
     /**
      * Returns whether or not this tag has an entry for the specified item.
diff --git a/src/main/java/org/bukkit/attribute/Attribute.java b/src/main/java/org/bukkit/attribute/Attribute.java
index 37f93b7e9f722e76631c3e7d3e770526ee0c3926..ee18e01cf5407ee1c6bd2626431bd283d0db56b2 100644
--- a/src/main/java/org/bukkit/attribute/Attribute.java
+++ b/src/main/java/org/bukkit/attribute/Attribute.java
@@ -9,6 +9,7 @@ import org.jetbrains.annotations.NotNull;
  */
 public enum Attribute implements Keyed, net.kyori.adventure.translation.Translatable { // Paper - Adventure translations
 
+    // Paper start - Generated/Attribute
     /**
      * Maximum health of an Entity.
      */
@@ -65,6 +66,7 @@ public enum Attribute implements Keyed, net.kyori.adventure.translation.Translat
      * Chance of a zombie to spawn reinforcements.
      */
     ZOMBIE_SPAWN_REINFORCEMENTS("zombie.spawn_reinforcements");
+    // Paper end - Generated/Attribute
 
     private final NamespacedKey key;
 
diff --git a/src/main/java/org/bukkit/block/Biome.java b/src/main/java/org/bukkit/block/Biome.java
index 5d8fa5b39a5d50cca48ba63af3a84b80f279b649..c6ea69c0d05aef5a29b758dbd3ea4c84a7587f6b 100644
--- a/src/main/java/org/bukkit/block/Biome.java
+++ b/src/main/java/org/bukkit/block/Biome.java
@@ -9,6 +9,7 @@ import org.jetbrains.annotations.NotNull;
  * Holds all accepted Biomes in the default server
  */
 public enum Biome implements Keyed, net.kyori.adventure.translation.Translatable { // Paper
+    // Paper start - Generated/Biome
     OCEAN,
     PLAINS,
     DESERT,
@@ -73,6 +74,7 @@ public enum Biome implements Keyed, net.kyori.adventure.translation.Translatable
     JAGGED_PEAKS,
     STONY_PEAKS,
     CHERRY_GROVE,
+    // Paper end - Generated/Biome
     /**
      * Represents a custom Biome
      */
diff --git a/src/main/java/org/bukkit/block/banner/PatternType.java b/src/main/java/org/bukkit/block/banner/PatternType.java
index 1c5c6303de815dd4ffa68f47dd2b6fa187fd4b70..09a0ab2e16a9ded3c011b0361c8568f9d7768244 100644
--- a/src/main/java/org/bukkit/block/banner/PatternType.java
+++ b/src/main/java/org/bukkit/block/banner/PatternType.java
@@ -10,6 +10,7 @@ import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 public enum PatternType implements Keyed {
+    // Paper start - Generated/PatternType
     BASE("b", "base"),
     SQUARE_BOTTOM_LEFT("bl", "square_bottom_left"),
     SQUARE_BOTTOM_RIGHT("br", "square_bottom_right"),
@@ -51,6 +52,7 @@ public enum PatternType implements Keyed {
     MOJANG("moj", "mojang"),
     GLOBE("glb", "globe"),
     PIGLIN("pig", "piglin");
+    // Paper end - Generated/PatternType
 
     private final String identifier;
     private final NamespacedKey key;
diff --git a/src/main/java/org/bukkit/damage/DamageType.java b/src/main/java/org/bukkit/damage/DamageType.java
index 5fef0f8b7957d5daef864de8547104f552215f29..de742b029409e2c4d74f854ca802130d6422ec88 100644
--- a/src/main/java/org/bukkit/damage/DamageType.java
+++ b/src/main/java/org/bukkit/damage/DamageType.java
@@ -20,6 +20,7 @@ import org.jetbrains.annotations.NotNull;
 @ApiStatus.Experimental
 public interface DamageType extends Keyed, Translatable {
 
+    // Paper start - Generated/DamageType
     public static final DamageType IN_FIRE = getDamageType("in_fire");
     public static final DamageType LIGHTNING_BOLT = getDamageType("lightning_bolt");
     public static final DamageType ON_FIRE = getDamageType("on_fire");
@@ -64,6 +65,7 @@ public interface DamageType extends Keyed, Translatable {
     public static final DamageType BAD_RESPAWN_POINT = getDamageType("bad_respawn_point");
     public static final DamageType OUTSIDE_BORDER = getDamageType("outside_border");
     public static final DamageType GENERIC_KILL = getDamageType("generic_kill");
+    // Paper end - Generated/DamageType
 
     @NotNull
     private static DamageType getDamageType(@NotNull String key) {
diff --git a/src/main/java/org/bukkit/entity/Boat.java b/src/main/java/org/bukkit/entity/Boat.java
index 2ac685fb1817f3ce06ebe6391cc863712d68367c..056d44476b2118b124ad2f7262aee8631d1fc0aa 100644
--- a/src/main/java/org/bukkit/entity/Boat.java
+++ b/src/main/java/org/bukkit/entity/Boat.java
@@ -136,6 +136,7 @@ public interface Boat extends Vehicle {
      * Represents the type of boats.
      */
     public enum Type {
+        // Paper start - Generated/BoatType
         OAK(Material.OAK_PLANKS),
         SPRUCE(Material.SPRUCE_PLANKS),
         BIRCH(Material.BIRCH_PLANKS),
@@ -146,6 +147,7 @@ public interface Boat extends Vehicle {
         MANGROVE(Material.MANGROVE_PLANKS),
         BAMBOO(Material.BAMBOO_PLANKS),
         ;
+        // Paper end - Generated/BoatType
 
         private final Material materialBlock;
 
@@ -170,11 +172,13 @@ public interface Boat extends Vehicle {
     public enum Status {
 
         NOT_IN_WORLD, // Paper
+        // Paper start - Generated/BoatStatus
         IN_WATER,
         UNDER_WATER,
         UNDER_FLOWING_WATER,
         ON_LAND,
         IN_AIR;
+        // Paper end - Generated/BoatStatus
     }
 
     // Paper start
diff --git a/src/main/java/org/bukkit/entity/Cat.java b/src/main/java/org/bukkit/entity/Cat.java
index d03adfaa4176617ef2ace2754fe02b63860e3aee..fceb2f0a452a1ce72988bbd1ecca599cfb6af6d9 100644
--- a/src/main/java/org/bukkit/entity/Cat.java
+++ b/src/main/java/org/bukkit/entity/Cat.java
@@ -46,6 +46,7 @@ public interface Cat extends Tameable, Sittable, io.papermc.paper.entity.CollarC
      * Represents the various different cat types there are.
      */
     public enum Type implements Keyed {
+        // Paper start - Generated/CatType
         TABBY("tabby"),
         BLACK("black"),
         RED("red"),
@@ -57,6 +58,7 @@ public interface Cat extends Tameable, Sittable, io.papermc.paper.entity.CollarC
         WHITE("white"),
         JELLIE("jellie"),
         ALL_BLACK("all_black");
+        // Paper end - Generated/CatType
 
         private final NamespacedKey key;
 
diff --git a/src/main/java/org/bukkit/entity/Fox.java b/src/main/java/org/bukkit/entity/Fox.java
index 473a7e36ad64f866d1d2e09e2ecb2e9881668faf..1fb12e0bc77247e9f6562bb7a8d0ff35eecb54a9 100644
--- a/src/main/java/org/bukkit/entity/Fox.java
+++ b/src/main/java/org/bukkit/entity/Fox.java
@@ -89,8 +89,10 @@ public interface Fox extends Animals, Sittable {
      * Represents the various different fox types there are.
      */
     public enum Type {
+        // Paper start - Generated/FoxType
         RED,
         SNOW;
+        // Paper end - Generated/FoxType
     }
 
     // Paper start - Add more fox behavior API
diff --git a/src/main/java/org/bukkit/entity/Frog.java b/src/main/java/org/bukkit/entity/Frog.java
index 0960bdbb3253bfe109dc6283244b8c9d95a850f7..d11cfd431cee09886ecd9d9416de85f56ac90462 100644
--- a/src/main/java/org/bukkit/entity/Frog.java
+++ b/src/main/java/org/bukkit/entity/Frog.java
@@ -46,6 +46,7 @@ public interface Frog extends Animals {
      */
     public enum Variant implements Keyed {
 
+        // Paper start - Generated/FrogVariant
         /**
          * Temperate (brown-orange) frog.
          */
@@ -58,6 +59,7 @@ public interface Frog extends Animals {
          * Cold (green) frog.
          */
         COLD;
+        // Paper end - Generated/FrogVariant
         private final NamespacedKey key;
 
         private Variant() {
diff --git a/src/main/java/org/bukkit/entity/Panda.java b/src/main/java/org/bukkit/entity/Panda.java
index aa5686df134185334a74429576ff0709a604dbfd..94ce1e46ef2fe4a8aa5ba79e1c941119b744a572 100644
--- a/src/main/java/org/bukkit/entity/Panda.java
+++ b/src/main/java/org/bukkit/entity/Panda.java
@@ -201,6 +201,7 @@ public interface Panda extends Animals, Sittable {
 
     public enum Gene {
 
+        // Paper start - Generated/PandaGene
         NORMAL(false),
         LAZY(false),
         WORRIED(false),
@@ -208,6 +209,7 @@ public interface Panda extends Animals, Sittable {
         BROWN(true),
         WEAK(true),
         AGGRESSIVE(false);
+        // Paper end - Generated/PandaGene
 
         private final boolean recessive;
 
diff --git a/src/main/java/org/bukkit/entity/Sniffer.java b/src/main/java/org/bukkit/entity/Sniffer.java
index 15a0a733b0e5804655b5957cbf20831290d52a08..53228454d68eb57f6f0e70760ac02fd7998312f8 100644
--- a/src/main/java/org/bukkit/entity/Sniffer.java
+++ b/src/main/java/org/bukkit/entity/Sniffer.java
@@ -74,6 +74,7 @@ public interface Sniffer extends Animals {
      * Represents the current state of the Sniffer.
      */
     public enum State {
+        // Paper start - Generated/SnifferState
         IDLING,
         FEELING_HAPPY,
         SCENTING,
@@ -81,5 +82,6 @@ public interface Sniffer extends Animals {
         SEARCHING,
         DIGGING,
         RISING;
+        // Paper end - Generated/SnifferState
     }
 }
diff --git a/src/main/java/org/bukkit/entity/TropicalFish.java b/src/main/java/org/bukkit/entity/TropicalFish.java
index 84bba32e80d755f094975b667f1bf2a132087f4f..4fec7b3bbeae7ab0538f3de6e83faa9c5aa0b8ad 100644
--- a/src/main/java/org/bukkit/entity/TropicalFish.java
+++ b/src/main/java/org/bukkit/entity/TropicalFish.java
@@ -60,6 +60,7 @@ public interface TropicalFish extends io.papermc.paper.entity.SchoolableFish { /
      */
     public static enum Pattern {
 
+        // Paper start - Generated/TropicalFishPattern
         KOB,
         SUNSTREAK,
         SNOOPER,
@@ -72,5 +73,6 @@ public interface TropicalFish extends io.papermc.paper.entity.SchoolableFish { /
         BLOCKFISH,
         BETTY,
         CLAYFISH;
+        // Paper end - Generated/TropicalFishPattern
     }
 }
diff --git a/src/main/java/org/bukkit/entity/Villager.java b/src/main/java/org/bukkit/entity/Villager.java
index 3bc24457d143449e6a338d79becf7c39b9f81054..b6f07418ab7267cc43d1174fc023930e04eba360 100644
--- a/src/main/java/org/bukkit/entity/Villager.java
+++ b/src/main/java/org/bukkit/entity/Villager.java
@@ -167,6 +167,7 @@ public interface Villager extends AbstractVillager {
      */
     public enum Type implements Keyed {
 
+        // Paper start - Generated/VillagerType
         DESERT,
         JUNGLE,
         PLAINS,
@@ -174,6 +175,7 @@ public interface Villager extends AbstractVillager {
         SNOW,
         SWAMP,
         TAIGA;
+        // Paper end - Generated/VillagerType
         private final NamespacedKey key;
 
         private Type() {
diff --git a/src/main/java/org/bukkit/generator/structure/Structure.java b/src/main/java/org/bukkit/generator/structure/Structure.java
index 8e39f282c771ddafe5d890dcf065c56f0c633647..5505705a63e9ed8a820f7f4e164d6c1dd2936b05 100644
--- a/src/main/java/org/bukkit/generator/structure/Structure.java
+++ b/src/main/java/org/bukkit/generator/structure/Structure.java
@@ -15,6 +15,7 @@ import org.jetbrains.annotations.NotNull;
  */
 public abstract class Structure implements Keyed {
 
+    // Paper start - Generated/Structure
     public static final Structure PILLAGER_OUTPOST = getStructure("pillager_outpost");
     public static final Structure MINESHAFT = getStructure("mineshaft");
     public static final Structure MINESHAFT_MESA = getStructure("mineshaft_mesa");
@@ -51,6 +52,7 @@ public abstract class Structure implements Keyed {
     @MinecraftExperimental
     @org.jetbrains.annotations.ApiStatus.Experimental // Paper - add missing annotation
     public static final Structure TRIAL_CHAMBERS = getStructure("trial_chambers");
+    // Paper end - Generated/Structure
 
     private static Structure getStructure(String name) {
         return Registry.STRUCTURE.get(NamespacedKey.minecraft(name));
diff --git a/src/main/java/org/bukkit/generator/structure/StructureType.java b/src/main/java/org/bukkit/generator/structure/StructureType.java
index 2f908d5545d76ea25e26f6a4be579460512b5faf..77a03c7c35da902a802a90a41d8583172a9ca992 100644
--- a/src/main/java/org/bukkit/generator/structure/StructureType.java
+++ b/src/main/java/org/bukkit/generator/structure/StructureType.java
@@ -14,6 +14,7 @@ import org.bukkit.Registry;
  */
 public abstract class StructureType implements Keyed {
 
+    // Paper start - Generated/StructureType
     public static final StructureType BURIED_TREASURE = getStructureType("buried_treasure");
     public static final StructureType DESERT_PYRAMID = getStructureType("desert_pyramid");
     public static final StructureType END_CITY = getStructureType("end_city");
@@ -30,6 +31,7 @@ public abstract class StructureType implements Keyed {
     public static final StructureType STRONGHOLD = getStructureType("stronghold");
     public static final StructureType SWAMP_HUT = getStructureType("swamp_hut");
     public static final StructureType WOODLAND_MANSION = getStructureType("woodland_mansion");
+    // Paper end - Generated/StructureType
 
     private static StructureType getStructureType(String name) {
         return Registry.STRUCTURE_TYPE.get(NamespacedKey.minecraft(name));
diff --git a/src/main/java/org/bukkit/inventory/meta/trim/TrimMaterial.java b/src/main/java/org/bukkit/inventory/meta/trim/TrimMaterial.java
index 178d83cb3ccff2d12477d3c13ca4f108fa17e619..86e5957c70bd369f08c86d8ed588d2dc285a61f9 100644
--- a/src/main/java/org/bukkit/inventory/meta/trim/TrimMaterial.java
+++ b/src/main/java/org/bukkit/inventory/meta/trim/TrimMaterial.java
@@ -10,6 +10,7 @@ import org.bukkit.Registry;
  */
 public interface TrimMaterial extends Keyed {
 
+    // Paper start - Generated/TrimMaterial
     /**
      * {@link Material#QUARTZ}.
      */
@@ -50,6 +51,7 @@ public interface TrimMaterial extends Keyed {
      * {@link Material#AMETHYST_SHARD}.
      */
     public static final TrimMaterial AMETHYST = Registry.TRIM_MATERIAL.get(NamespacedKey.minecraft("amethyst"));
+    // Paper end - Generated/TrimMaterial
     // Paper start
     /**
      * @deprecated use {@link Registry#getKey(Keyed)} and {@link Registry#TRIM_MATERIAL}. TrimMaterials
diff --git a/src/main/java/org/bukkit/inventory/meta/trim/TrimPattern.java b/src/main/java/org/bukkit/inventory/meta/trim/TrimPattern.java
index e29fc42ae2b9c555db63d10d20552748e28ba60e..ae7f73f391b514266371021e4171fde6e8fda6b3 100644
--- a/src/main/java/org/bukkit/inventory/meta/trim/TrimPattern.java
+++ b/src/main/java/org/bukkit/inventory/meta/trim/TrimPattern.java
@@ -10,6 +10,7 @@ import org.bukkit.Registry;
  */
 public interface TrimPattern extends Keyed {
 
+    // Paper start - Generated/TrimPattern
     /**
      * {@link Material#SENTRY_ARMOR_TRIM_SMITHING_TEMPLATE}.
      */
@@ -74,6 +75,7 @@ public interface TrimPattern extends Keyed {
      * {@link Material#HOST_ARMOR_TRIM_SMITHING_TEMPLATE}.
      */
     public static final TrimPattern HOST = Registry.TRIM_PATTERN.get(NamespacedKey.minecraft("host"));
+    // Paper end - Generated/TrimPattern
     // Paper start
     /**
      * @deprecated use {@link Registry#getKey(Keyed)} and {@link Registry#TRIM_PATTERN}. TrimPatterns
diff --git a/src/main/java/org/bukkit/inventory/recipe/CookingBookCategory.java b/src/main/java/org/bukkit/inventory/recipe/CookingBookCategory.java
index 9990a7abf8f910ff5da45b6e5d94902e825e2096..54cd8b54e9516f6d900f9a550fc220063b494166 100644
--- a/src/main/java/org/bukkit/inventory/recipe/CookingBookCategory.java
+++ b/src/main/java/org/bukkit/inventory/recipe/CookingBookCategory.java
@@ -5,7 +5,9 @@ package org.bukkit.inventory.recipe;
  */
 public enum CookingBookCategory {
 
+    // Paper start - Generated/CookingBookCategory
     FOOD,
     BLOCKS,
     MISC;
+    // Paper end - Generated/CookingBookCategory
 }
diff --git a/src/main/java/org/bukkit/inventory/recipe/CraftingBookCategory.java b/src/main/java/org/bukkit/inventory/recipe/CraftingBookCategory.java
index 09a6a1b9b9ffb1e269bdaa1efc19cd8bedc2b499..3eecca87f204dcaebf580b7284859c4300057715 100644
--- a/src/main/java/org/bukkit/inventory/recipe/CraftingBookCategory.java
+++ b/src/main/java/org/bukkit/inventory/recipe/CraftingBookCategory.java
@@ -5,8 +5,10 @@ package org.bukkit.inventory.recipe;
  */
 public enum CraftingBookCategory {
 
+    // Paper start - Generated/CraftingBookCategory
     BUILDING,
     REDSTONE,
     EQUIPMENT,
     MISC;
+    // Paper end - Generated/CraftingBookCategory
 }
diff --git a/src/main/java/org/bukkit/map/MapCursor.java b/src/main/java/org/bukkit/map/MapCursor.java
index 599399e4d8ca7c46747439ad2dd0186d0a289a19..da69e7750ccac139ec2e8e240fbb7ef53d136683 100644
--- a/src/main/java/org/bukkit/map/MapCursor.java
+++ b/src/main/java/org/bukkit/map/MapCursor.java
@@ -282,6 +282,7 @@ public final class MapCursor {
      * resource pack.
      */
     public enum Type {
+        // Paper start - Generated/MapCursorType
         WHITE_POINTER(0),
         GREEN_POINTER(1),
         RED_POINTER(2),
@@ -317,6 +318,7 @@ public final class MapCursor {
         JUNGLE_TEMPLE(32),
         SWAMP_HUT(33),
         ;
+        // Paper end - Generated/MapCursorType
 
         private byte value;
 
diff --git a/src/main/java/org/bukkit/map/MapPalette.java b/src/main/java/org/bukkit/map/MapPalette.java
index c80faa079eca1564847070f0338fc98024639829..825b6c7b4e6b24f8d49da7b44f37e8a5ddc2e2c1 100644
--- a/src/main/java/org/bukkit/map/MapPalette.java
+++ b/src/main/java/org/bukkit/map/MapPalette.java
@@ -41,6 +41,7 @@ public final class MapPalette {
 
     @NotNull
     static final Color[] colors = {
+        // Paper start - Generated/MapPalette#colors
         c(0, 0, 0, 0), c(0, 0, 0, 0), c(0, 0, 0, 0), c(0, 0, 0, 0),
         c(89, 125, 39), c(109, 153, 48), c(127, 178, 56), c(67, 94, 29),
         c(174, 164, 115), c(213, 201, 140), c(247, 233, 163), c(130, 123, 86),
@@ -103,6 +104,7 @@ public final class MapPalette {
         c(70, 70, 70), c(86, 86, 86), c(100, 100, 100), c(52, 52, 52),
         c(152, 123, 103), c(186, 150, 126), c(216, 175, 147), c(114, 92, 77),
         c(89, 117, 105), c(109, 144, 129), c(127, 167, 150), c(67, 88, 79)
+        // Paper end - Generated/MapPalette#colors
     };
 
     // Interface
diff --git a/src/main/java/org/bukkit/potion/PotionType.java b/src/main/java/org/bukkit/potion/PotionType.java
index 70456f08668733a9a2840be6b08b078ad63e6825..13e1db8778adea4591dbb85870ff366fd565aa9e 100644
--- a/src/main/java/org/bukkit/potion/PotionType.java
+++ b/src/main/java/org/bukkit/potion/PotionType.java
@@ -15,6 +15,7 @@ import org.jetbrains.annotations.Nullable;
  * the Creative mode inventory
  */
 public enum PotionType implements Keyed {
+    // Paper start - Generated/PotionType
     UNCRAFTABLE("empty"),
     WATER("water"),
     MUNDANE("mundane"),
@@ -59,6 +60,7 @@ public enum PotionType implements Keyed {
     SLOW_FALLING("slow_falling"),
     LONG_SLOW_FALLING("long_slow_falling"),
     ;
+    // Paper end - Generated/PotionType
 
     private final NamespacedKey key;
     private final Supplier<InternalPotionData> internalPotionDataSupplier;
diff --git a/src/main/java/org/bukkit/scoreboard/DisplaySlot.java b/src/main/java/org/bukkit/scoreboard/DisplaySlot.java
index 21cd2ba659504c3a1eb95226539a5701d0c324db..3f6044892ad17bb99786e0009c937d294e90aa09 100644
--- a/src/main/java/org/bukkit/scoreboard/DisplaySlot.java
+++ b/src/main/java/org/bukkit/scoreboard/DisplaySlot.java
@@ -6,6 +6,7 @@ import net.kyori.adventure.text.format.NamedTextColor; // Paper
  */
 public enum DisplaySlot {
     // Paper start
+    // Paper start - Generated/DisplaySlot
     BELOW_NAME("below_name"),
     PLAYER_LIST("list"),
     SIDEBAR("sidebar"),
@@ -25,6 +26,7 @@ public enum DisplaySlot {
     SIDEBAR_TEAM_LIGHT_PURPLE(NamedTextColor.LIGHT_PURPLE),
     SIDEBAR_TEAM_YELLOW(NamedTextColor.YELLOW),
     SIDEBAR_TEAM_WHITE(NamedTextColor.WHITE);
+    // Paper end - Generated/DisplaySlot
 
     public static final net.kyori.adventure.util.Index<String, DisplaySlot> NAMES = net.kyori.adventure.util.Index.create(DisplaySlot.class, DisplaySlot::getId);
 
