From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Thu, 11 Apr 2024 21:31:18 +0200
Subject: [PATCH] fix EntityType#isSpawnable


diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
index 48843885feebcf26a48a23f8085fcb246b6bbe18..e32b001ea23d1bf71a5774cfd99279ef08486dd7 100644
--- a/src/main/java/org/bukkit/entity/EntityType.java
+++ b/src/main/java/org/bukkit/entity/EntityType.java
@@ -34,7 +34,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
      * Spawn with {@link World#dropItem(Location, ItemStack)} or {@link
      * World#dropItemNaturally(Location, ItemStack)}
      */
-    DROPPED_ITEM("item", Item.class, 1, false),
+    DROPPED_ITEM("item", Item.class, 1),
     /**
      * An experience orb.
      */
@@ -94,7 +94,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
     /**
      * A flying splash potion.
      */
-    SPLASH_POTION("potion", ThrownPotion.class, 16, false),
+    SPLASH_POTION("potion", ThrownPotion.class, 16),
     /**
      * A flying experience bottle.
      */
@@ -114,11 +114,11 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
     /**
      * A block that is going to or is about to fall.
      */
-    FALLING_BLOCK("falling_block", FallingBlock.class, 21, false),
+    FALLING_BLOCK("falling_block", FallingBlock.class, 21),
     /**
      * Internal representation of a Firework once it has been launched.
      */
-    FIREWORK("firework_rocket", Firework.class, 22, false),
+    FIREWORK("firework_rocket", Firework.class, 22),
     /**
      * @see Husk
      */
@@ -302,7 +302,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
      * <p>
      * Spawn with {@link World#strikeLightning(Location)}.
      */
-    LIGHTNING("lightning_bolt", LightningStrike.class, -1, false),
+    LIGHTNING("lightning_bolt", LightningStrike.class, -1),
     PLAYER("player", Player.class, -1, false),
     // Paper end - Generated/EntityType
     /**
