From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Thu, 11 Apr 2024 21:31:18 +0200
Subject: [PATCH] fix EntityType#isSpawnable


diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
index 65fdf77ca89eadae84bf1025215ebd3338a5acf5..0b9cf50dff279abd3209460e97799828374c279c 100644
--- a/src/main/java/org/bukkit/entity/EntityType.java
+++ b/src/main/java/org/bukkit/entity/EntityType.java
@@ -35,7 +35,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
      * Spawn with {@link World#dropItem(Location, ItemStack)} or {@link
      * World#dropItemNaturally(Location, ItemStack)}
      */
-    ITEM("item", Item.class, 1, false),
+    ITEM("item", Item.class, 1),
     /**
      * An experience orb.
      */
@@ -95,7 +95,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
     /**
      * A flying splash potion.
      */
-    POTION("potion", ThrownPotion.class, 16, false),
+    POTION("potion", ThrownPotion.class, 16),
     /**
      * A flying experience bottle.
      */
@@ -115,11 +115,11 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
     /**
      * A block that is going to or is about to fall.
      */
-    FALLING_BLOCK("falling_block", FallingBlock.class, 21, false),
+    FALLING_BLOCK("falling_block", FallingBlock.class, 21),
     /**
      * Internal representation of a Firework once it has been launched.
      */
-    FIREWORK_ROCKET("firework_rocket", Firework.class, 22, false),
+    FIREWORK_ROCKET("firework_rocket", Firework.class, 22),
     /**
      * @see Husk
      */
@@ -313,7 +313,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
      * <p>
      * Spawn with {@link World#strikeLightning(Location)}.
      */
-    LIGHTNING_BOLT("lightning_bolt", LightningStrike.class, -1, false),
+    LIGHTNING_BOLT("lightning_bolt", LightningStrike.class, -1),
     PLAYER("player", Player.class, -1, false),
     // Paper end - Generated/EntityType
     /**
