--- a/net/minecraft/CrashReport.java
+++ b/net/minecraft/CrashReport.java
@@ -33,9 +33,10 @@
     private StackTraceElement[] uncategorizedStackTrace = new StackTraceElement[0];
     private final SystemReport systemReport = new SystemReport();
 
-    public CrashReport(String s, Throwable throwable) {
-        this.title = s;
-        this.exception = throwable;
+    public CrashReport(String title, Throwable exception) {
+        this.title = title;
+        this.exception = exception;
+        this.systemReport.setDetail("CraftBukkit Information", new org.bukkit.craftbukkit.CraftCrashReport()); // CraftBukkit
     }
 
     public String getTitle() {
