From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Sun, 3 Mar 2024 18:05:10 +0100
Subject: [PATCH] server gen test


diff --git a/build.gradle.kts b/build.gradle.kts
index 75194254bc0ca416fb4590ca0886e4c1136cb8e1..557be470c33e5c249daa4f9702efcf676b9e0f2f 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -176,14 +176,18 @@ fun TaskContainer.registerRunTask(
 }
 
 // Paper start
+val generatedServerFolder: File = rootProject.file("paper-api-generator/generatedServerTest")
+
 sourceSets.main {
     java {
+        srcDir(generatedServerFolder)
         exclude("excluded/**")
     }
 }
 
 idea {
     module {
+        generatedSourceDirs.add(generatedServerFolder)
         excludeDirs.add(projectDir.resolve("src/main/java/excluded"))
     }
 }
