From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Date: Sun, 3 Mar 2024 18:05:10 +0100
Subject: [PATCH] server gen test


diff --git a/build.gradle.kts b/build.gradle.kts
index 3c80c2588497c95cd3c7253063fca95ef48d46f8..0e14f013b024b165e155ab08f4580c4a9d0558cc 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -182,14 +182,18 @@ fun TaskContainer.registerRunTask(
 }
 
 // Paper start
+val generatedServerFolder: File = rootProject.file("paper-api-generator/generatedServerTest")
+
 sourceSets.main {
     java {
+        srcDir(generatedServerFolder)
         exclude("excluded/**")
     }
 }
 
 idea {
     module {
+        generatedSourceDirs.add(generatedServerFolder)
         excludeDirs.add(projectDir.resolve("src/main/java/excluded"))
     }
 }
