--- a/net/minecraft/world/entity/monster/Spider.java
+++ b/net/minecraft/world/entity/monster/Spider.java
@@ -195,8 +181,8 @@
             Spider.SpiderEffectsGroupData spider_spidereffectsgroupdata = (Spider.SpiderEffectsGroupData) object;
             MobEffect mobeffect = spider_spidereffectsgroupdata.effect;
 
-            if (mobeffect != null) {
-                this.addEffect(new MobEffectInstance(mobeffect, -1));
+            if (mobeffectlist != null) {
+                this.addEffect(new MobEffectInstance(mobeffectlist, -1), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.SPIDER_SPAWN); // CraftBukkit
             }
         }
 
