--- a/net/minecraft/world/entity/monster/Witch.java
+++ b/net/minecraft/world/entity/monster/Witch.java
@@ -140,7 +134,7 @@
                             while (iterator.hasNext()) {
                                 MobEffectInstance mobeffectinstance = (MobEffectInstance) iterator.next();
 
-                                this.addEffect(new MobEffectInstance(mobeffectinstance));
+                                this.addEffect(new MobEffectInstance(mobeffect), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
                             }
                         }
                     }
