--- a/net/minecraft/world/entity/projectile/SpectralArrow.java
+++ b/net/minecraft/world/entity/projectile/SpectralArrow.java
@@ -11,34 +11,37 @@
 import net.minecraft.world.level.Level;
 
 public class SpectralArrow extends AbstractArrow {
+
     private static final ItemStack DEFAULT_ARROW_STACK = new ItemStack(Items.SPECTRAL_ARROW);
-    private int duration = 200;
+    public int duration = 200;
 
     public SpectralArrow(EntityType<? extends SpectralArrow> entityType, Level level) {
-        super(entityType, level, DEFAULT_ARROW_STACK);
+        super(entityType, level, SpectralArrow.DEFAULT_ARROW_STACK);
     }
 
-    public SpectralArrow(Level level, LivingEntity livingEntity, ItemStack itemStack) {
-        super(EntityType.SPECTRAL_ARROW, livingEntity, level, itemStack);
+    public SpectralArrow(Level world, LivingEntity entityliving, ItemStack itemstack) {
+        super(EntityType.SPECTRAL_ARROW, entityliving, world, itemstack);
     }
 
-    public SpectralArrow(Level level, double d, double d1, double d2, ItemStack itemStack) {
-        super(EntityType.SPECTRAL_ARROW, d, d1, d2, level, itemStack);
+    public SpectralArrow(Level world, double d0, double d1, double d2, ItemStack itemstack) {
+        super(EntityType.SPECTRAL_ARROW, d0, d1, d2, world, itemstack);
     }
 
     @Override
     public void tick() {
         super.tick();
         if (this.level().isClientSide && !this.inGround) {
-            this.level().addParticle(ParticleTypes.INSTANT_EFFECT, this.getX(), this.getY(), this.getZ(), 0.0, 0.0, 0.0);
+            this.level().addParticle(ParticleTypes.INSTANT_EFFECT, this.getX(), this.getY(), this.getZ(), 0.0D, 0.0D, 0.0D);
         }
+
     }
 
     @Override
     protected void doPostHurtEffects(LivingEntity living) {
         super.doPostHurtEffects(living);
-        MobEffectInstance mobEffectInstance = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
-        living.addEffect(mobEffectInstance, this.getEffectSource());
+        MobEffectInstance mobeffect = new MobEffectInstance(MobEffects.GLOWING, this.duration, 0);
+
+        living.addEffect(mobeffect, this.getEffectSource(), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
     }
 
     @Override
@@ -47,6 +50,7 @@
         if (compound.contains("Duration")) {
             this.duration = compound.getInt("Duration");
         }
+
     }
 
     @Override
