--- a/net/minecraft/world/entity/monster/Phantom.java
+++ b/net/minecraft/world/entity/monster/Phantom.java
@@ -508,7 +549,10 @@
                             cat.hiss();
                         }
 
-                        this.isScaredOfCat = !entitiesOfClass.isEmpty();
+                        if (Phantom.this.canAttack(entityhuman, TargetingConditions.DEFAULT)) {
+                            Phantom.this.setTarget(entityhuman, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER, true); // CraftBukkit - reason
+                            return true;
+                        }
                     }
 
                     return !this.isScaredOfCat;
