--- a/net/minecraft/world/entity/monster/Spider.java
+++ b/net/minecraft/world/entity/monster/Spider.java
@@ -173,10 +177,12 @@
             }
         }
 
-        if (var9 instanceof Spider.SpiderEffectsGroupData spiderEffectsGroupData) {
-            MobEffect mobEffect = spiderEffectsGroupData.effect;
-            if (mobEffect != null) {
-                this.addEffect(new MobEffectInstance(mobEffect, -1));
+        if (object instanceof Spider.SpiderEffectsGroupData) {
+            Spider.SpiderEffectsGroupData entityspider_groupdataspider = (Spider.SpiderEffectsGroupData) object;
+            MobEffect mobeffectlist = entityspider_groupdataspider.effect;
+
+            if (mobeffectlist != null) {
+                this.addEffect(new MobEffectInstance(mobeffectlist, -1), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.SPIDER_SPAWN); // CraftBukkit
             }
         }
 
