--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -320,8 +349,16 @@
                 var10000 = false;
             }
 
-            boolean flag1 = var10000;
-            this.setSuffocating(!flag || flag1);
+            boolean flag2 = flag1;
+
+            // CraftBukkit start
+            boolean suffocating = !flag || flag2;
+            if (suffocating ^ this.isSuffocating()) {
+                if (org.bukkit.craftbukkit.event.CraftEventFactory.callStriderTemperatureChangeEvent(this, suffocating)) {
+                    this.setSuffocating(suffocating);
+                }
+            }
+            // CraftBukkit end
         }
 
         super.tick();
