--- a/net/minecraft/world/entity/monster/Vex.java
+++ b/net/minecraft/world/entity/monster/Vex.java
@@ -362,8 +403,9 @@
         }
 
         @Override
-        public boolean canContinueToUse() {
-            return false;
+        public void start() {
+            Vex.this.setTarget(Vex.this.owner.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit
+            super.start();
         }
 
         @Override
