--- a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
@@ -19,6 +20,13 @@
 import net.minecraft.world.level.block.BarrelBlock;
 import net.minecraft.world.level.block.state.BlockState;
 
+// CraftBukkit start
+import java.util.ArrayList;
+import java.util.List;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+// CraftBukkit end
+
 public class BarrelBlockEntity extends RandomizableContainerBlockEntity {
     private NonNullList<ItemStack> items = NonNullList.withSize(27, ItemStack.EMPTY);
     private final ContainerOpenersCounter openersCounter = new ContainerOpenersCounter() {
@@ -28,11 +29,9 @@
             BarrelBlockEntity.this.updateBlockState(state, true);
         }
 
-        @Override
-        protected void onClose(Level level, BlockPos pos, BlockState state) {
-            BarrelBlockEntity.this.playSound(state, SoundEvents.BARREL_CLOSE);
-            BarrelBlockEntity.this.updateBlockState(state, false);
-        }
+    // CraftBukkit start - add fields and methods
+    public List<HumanEntity> transaction = new ArrayList<>();
+    private int maxStack = MAX_STACK;
 
         @Override
         protected void openerCountChanged(Level level, BlockPos pos, BlockState state, int count, int openCount) {
@@ -49,7 +43,30 @@
         }
     };
 
-    public BarrelBlockEntity(BlockPos pos, BlockState blockState) {
+    @Override
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    @Override
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+       return maxStack;
+    }
+
+    @Override
+    public void setMaxStackSize(int i) {
+        maxStack = i;
+    }
+    // CraftBukkit end
+    private NonNullList<ItemStack> items;
+    public final ContainerOpenersCounter openersCounter;
+
+    public BarrelBlockEntity(BlockPos pos, IBlockData blockState) {
         super(BlockEntityType.BARREL, pos, blockState);
     }
 
