--- a/net/minecraft/world/level/block/entity/ChestBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChestBlockEntity.java
@@ -22,6 +22,11 @@
 import net.minecraft.world.level.block.ChestBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.properties.ChestType;
+// CraftBukkit start
+import java.util.List;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+// CraftBukkit end
 
 public class ChestBlockEntity extends RandomizableContainerBlockEntity implements LidBlockEntity {
     private static final int EVENT_SET_OPEN_COUNT = 1;
@@ -32,10 +35,9 @@
             ChestBlockEntity.playSound(level, pos, state, SoundEvents.CHEST_OPEN);
         }
 
-        @Override
-        protected void onClose(Level level, BlockPos pos, BlockState state) {
-            ChestBlockEntity.playSound(level, pos, state, SoundEvents.CHEST_CLOSE);
-        }
+    // CraftBukkit start - add fields and methods
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
 
         @Override
         protected void openerCountChanged(Level level, BlockPos pos, BlockState state, int count, int openCount) {
@@ -55,7 +47,25 @@
     };
     private final ChestLidController chestLidController = new ChestLidController();
 
-    protected ChestBlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState blockState) {
+    public void onClose(CraftHumanEntity who) {
+        transaction.remove(who);
+    }
+
+    public List<HumanEntity> getViewers() {
+        return transaction;
+    }
+
+    @Override
+    public int getMaxStackSize() {
+        return maxStack;
+    }
+
+    public void setMaxStackSize(int size) {
+        maxStack = size;
+    }
+    // CraftBukkit end
+
+    protected ChestBlockEntity(BlockEntityType<?> type, BlockPos pos, IBlockData blockState) {
         super(type, pos, blockState);
     }
 
@@ -182,4 +233,11 @@
         Block block = state.getBlock();
         level.blockEvent(pos, block, 1, eventParam);
     }
+
+    // CraftBukkit start
+    @Override
+    public boolean onlyOpCanSetNbt() {
+        return true;
+    }
+    // CraftBukkit end
 }
