--- a/net/minecraft/world/level/block/entity/SkullBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SkullBlockEntity.java
@@ -184,10 +189,21 @@
         }
     }
 
-    public static void resolveGameProfile(CompoundTag compoundTag) {
-        String string = compoundTag.getString("SkullOwner");
-        if (!Util.isBlank(string)) {
-            resolveGameProfile(compoundTag, string);
+    public static void resolveGameProfile(CompoundTag nbttagcompound) {
+        String s = nbttagcompound.getString("SkullOwner");
+
+        if (!Util.isBlank(s)) {
+            resolveGameProfile(nbttagcompound, s);
+            // CraftBukkit start
+        } else {
+            net.minecraft.nbt.ListTag textures = nbttagcompound.getCompound("SkullOwner").getCompound("Properties").getList("textures", 10); // Safe due to method contracts
+            for (int i = 0; i < textures.size(); i++) {
+                if (textures.get(i) instanceof CompoundTag && !((CompoundTag) textures.get(i)).contains("Signature", 8) && ((CompoundTag) textures.get(i)).getString("Value").trim().isEmpty()) {
+                    nbttagcompound.remove("SkullOwner");
+                    break;
+                }
+            }
+            // CraftBukkit end
         }
     }
 
