From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MiniDigger | Martin <admin@benndorf.dev>
Date: Sat, 6 Jan 2024 15:41:33 +0100
Subject: [PATCH] Example Feature Patch


diff --git a/io/papermc/paper/Feature.java b/io/papermc/paper/Feature.java
new file mode 100644
index 0000000000000000000000000000000000000000..8f5c666bb2897b6e6515fb6f8b085797751d71dc
--- /dev/null
+++ b/io/papermc/paper/Feature.java
@@ -0,0 +1,4 @@
+package io.papermc.paper;
+
+public class Feature {
+}
