--- a/net/minecraft/world/item/crafting/Ingredient.java
+++ b/net/minecraft/world/item/crafting/Ingredient.java
@@ -35,6 +35,7 @@
     private ItemStack[] itemStacks;
     @Nullable
     private IntList stackingIds;
+    public boolean exact; // CraftBukkit
     public static final Codec<Ingredient> CODEC = codec(true);
     public static final Codec<Ingredient> CODEC_NONEMPTY = codec(false);
 
@@ -73,7 +73,16 @@
             for (int j = 0; j < i; ++j) {
                 ItemStack itemstack1 = aitemstack[j];
 
-                if (itemstack1.is(itemstack.getItem())) {
+                // CraftBukkit start
+                if (exact) {
+                    if (itemstack1.getItem() == stack.getItem() && ItemStack.isSameItemSameTags(stack, itemstack1)) {
+                        return true;
+                    }
+
+                    continue;
+                }
+                // CraftBukkit end
+                if (itemstack1.is(stack.getItem())) {
                     return true;
                 }
             }
