--- a/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
+++ b/net/minecraft/world/entity/ai/behavior/HarvestFarmland.java
@@ -19,10 +21,15 @@
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.state.IBlockData;
+import net.minecraft.world.level.gameevent.GameEvent;
+
+// CraftBukkit start
+import net.minecraft.world.level.block.Blocks;
 import net.minecraft.world.level.block.CropBlock;
 import net.minecraft.world.level.block.FarmBlock;
-import net.minecraft.world.level.block.state.BlockState;
-import net.minecraft.world.level.gameevent.GameEvent;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
 
 public class HarvestFarmland extends Behavior<Villager> {
     private static final int HARVEST_DURATION = 200;
@@ -109,8 +105,11 @@
                 BlockState blockState = level.getBlockState(this.aboveFarmlandPos);
                 Block block = blockState.getBlock();
                 Block block1 = level.getBlockState(this.aboveFarmlandPos.below()).getBlock();
-                if (block instanceof CropBlock && ((CropBlock)block).isMaxAge(blockState)) {
+
+                if (block instanceof CropBlock && ((CropBlock) block).isMaxAge(iblockdata)) {
+                    if (CraftEventFactory.callEntityChangeBlockEvent(owner, this.aboveFarmlandPos, Blocks.AIR.defaultBlockState())) { // CraftBukkit
                     level.destroyBlock(this.aboveFarmlandPos, true, owner);
+                    } // CraftBukkit
                 }
 
                 if (blockState.isAir() && block1 instanceof FarmBlock && owner.hasFarmSeeds()) {
@@ -119,11 +118,20 @@
                     for (int i = 0; i < inventory.getContainerSize(); i++) {
                         ItemStack item = inventory.getItem(i);
                         boolean flag = false;
-                        if (!item.isEmpty() && item.is(ItemTags.VILLAGER_PLANTABLE_SEEDS) && item.getItem() instanceof BlockItem blockItem) {
-                            BlockState blockState1 = blockItem.getBlock().defaultBlockState();
-                            level.setBlockAndUpdate(this.aboveFarmlandPos, blockState1);
-                            level.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(owner, blockState1));
-                            flag = true;
+
+                        if (!itemstack.isEmpty() && itemstack.is(ItemTags.VILLAGER_PLANTABLE_SEEDS)) {
+                            Item item = itemstack.getItem();
+
+                            if (item instanceof BlockItem) {
+                                BlockItem itemblock = (BlockItem) item;
+                                IBlockData iblockdata1 = itemblock.getBlock().defaultBlockState();
+
+                                if (CraftEventFactory.callEntityChangeBlockEvent(owner, this.aboveFarmlandPos, iblockdata1)) { // CraftBukkit
+                                level.setBlockAndUpdate(this.aboveFarmlandPos, iblockdata1);
+                                level.gameEvent(GameEvent.BLOCK_PLACE, this.aboveFarmlandPos, GameEvent.Context.of(owner, iblockdata1));
+                                flag = true;
+                                } // CraftBukkit
+                            }
                         }
 
                         if (flag) {
