--- a/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerMakeLove.java
@@ -16,6 +17,10 @@
 import net.minecraft.world.entity.ai.village.poi.PoiTypes;
 import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.level.pathfinder.Path;
+// CraftBukkit start
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+// CraftBukkit end
 
 public class VillagerMakeLove extends Behavior<Villager> {
     private static final int INTERACT_DIST_SQR = 5;
@@ -113,13 +120,19 @@
         if (breedOffspring == null) {
             return Optional.empty();
         } else {
+            entityvillager2.setAge(-24000);
+            entityvillager2.moveTo(parent.getX(), parent.getY(), parent.getZ(), 0.0F, 0.0F);
+            // CraftBukkit start - call EntityBreedEvent
+            if (CraftEventFactory.callEntityBreedEvent(entityvillager2, parent, partner, null, null, 0).isCancelled()) {
+                return Optional.empty();
+            }
+            // Move age setting down
             parent.setAge(6000);
             partner.setAge(6000);
-            breedOffspring.setAge(-24000);
-            breedOffspring.moveTo(parent.getX(), parent.getY(), parent.getZ(), 0.0F, 0.0F);
-            level.addFreshEntityWithPassengers(breedOffspring);
-            level.broadcastEntityEvent(breedOffspring, (byte)12);
-            return Optional.of(breedOffspring);
+            level.addFreshEntityWithPassengers(entityvillager2, CreatureSpawnEvent.SpawnReason.BREEDING);
+            // CraftBukkit end
+            level.broadcastEntityEvent(entityvillager2, (byte) 12);
+            return Optional.of(entityvillager2);
         }
     }
 
