--- a/net/minecraft/server/commands/ScheduleCommand.java
+++ b/net/minecraft/server/commands/ScheduleCommand.java
@@ -55,9 +53,9 @@
         if (i == 0) {
             throw ScheduleCommand.ERROR_SAME_TICK.create();
         } else {
-            long j = commandsourcestack.getLevel().getGameTime() + (long) i;
-            ResourceLocation resourcelocation = (ResourceLocation) pair.getFirst();
-            TimerQueue<MinecraftServer> timerqueue = commandsourcestack.getServer().getWorldData().overworldData().getScheduledEvents();
+            long j = source.getLevel().getGameTime() + (long) time;
+            ResourceLocation minecraftkey = (ResourceLocation) function.getFirst();
+            TimerQueue<MinecraftServer> customfunctioncallbacktimerqueue = source.getLevel().serverLevelData.overworldData().getScheduledEvents(); // CraftBukkit - SPIGOT-6667: Use world specific function timer
 
             ((Either) pair.getSecond()).ifLeft((commandfunction) -> {
                 String s = resourcelocation.toString();
