--- a/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
+++ b/net/minecraft/world/entity/vehicle/AbstractMinecartContainer.java
@@ -17,6 +17,13 @@
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.level.Level;
+// CraftBukkit start
+import java.util.List;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
+// CraftBukkit end
 
 public abstract class AbstractMinecartContainer extends AbstractMinecart implements ContainerEntity {
 
@@ -25,9 +32,12 @@
     private ResourceLocation lootTable;
     private long lootTableSeed;
 
-    protected AbstractMinecartContainer(EntityType<?> entitytype, Level level) {
-        super(entitytype, level);
-        this.itemStacks = NonNullList.withSize(36, ItemStack.EMPTY);
+    // CraftBukkit start
+    public List<HumanEntity> transaction = new java.util.ArrayList<HumanEntity>();
+    private int maxStack = MAX_STACK;
+
+    public List<ItemStack> getContents() {
+        return this.itemStacks;
     }
 
     protected AbstractMinecartContainer(EntityType<?> entitytype, double d0, double d1, double d2, Level level) {
@@ -41,7 +71,18 @@
         super.destroy(damagesource);
         this.chestVehicleDestroyed(damagesource, this.level(), this);
     }
+    // CraftBukkit end
 
+    protected AbstractMinecartContainer(EntityType<?> entityType, Level level) {
+        super(entityType, level);
+        this.itemStacks = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY); // CraftBukkit - SPIGOT-3513
+    }
+
+    protected AbstractMinecartContainer(EntityType<?> entityType, double x, double d1, double y, Level world) {
+        super(entityType, world, x, d1, y);
+        this.itemStacks = NonNullList.withSize(this.getContainerSize(), ItemStack.EMPTY); // CraftBukkit - SPIGOT-3513
+    }
+
     @Override
     @Override
     public ItemStack getItem(int i) {
